---
title: "Analyse sRNA populations using mobileRNA"
author: Katie Jeynes-Cupper, Marco Catoni
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
  toc: true
toc_depth: 3
number_sections: true
theme: united
highlight: tango
fig_width: 7
vignette: >
 %\VignetteIndexEntry{mobileRNA_standard}
 %\VignetteEngine{knitr::rmarkdown}}
 %\VignetteEncoding{UTF-8}
 %\VignettePackage{mobileRNA}
 %\VignetteDepends{mobileRNA}
 %\usepackage[utf8]{inputenc}

---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  fig.align='center',
  external=TRUE,
  echo=TRUE,
  warning=FALSE,
  comment = "#>"
)
```


# Introduction

In plants, systemic signalling is an elaborated molecular system which allows to
coordinate plant development at the entire organism level, integrating and
transmitting the information perceived from environment to distant organs.
RNA molecules play a key role in this extensive network, both across 
short and long distances. Of which, small RNAs (sRNA) have been shown to be
involved in post-transcriptional gene silencing and RNA-directed DNA methylation.
Alterations such as these have been linked to changes in gene expression and 
phenotypical changes or trait. Here, we introduce the use of the mobileRNA 
R package as a tool for sRNAseq analysis to identify sRNA molecules in a 
biological system, regardless of whether the system is chimeric or not. 

While, other RNA molecules, such as messenger RNA, are also renown be play a 
role in systemic signalling this tool currently focuses on the identification 
of local sRNA populations. 
<br>


# Method
Small RNAs (sRNAs) range from 20-24 nucleotides in length, but regardless of 
their short length the different groups have very specific roles in genetic 
regulation in plants. There are two key groups of interest; micro-RNAs (miRNAs)
with a size between 21-22nt and small-interfering-RNAs (siRNAs) with a size of 
24nt. miRNAs are renown to play a role in post-transcriptional regulation, 
while siRNAs are key components in the RNA-directed DNA methylation pathway. 
Therefore, understanding the population dynamics of sRNAs within your
biological system could shed light on gene expression and DNA methylation 
patterns, bringing light to the molecular mechanisms underpinning biological 
processes. 

The workflow can be utilise to explore sRNA populations in any species, and 
can be utilise further if you have a treatment vs. control experimental design.
The workflow will aid the identification of unique sRNA populations and 
differences in sRNA abundances between treatment and control conditions. 

This guide suggests mapping steps, analysis steps, and functional analysis 
steps.

# Installation

The latest version of `mobileRNA` can be installed via GitHub
using the `devtools` package:

```{r,  eval=FALSE, message=FALSE}
if (!require("devtools")) install.packages("devtools")
devtools::install_github("KJeynesCupper/mobileRNA", ref = "main")

library(mobileRNA)
```

# Pre-processing 
Assuming your sRNAseq raw samples have been cleaned to remove low quality reads 
and remove adapter sequences, we can align the samples to the genome assembly
and detect sRNA clusters using the Linux `ShortStack` program
(https://github.com/MikeAxtell/ShortStack). 

## Auto-Detection of sRNA Cluster  

### Step 1 - Cluster analysis with ShortStack

``` bash
ShortStack \
--readfile <control_1.fastq> \
--genomefile <merged_reference.fa> \
--bowtie_cores 6 \
--mmap n \
--mismatches 0 \
--nohp \
--outdir <./output/directory>

```


#### Step 2 - Build sRNA cluster list 

Now, we collate all the sRNA loci information from each sample into a text file. 

```{r, eval = FALSE}
# location of step 1 output
folder <- "./output/directory/from/step/1/"

# name and location to save output file to (must be .txt)
save_folder <- "./output/directory/ClustersInfo.txt"

# names of samples (ie. folder names)
sample_names <- c("<treatment_1>", "<treatment_2>", "<control_1>","<control_2>")


loci_info <- RNAloci(files = folder, 
             out = save_folder,
             samples = sample_names)
```


## Mapping   
Each sample is mapped to the merged reference genome with the list of sRNA 
clusters. 


``` bash
ShortStack \
--readfile <control_1.fastq> \
--genomefile <merged_reference.fa> \
--locifile <./output/directory/ClustersInfo.gff3> \
--bowtie_cores 6 \
--mmap n \
--mismatches 0 \
--nohp \
--mincov 5 \
--outdir <./output/directory/step2/>

```

# Analysis 
The aligned data can now be analysed in R, and the native population of sRNAs
can be explored. This following analysis workflow is designed to facilitate a 
treatment vs control experimental design. 

For the analysis we will use the semi-synthetic data, although there are two 
different genotypes in this analysis it can still be utilised to show 
different sRNA population dynamics. 


```{r Load, message=FALSE}
library(mobileRNA)
data("sRNA_data")

```


## Step 1: Import data 
State the type of data (which is "sRNA" here), the location of the mapping 
results in your directory, and the sample names. 
```{r, eval = FALSE}
# Directory containing results
results_dir <-  "<./output/directory/step2/>"

# Sample names and total number of reads, in the same order. 
sample_names <- c("heterograft_1", "heterograft_2" , "heterograft_3", 
                  "selfgraft_1", "selfgraft_2", "selfgraft_3")


sRNA_data <- RNAimport(input = "sRNA", 
                       directory = results_dir,
                       samples = sample_names)
                           
```
<br>


## Step 2: Quality control 
A handy step in the analysis is to assess the overall similarity and difference
between samples within a conditions and between the conditions. This is to 
identify where most of the variation is introduced in the data set, as well as 
understanding whether the data set meets your expectations, quality-wise. 
It is expected that between the conditions, the sample replicates show enough 
variation to suggest that the replicates are from different groups. Here we
will use the following methods: 
* Distribution of RNA classes within each sample
* Principle component analysis (PCA)
* Heatmap using Hierarchical clustering 

This will show us how well samples within each condition cluster together, 
which may highlight outliers. Plus, to show whether our experimental conditions 
represent the main source of variation in the data set.

### Principle component analysis (PCA)
Principal Component Analysis (PCA) is a useful technique to illustrate sample 
distance as it emphasizes the variation through the reduction of dimensions in 
the data set. Here, we introduce the function plotSamplePCA():

```{r, results='hide', message=FALSE}
group <- c("Heterograft", "Heterograft", "Heterograft",
            "Selfgraft", "Selfgraft", "Selfgraft")

plotSamplePCA(sRNA_data, group)
```


### Hierarchical clustered heatmap to assess sample distance
Similarly, to a PCA plot, the plotSampleDistance() function undertakes 
hierarchical clustering with an unbiased log transformation to calculate sample 
distance and is plotted in the form of a heatmap.

```{r, results='hide', message=FALSE}
plotSampleDistance(sRNA_data)
```


### Plot the distribution of RNA classes within each sample
Explore the number of each RNA class identified within each sample using the 
`RNAdistribution` function which produces a table and plots the results. 
The results can be plotted as either a bar chart (style = "bar") or a line graph 
(style = "line") for each sample and can be shown in a bar chart facet 
(facet = TRUE) or in as a single line graph, where each line represents a sample 
(together=TRUE). Alternatively, the results can be plotted individually for each 
sample in either format.

```{r}
sample_distribution <- RNAdistribution(sRNA_data,
                                            style = "line",
                                            together = FALSE)
```

Lets look at the plot and the data:
```{r, fig.show="hold"}
# view plot 
sample_distribution$plot
```


```{r, fig.show="hold"}
# view data 
sample_distribution$data

```

## Step 3: Calculate the consensus of each sRNA cluster  
For a given sRNA cluster, each replicate has determined the class (20-24nt) 
based on the most abundant small RNA size. Replicates within the same condition 
are expected to class a given sRNA similarly. 

The `RNAdicercall()` function is used to define the dicer class of a sRNA 
cluster based on the consensus across (specific) replicates. 


```{r}
# calculate dicercall consensus 
sRNA_data_summary <- RNAdicercall(data = sRNA_data,  
                                 tidy=TRUE)
# plot results 
plot_distribution <- RNAdistribution(data = sRNA_data_consensus, 
                        consensus = TRUE)
# View figure
plot_distribution$plot
```


## Step 4: Statistical analysis using DESeq2 or edgeR
Statistical analysis can be helpful to understand the differential abundances of 
sRNA clusters found within your system. Here we can utilize either the DESeq2 or 
edgeR algorithms to support your analysis, and can be used in the next steps. 


```{r, results='hide', message=FALSE}
# sample conditions in order within dataframe
groups <- c("Heterograft", "Heterograft", "Heterograft",
            "Selfgraft", "Selfgraft", "Selfgraft")

## Differential analysis using the DEseq2 method 
sRNA_DESeq2 <- RNAanalysis(data = sRNA_data_summary,
                              group = groups,
                              method = "DESeq2")

## Differential analysis using the edgeR method 

sRNA_edgeR <- RNAanalysis(data = sRNA_data_summary,
                              group = groups,
                              method = "edgeR")
```

This will add differential analysis values for each sRNA cluster to the working
dataframe. We can then easily manipulate this data and plot it as heatmap. 


### Plot a volcano plot 
A Volcano plots can be plotted to show the relationship between fold change and 
the adjusted p-value (i.e. FDR) across the dataset. Each point represents an 
individual sRNA cluster. Pointes are coloured according to their significance 
level thresholds (0.05, 0.01, and 0.001 or unchanged). 



```{r}
# plot with the top 10 clusters labelled
vol_plot <- plotVolcano(data = sRNA_DESeq2, labels = TRUE)

vol_plot$plot
```

## Step 5: Explore difference in sRNA abundances
When comparing treatment to control conditions, it might be the case that 
the same sRNA clusters are found within both yet there could be difference in 
the total abundance of the shared clusters. For instance, for a given sRNA 
cluster the samples in the treatment condition might have a greater abundance 
than the samples in the control condition. These difference could attribute to 
observed differences. 

The statistical analysis calculated the log2FC values for each sRNA cluster by 
comparing the normalised counts between treatment and control. Here, 
a positive log2FC indicates an increased abundance of transcripts for a given 
sRNA cluster in the treatment compared to the control, while negative log2FC 
indicates decreased abundance of transcripts for a given sRNA cluster. The 
statistical significance of the log2FC is determined by the (adjusted) pvalue
also calculated in the previous step. 

Here we will filter the data to select sRNA clusters which are statistically 
significant, and then plot the results as a heatmap to compare the conditions. 

```{r, fig.show='hide'}
# select only signifcant sRNAs
significant_sRNAs <- RNAsignificant(sRNA_DESeq2)

#plot
p1 <- plotHeatmap(significant_sRNAs, row.names = FALSE)

```


Lets look at the plot:

```{r}
p1$plot 
```
<br>


## Step 6: Identify absolute sRNA population difference 
Finally, the  `RNApopulation()` function can be utilised to identify unique 
sRNA populations found in the treatment or control condition.

First lets look at the sRNA clusters unique to the treatment condition:
```{r}
# select sRNA clusters only found in treatment & not in the control samples
treatment_reps <- c("heterograft_1", "heterograft_2" , "heterograft_3")
unique_treatment <- RNApopulation(data = sRNA_data_consensus, 
                                   conditions = treatment_reps)

# look at number of sRNA cluster only found in treatment 
nrow(unique_treatment)
```


Now, the sRNA clusters unique to the control condition:
```{r}
# select sRNA clusters only found in control & not in the treatment samples
control_reps <- c("selfgraft_1", "selfgraft_2" , "selfgraft_3")
unique_control <- RNApopulation(data = sRNA_data_consensus, 
                                   conditions = control_reps)
# look at number of sRNA cluster only found in control  
nrow(unique_control) 
```
<br>


### Plotting heterograft-only sRNA populations
Here we located 148 sRNA clusters which are uniquely present to the 
heterografts, and not found in the self-graft controls. These results can 
be plotted as a distribution plot and heatmap. 

```{r, fig.show='hide'}
# plot distribution of sRNA classes only found in treatment 
p2 <- RNAdistribution(data = unique_treatment, style = "line", together = FALSE)

# plot heatmap of treatment-only sRNA clusters
p3 <- plotHeatmap(data = unique_treatment, row.names = FALSE)

```


Lets look at the plots:

Here we can see the distribution of sRNA classes only found in treatment 
```{r}
p2$plot
```

Here we can compare the difference and similarities between the treatment 
and controls using a heatmap: 
```{r}
p3$plot
```


## Step 6: Functional analysis 

### Target sequences
sRNAs target complementary sequences in the genome to mediate silencing via 
post-transcriptional or pre-transcriptional mechanisms. Here we will identify 
the complementary sequences matching the most abundant sRNA within a cluster. 

```{r}
target_sequences <- RNAsequences(unique_treatment)
```

This dataframe can be exported the results loaded into BLAST to identify 
regions within the genome that these complementary sequences are present. There 
are tools, such as `TargetFinder` or `sRNATargetDigger`, which use a predictive 
algorithm to identify the most likely locations in the genome where the sRNAs 
might target. 

Lets export the data frame :
```{r, eval=FALSE}

write.table(target_sequences, "./data/mobile_sequences.txt")

```

### Overlapping regions
Finally, we can identify regions in the genome annotation where the sRNA cluster
overlaps with based on the chromosome and the start/end coordinated. This can 
be explored using two functions: [RNAattributes()] and [RNAfeatures()].


The `RNAattributes` function compared a genome annotation with the sRNA clusters
and, if an overlap is present, the information is joined. This is accomplished 
by adding the the information from the annotation file to the sRNA working 
dataframe as additional columns. 

```{r, eval = FALSE}
attributes <- RNAattributes(data = unique_treatment,
                            annotation = "./annotation/origin_annotation.gff3")
```

The `RNAfeatures` functions is similar to the `RNAattributes` function, but 
instead it returns either the absolute number or percentage of sRNA clusters 
which overlap with specific regions in the genome such as repeats, introns, 
exons,  UTR and promoter regions. It could be helpful to subset your dataset
to only explore a specific sRNA class, see [mobileRNA::RNAsubset()] for more
information. 


```{r, eval = FALSE}
features <- RNAfeatures(data = unique_treatment,
                            annotation = "./annotation/origin_annotation.gff3", 
                      repeats = "./annotation/origin_annotation_repeats.gff3")

```
<br>



# Session information
```{r}
sessionInfo()
```