---
title: "Overview of the mobileRNA R package for mobile sRNA identification between genotypes"
author: Katie Jeynes-Cupper, Marco Catoni
date: "`r Sys.Date()`"
output:
  html_document:
  toc: true
toc_depth: 3
number_sections: true
theme: united
highlight: tango
fig_width: 7
vignette: >
 %\VignetteIndexEntry{mobileRNA}
 %\VignetteEngine{knitr::rmarkdown}}
 %\VignetteEncoding{UTF-8}
 %\VignettePackage{mobileRNA}
 %\VignetteDepends{mobileRNA}
 %\usepackage[utf8]{inputenc}

---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  fig.align='center',
  external=TRUE,
  echo=TRUE,
  warning=FALSE,
  comment = "#>"
)
```


# Introduction
In plants, systemic signalling is an elaborated molecular system which allows to
coordinate plant development at the entire organism level, integrating and
transmitting the information perceived from environment to distant organs.
An important role in long-distance signalling is played by small RNA molecules
(sRNAs). These sRNAs can be produced in a tissue and then being transported
systemically across the vascular system into recipient organs, where they can
induce a molecular response and coordinate physiological changes.
Being mobile sRNA able to move across a graft junction, their study can bring to
novel methods for transferring traits into crops, for example by combining
genotypes by grafting, rather than using breeding or transgenic approaches.
Several genomics studies used grafting to explore and
characterise mobile sRNAs, however, the approach used to distinguish these
mobile molecules from the sequencing noise is not standardised, and for
each study it is necessary to adapt general RNA sequencing pipelines to the
specific study of mobile molecules.  Here we introduce the R Package
\code[mobileRNA], to aid the identification of mobile RNA molecules in a plant
graft system, enable the visualization of the results, and perform functional
analysis. The package generates mobile molecules through a consensus
classification system, in hand with the use of differential expression analysis
and can achieve results on a range of RNA molecule types.


# Methods
This manual offers a pipeline for the analysis of sRNAseq data from plant
grafting experiments, specifically to identify the movement of sRNA molecules
across a graft junction from one genotype to another.

The pre-processing method is different to others, as it involves mapping the
samples to a merged reference genome, comprising of the two genotypes in the
heterograft system. In addition, all samples are analysed against a list of all
identified sRNA cluster loci to ensure that all sRNA clusters are fairly
recognised in the analysis

The analysis method takes the results produced by ShortStack mapping and cluster
analysis and identified potential mobile sRNA by removing those mapped to the
destination tissue genotype and removing additional falsely mapped sRNA.
Previous methods have not employed statistical analysis to select mobile sRNA,
here we have incorporated this into the analysis options. This adds greater
flexibility and control to the user. By the end of the analysis using mobileRNA,
the user could have produced a number of figures to represent the data. This
includes plots to represent the:
  * Distribution of sRNA sizes across samples, total dataset consensus sRNA,
or across the potential mobile sRNA.
* Distribution of sRNA across genomic features
* Sample distance, including a PCA and heatmap
* Heatmap of mobile sRNA expression



# Installation

The latest version of `mobileRNA` can be installed via GitHub
using the `devtools` package:

```{r,  eval=FALSE, message=FALSE}

if (!require("devtools")) install.packages("devtools")
devtools::install_github("KJeynesCupper/mobileRNA")
```


# Installation of Linux Dependencies
The majority of the pre-processing functions require a system call
to Linux software. The following dependencies need to be installed before you
begin:

- `FastQC` (https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
- `bedtools` (https://bedtools.readthedocs.io/en/latest/)
- `Samtools` (http://www.htslib.org/)
- `sed` (https://www.gnu.org/software/sed/manual/sed.html)
- `ShortStack` (https://github.com/MikeAxtell/ShortStack)


# Load the package and the example data into R session

```{r Load, message=FALSE}

library(mobileRNA)
data("sRNA_data")

```

Once installed, you will have access to the following functions:

  **Pre-processing**
- `RNAmergeGenomes()`
- `RNAmergeAnnotations()`
- `RNAloci()`

**Analysis**
- `RNAimport()`
- `RNAconsensus()`
- `RNAsubset()`
- `RNAdistribution()`
- `plotSamplePCA()`
- `plotSampleDistribution()`
- `RNAanalysis()`
- `RNAmobile()`
- `plotHeatmap()`
- `RNAmean()`
- `RNAfeatures()`
- `RNAattributes()`



Further details on how each function operates will be provided later on in the
documentation. For some functions there are recurring arguments.
In the pre-processing steps, the `files` argument occurs frequently and tells
the function a local directory path to a folder holding a particular file(s).
Similarly, the argument `out` instructs the function of the location to save the
output file(s).  In the analysis steps, some important arguments include `samples`
which refers to the a character vector contains the names of all samples in the
experiment, `conditions` which refers to a character vector containing a select
few samples depending on your analysis preference, `groups` which refers to
a character vector of the two experimental conditions and lastly `controls`
which refers to the control samples in the analysis. A recurring argument in the
analysis pipeline is the `method` argument which tells the function
how to process the data from the choice of two analytical methods
(i.e.`"DESeq"`, or `"edgeR"`). All other arguments are discussed
in further detail by looking at the a functions help file.

## An overview of the data used
For the following examples, an example small RNA-seq data set from a Tomato
(**Solanium lycopersicum**) & Eggplant (**Solanium melongena**) grafting
experiment was utilised. The heterografts are composed of an tomato scion and
a eggplant rootstock. While in the self-graft, tomato has been used as the
control with the aim to analyse whether a the eggplant rootstock can send
sRNA molecules across the graft junction to the tomato scion.

The data set, called `sRNA_data`, stores a matrix containing the
pre-processed data from the experiment. As a
user, this allows you to see what a full data set might look like and how
you might expect it to change across the analysis.

These can be loaded in these R workspace by using the following command:


  ``` r
data("sRNA_data")
```



## Data organisation
There are Two key elements required for the pipeline analysis:
* sRNA-seq sample replicates  (.fasta/.fq)
* Two reference genomes (.fasta/.fa)

And one additional element; to improve functional analysis:
  * Two reference genome annotations (.gff/.gff3)

### The raw data files

It is recommended to rename your files to names you wish for them to
be represented as within the analysis and shown as labels in plots.
Plus, it makes the analysis easier!

  For example, instead of names such as:
* `sample_1.fq`
* `sample_2.fq`
* `sample_3.fq`
* `sample_4.fq`
* `sample_5.fq`
* `sample_6.fq`

For the example data set included in the package, here we have renamed the
files based on the condition (treatment or control).
For the hetero-grafts, where the is a tomato scion and an eggplant rootstock:
* `TomEgg_1.fq`
* `TomEgg_2.fq`
* `TomEgg_3.fq`

and for the tomato self-grafts:
* `TomTom_1.fq`
* `TomTom_2.fq`
* `TomTom_3.fq`


### The reference genomes and annotation files (.fasta/.fa)
The pipeline is designed to analyse grafting systems with two
distinct genomes, here tomato and eggplant.


# Pre-Processing
The pre-processing step involves cleaning raw data and aligning data to the
merged genome.  We recommend installing and using the `ShortStack` (https://github.com/MikeAxtell/ShortStack)
program to aligns samples to the merged genome and undertake cluster analysis.
This program is specifically formulated for sRNAseq analysis, utilising Bowtie
(Version 1) to map samples and a specifically formulated algorithm to cluster
sRNAs.

Here, we introduce an alternative mapping method for the analysis of plant
heterograft samples. The heterograft system involves two genotypes; here
the two genome references are merged into a single reference to which
samples are aligned to.

`mobileRNA` offers a function to merge two FASTA reference genomes into one.
To distinguish between the reference genomes in a merged file, it is important
to make sure the chromosome names between the genomes are different and
distinguishable. The function below added a particular character string to the
start of each chromosome name in each reference genome. As standard, the string
"A_" is added to the reference genome supplied to "genomeA" and "B_" is added
to the reference genome supplied to "genomeB". These can be customised to the
users preference, see manual for more information.


Going forward, the pipeline has assumed that the sRNA-seq samples have met quality
control standards.

## Pre-mapping

### Preparing the merged genome
Here we merge the two reference genomes into a single merged genome using the
[mobileRNA::RNAgenomeMerge()] function. As default, this function changes the
chromosome names of each genome to ensure they are distinguishable. To do so,
the function requires an input of the initials of each organism's Latin name.
But why do this:

* If the two genomes use the same pattern to name the chromosomes, the user will
not be able to differentiate the chromosomes from one another in the merged
genome. This could be solves by adding letters to the chromosomes of one of the
genomes, for example, "SM" to represent the Latin name of eggplant.

* If a chromosome naming pattern contains punctuation, the mapping step will not
work.


In the example, the **Solanum lycopersicum** (version 4) genome contains a
full-stop/period within each chromosome name which needs to be removed as well.
Here we rename the chromosomes of the tomato genome to `SL` and the chromosomes
related to the eggplant genome (**Solanum melongena**) to `SM`.

```{r, eval = FALSE}
RNAmergeGenomes(genomeA = "/Users/user1/projectname/workplace/reference/ref1.fa",
               genomeB = "/Users/user1/projectname/workplace/reference/ref2.fa",
               abbreviationGenomeA = "SL",
               abbreviationGenomeB = "SM",
               out_dir = "/Users/user1/projectname/workplace/reference/merge/
               ref_merged.fa")
```




## Identify genomic locations of sRNA
To identify associates between sRNA clusters and genomic features, the genomic
annotation files (GFF) of each reference genome must be merged in the same way
the reference genomes were. This includes following the same naming patterns for
the chromosomes in each genome as undertaken earlier with the `RNAmergeGenomes()`
function. To undertake a similar merging between the two GFF files, use the
`RNAmergeAnnotations` function.

The overlap between the merged annotation and the loci of clusters can be undertaken
later in analysis using the functions `RNAfeatures` and `RNAattributes`. Where
`RNAfeatures` computes the proportion of sRNA overlapping specific features,
such as 2kb promoter regions, transposable element regions and exons/introns.
While the `RNAattributes` function, adds the specific genomic features or regions
a sRNA cluster exactly overlaps with to the working dataset.

```{r, eval = FALSE}
## merge the annotation files into a single annotation file
RNAmergeAnnotation(annotationA = "/Users/user1/projectname/workplace/annotation/annotation_1.gff3",
                   annotationB = "/Users/user1/projectname/workplace/annotation/annotation_1.gff3",
                   abbreviationAnnoA = "SL",
                   abbreviationAnnoB = "SM"
                  out_dir = "/Users/user1/projectname/workplace/annotation/merge/anno_merged.gff")

```



## Mapping
The mapping occurs in 3 steps:

* *Step 1 :* Aligns samples with the merged reference genome to identify the
locations of the sRNA clusters in each sample.
* *Step 2 :* Creates a plain text file and .gff3 file containing all the
identified sRNA clusters across samples.
* *Step 3:* Aligns the samples to the merged reference genome with the sRNA
cluster location.


The analysis pipeline is formulated to analyse mapping and clustering results
produced by `ShortStack` (https://github.com/MikeAxtell/ShortStack).

To distinguish between the reference genomes in a merged file, it is important
to make sure the chromosome names between the genomes are different and
distinguishable.

Here, we recommend a double-mapping process using `ShortStack`, the steps are
as follow:

### Step 1 - Identify loci of dicer-derived sRNA cluster in each sample

The bash script below suggests an appropriate method to undertaking the first
mapping step using `ShortStack`. The samples are mapped uniquely ( ie. no
multi-mapping) to the merged genome. This script can easily be converted into a
for loop.

``` bash
ShortStack \
--readfile <control_1.fastq> \
--genomefile <merged_reference.fa> \
--bowtie_cores 6 \
--mmap n \
--mismatches 0 \
--nohp \
--outdir <./output/directory>

```

### Step 2 - Create annotation file containing all identified sRNA cluster loci across samples

From the output of Step 1, the `RNAloci()` function can collate all
the loci information into a single `txt` file and the dataframe is saved to the
global object.

For each sample, in step one, ShortStack identifies sRNA cluster
(referred to as dicer-derived clusters) which aligned with locations across the
merged reference genome. The samples will share many of these cluster, although,
there will be some unique clusters only found in the treatment and not in the
controls, and potentially vice versa. These will likely correspond to potential
mobile molecules, but we cannot be sure till we have fully analysed the data.


``` r
sample_names <- c("<treatment_1>", "<treatment_2>", "<control_1>","<control_2>")

folder <- <./output/directory/from/step/1/>
save_folder <- <./output/directory/ClustersInfo.txt>


loci_info <- mobileRNA::RNAloci(files = folder,
             out = save_folder,
             samples = sample_names)
```


### Step 3 - Uniquely map samples to all loci of identified sRNA clusters

The final step in mapping is similar to the first step, however, this time we
include the newly created annotation file containing all the sRNA cluster
information as well as an additional parameter to reduce noise in the data.
This parameter (`--mincov 5`) sets a minimum read coverage of 5 (i.e. there must
be a minimum of 5 reads per cluster in a sample).

``` bash
ShortStack \
--readfile <control_1.fastq> \
--genomefile <merged_reference.fa> \
--locifile <./output/directory/ClustersInfo.gff3> \
--bowtie_cores 6 \
--mmap n \
--mismatches 0 \
--nohp \
--mincov 5 \
--outdir <./output/directory/step2/>

```

# Analysis
Here, the analysis of the pre-processed data occurs, with the aim to
identify if any mobile sRNA molecules are traveling from one genotype to
the other in your grafting system.

## Import Data
In the pre-processing steps, the data was cleaned and aligned to
the merged reference genome. During the mapping step, a folder for each sample
is created which stores all the results in. The analysis steps requires the
information which is stored in the `Results.txt`.

The `RNAimport()` function imports the data from the folder containing
all sample folders. The function extracts the required information, stores it
in a matrix.

The function requires some information to coordinate the importation and
calculations. It requires a directory path to your processed samples;
this is the path to the folder containing all the
individual sample folders which is stores in argument `directory`. The second
directory path is to the loci file containing information on the loci of the
sRNA clusters across the analysis, stored in the argument `loci`.
The last pieces of information the function requires a vector containing the
sample names, both treatment and control replicates which is stored in the argument
`samples`. The sample names must match the names of the folders produced in
the mapping, and stored in the directory supplied to the argument `directory`.


```{r, eval = FALSE, message=FALSE}
## Import & organise data.
results_dir <-  "./analysis/alignment_unique_two/"
sample_names <- c("TomEgg_1","TomEgg_2", "TomEgg_3",
                "TomTom_1", "TomTom_2", "TomTom_3")
clusters <- utils::read.table(file = "./data/reference/ClustersInfo.txt",
                              header = TRUE, sep = "\t",
                              stringsAsFactors = TRUE,comment.char="")

sRNA_data <- RNAimport(loci = loci_info,
                       directory = results_dir,
                           samples = sample_names)

```

## Sample-level quality control
A handy step in the analysis is to assess the overall similarity between
sample replicates to understand which samples are the most similar and which are
different, and where the most variation is introduced in the data set.
As well as understanding whether the data set meets your expectations.
It is expected that between the conditions, the sample replicates show
enough variation to suggest that the replicates are from different groups.

To investigate the sample similarity/difference, we will undertake sample-level
quality control using three different methods:

- Distribution of RNA classes within each sample
- Principle component analysis (PCA)
- hierarchical clustering Heatmap

This will show us how well samples within each condition cluster together, which
may highlight outliers. Plus, to show whether our experimental conditions
represent the main source of variation in the data set.

Here we will be employing an unsupervised clustering methods for the PCA and
hierarchical clustering Heatmap. This involves an unbiased log2-transformation of
the counts which will emphasis the the sample clustering to improve
visualization. The DESeq2 package contains a particularly useful function to
undertake regularized log transformation (rlog) which controls the variance
across the mean, and in this package we have utilized this for the quality
control steps.

### Plot the distribution of RNA classes within each sample
Explore the number of each RNA class identified within each sample using the
`RNAdistribution` function which produces a table and plots the results. The
results can be plotted as either a bar chart (`style = "bar"`)  or a line graph
(`style = "line"`) for each sample and can be shown in a bar chart facet
(`facet = TRUE`) or in as a single line graph, where each line represents a
sample (`together=TRUE`). Alternatively, the results can be plotted individually
for each sample in either format.

##### Table 1. The total number of each RNA class found within each sample,
##### where N represents unclassified RNAs.

```{r, echo=FALSE}
cap9 <- "An example line graph (A) and bar graph (B), to show the distribution of
RNA classes within each sample"

```

```{r, message=FALSE, fig.cap=cap9, fig.show="hold", fig.subcap= c("A","B")}

sample.distribution_line <- RNAdistribution(sRNA_data,
                                            style = "line",
                                            together = FALSE,
                                            facet = TRUE)

sample.distribution_bar <- RNAdistribution(sRNA_data,
                                           style = "bar",
                                           facet = TRUE)

```

### Principle component analysis to assess sample distance
Principal Component Analysis (PCA) is a useful technique to illustrate sample
distance as it emphasizes the variation through the reduction of dimensions in
the data set. Here, we introduce the function `plotSamplePCA()`

```{r, echo=FALSE}
cap3 <-"An example of a PCA, illustracting the sRNA data set sample similarity"

```
```{r ,message=FALSE, fig.cap=cap3, fig.show="hold", out.width="50%"}

group <- c("Tomato/Eggplant", "Tomato/Eggplant", "Tomato/Eggplant",
           "Tomato/Tomato",   "Tomato/Tomato", "Tomato/Tomato")

plotSamplePCA(sRNA_data, group)

```

### Hierarchical clustered heatmap to assess sample distance

Similarly, to a PCA plot, the `plotSampleDistance()` function undertakes
hierarchical clustering with an unbiased log transformation to calculate sample
distance and is plotted in the form of a heatmap.

```{r, echo=FALSE}
cap4 <-"An example of a heatmap,illustracting the sRNA data set sample similarity"

```
```{r ,message=FALSE, fig.cap=cap4, fig.show="hold", out.width="50%"}
plotSampleDistance(sRNA_data)

```

## Define the consensus RNA class
Have a look at the `sRNA_data` object, you will see that for each sample the
class of sRNA has been determined (see columns with names starting with
DicerCall). The result will either state a number from 20-24 which represents
the nucleotide length of the sRNA, or, state NA/N which represents that the
sRNA was unclassified in the sample. There is usually some discrepancies between
the classification of a sRNA, therefore, a consensus class can be
assigned based off the information from the treatment samples.
The `RNAconsensus()` function is used to compute this by counting the number
of occurrences of classifications for a sRNA. The consensus is determined
based on the class with the most counts.

The function requires the user to state which samples to base the consensus
calculation on (see argument `conditions`). It is recommend to state the treatment
samples only, because these are the ones of interest.
However, if you have less than two treatment replicate, calculating
the consensus class this way could dramatically shrink your data - although this
could be a benefit as it'll make the analysis more stringent. Therefore, in this
particular circumstance, it is suggested that the user tries defining the
consensus class with all samples in parallel to defining the consensus with the
treatment samples.


The function also works to remove noise from your data. If the user wishes to
keep the noise, set `tidy=FALSE` to  prevent the removal of unclassified
clusters.

For the downstream analysis, it can be useful to define the distinct 24-nt and
21/22-nt populations of sRNA molecules at the beginning of the analysis.
Hence, the data should be subset. To subset the data, use the `RNAsubset()`
function to choose which sRNA populations to include:
  20-nt, 21-nt, 22-nt, 23-nt, or 24-nt. It is recommended to subset the data into
two dataframes: one which includes only 24-nt sRNA and the other to include both
21-nt and 22-nt sRNAs. To instruct the `RNAsubset()` function to select
specific groups, use the `type` argument.

```{r, message=FALSE}
##  define consensus sRNA classes based of treatment/hetero-graft samples
samples <- c("TomEgg_1", "TomEgg_2", "TomEgg_3")

# define consensus, store as a data summary file.
sRNA_data_summary <- RNAconsensus(data = sRNA_data,
                                  conditions = samples,
                                  tidy=TRUE)
# Subset data for analysis: 24-nt sRNAs
sRNA_24 <- RNAsubset(sRNA_data_summary, type = 24)
# Subset data for analysis: 24 21/22-nt sRNAs
sRNA_2122 <- RNAsubset(sRNA_data_summary, type = c(21, 22))

```

### Plot consensus classification of sRNA

Use `RNAdistriution()` to visualize the distribution of the consensus sRNA
classes within the data. This step can be used as a quality control step to
illustrate the distribution of each sRNA class within the data. The function
can plot the absolute or relative frequency, use argument `relative`. It can be
beneficial to view the absolute frequency as well as the relative to observe
whether the data meets your expectations.

```{r}
consensus_plot <- RNAdistribution(data = sRNA_data_summary,
                                  consensus = TRUE, style = "consensus")
```



## Differential Expression analysis with DESeq2 or edgeR
Differential expression (DE) analysis is undertaken to identify sRNA which are
statistically significant to discover quantitative changes in the expression
levels between the treatment (hetero-grafting) and the control (self-grafting)
groups. This technique can be undertaken with a variety of tools, in `mobileRNA`
users have the option to use the `DESeq2` or `edgeR` analytical method.

Although both methods will produce different results, it is typically they case
that that at the step where we isolate the mobile molecules, the same results
are produce. What is meant by this, is that the same mobile molecules are
typically identified at the end of the analysis regardless of the analytical
method. However, in particular case, one method will be preferred over the other.
For instance, the `DESeq2` method is not appropriate when the experiment does not
contain replicate (ie. one sample replicate per condition). On the other hand,
edgeR can be used. Here, we have included the recommend practice for edgeR when
the data set does not contain replicates. This option can be employed by setting
a custom dispersion value, see argument `dispersionValue`.

```{r DEprep, message = FALSE, warning = FALSE}
# sample conditions.
groups <- c("Tomato/Eggplant", "Tomato/Eggplant", "Tomato/Eggplant",
            "Tomato/Tomato", "Tomato/Tomato", "Tomato/Tomato")


## Differential analysis: DEseq2 method
# 24-nt sRNA data-set
sRNA_24_DESeq2 <- RNAanalysis(data = sRNA_24,
                              group = groups,
                              method = "DESeq2" )
# 2122-nt sRNA data-set
sRNA_2122_DESeq2 <- RNAanalysis(data = sRNA_2122,
                                group = groups,
                                method = "DESeq2" )


## Differential analysis: edgeR method
sRNA_24_edgeR <- RNAanalysis(data = sRNA_24,
                             group = groups,
                             method = "edgeR" )

sRNA_2122_edgeR <- RNAanalysis(data = sRNA_2122 ,
                               group = groups,
                               method = "edgeR" )
```


## Identify the mobile molecules
It is essential to remove the noise from the data to isolate potential mobile
molecules which include falsely mapped clusters and, in this case, the clusters
which map the the scion genotype when the aim is to identify molecules traveling
from the rootstock. In the data, the aim is to explore whether there are tomato
sRNA traveling up through the graft junction and exiting into the
eggplant leaf tissues. We can identify potential mobile molecule by using
the `RNAmobile()` function. It selects clusters mapping to the rootstock
genome in the hetero-graft. Therefore, we will remove clusters mapped to
chromosome in the genotype used in the self-graft controls.


In the example, the tomato genome has chromosome labeled with
"SL40" whilst the the eggplant genome has no labels. To remove clusters
associated with the eggplant genome and keep only clusters associated with the
tomato genome, set the `"task"` argument as `"keep"`.

The `RNAmobile()` function can take into account the statistically significance.
If set to consider this, the function will filter sRNA based on an adjusted
p-value threshold = 0.05. The threshold can be changed to make it less stringent
by using the argument `padj`.

Similarly, if you would prefer to extract the mobile RNA based on the p-value,
rather than the adjusted p-values, the argument a numeric threshold can be
set for the argument `p.value`. If you would like to explore all mobile
molecules, both significant and not, the statistical analysis can be removed by
setting `stat=FALSE`. This might be useful if you wish to compare the data sets.

```{r , message=FALSE}
# vector of control names
control_names <- c("TomTom_1", "TomTom_2", "TomTom_3")




## Mobile Molecules: DEseq2 method
# remove clusters associated to tomato in 24nt RNA database
sRNA_24_mobile_DEseq <- RNAmobile(data = sRNA_24_DESeq2,
                                  controls = control_names,
                                  id = "SL40",
                                  task = "remove")

# remove clusters associated to tomato in 2122nt RNA database
sRNA_2122_mobile_DEseq  <- RNAmobile(data = sRNA_2122_DESeq2,
                                     controls = control_names,
                                     id = "SL40",
                                     task = "remove")





## Mobile Molecules: edgeR method
# remove clusters associated to tomato in 24nt RNA database
sRNA_24_mobile_edgeR <- RNAmobile(data = sRNA_24_edgeR,
                                  controls = control_names,
                                  id = "SL40", task = "remove")

# remove clusters associated to tomato in 2122nt RNA database
sRNA_2122_mobile_edgeR <- RNAmobile(data = sRNA_2122_edgeR,
                                    controls = control_names,
                                    id = "SL40",
                                    task = "remove")
```


### Heatmap plots to represent mobile molecules
Here we can use a hierarchical clustering heatmap to analysis the potential
statistically significant mobile sRNAs we have identified.
It is expected that the log2 fold changes should be opposite between your
treatment and controls where the treatment samples shows up-regulation and the
control samples show down-regulation.


### Mobile molecules from the 24-nt sRNA dataset
```{r, echo=FALSE}
cap7 <- "An example heatmap of the potential mobile 24-nt RNA molecules
A) Identifed by the `DESeq2` method. B) Identifed by the `edgeR` method."

cap8<- "An example heatmap of the potential mobile 21/22-nt RNA molecules,
A) Identifed by the `DESeq2` method. B) Identifed the `edgeR` method."

```

```{r,fig.cap=cap7, fig.show="hold", out.width="50%", fig.subcap= c("A DESeq2","B:edgeR") }

p10 <- plotHeatmap(sRNA_24_mobile_DESeq2)
p11 <- plotHeatmap(sRNA_24_mobile_edgeR)

```


### Mobile molecules from the 21/22-nt sRNA dataset
```{r, message=FALSE, fig.cap=cap8, fig.show="hold", out.width="50%", fig.subcap=  c("A:DESeq2","B:edgeR")}

p12 <- plotHeatmap(sRNA_2122_mobile_DESeq2)
p13 <- plotHeatmap(sRNA_2122_mobile_edgeR)
```


# Additional features

The *mobileRNA* package offers several of other helpful functions to aid
analysis. This includes a function to calculate the mean FPKM and counts across
specific samples and also calculate the overlap between genomic features and
sRNA clusters.

## Calculate RPM and Count means for specific samples
This function calculates the RPM mean and Count mean across specific samples.

```{r}
# calculate mean RPM and mean Counts for heterograft samples
selected_samples <- c("TomEgg_1", "TomEgg_2", "TomEgg_3")
means <- RNAmean(data = sRNA_data, conditions = selected_samples)

```


## Calculate distribution of sRNA clusters across genomic features.

The distribution of sRNA clusters across genomic features should show a similar
distribution as size distribution. It is expected that 24-nt sRNA are associated
with repeat regions and promoter regions, while 21/22-nt sRNA are likely to be
associated with coding sequence.

This functions calculates the number of sRNA clusters associated to different
genomic regions including transposable elements, promoters, introns, exons and
untranslated regions. The results can either be displayed in the matrix as an
absolute value or as a percentage of the total.

The output can be utilsed and passed into a plot of your choice. Similarly,
this features can be utilsed across different data sets. For example, across
the total data set or specific mobile sRNA subsets.

It is important to recognise, that if you alter the chromosome names in either
of the mapping reference genomes, the chromosome names will need to match in the
annotation files(s).

```{r, eval = FALSE}
# matrix containing percentage of sRNA clusters associated to different genomic
# features
dis_features <- RNAfeatures(data = sRNA_24_mobile_DESeq2,
                            annotation = "./annotation/eggplant_genome.gff3",
                            repeats = "./annotation/eggplant_genome_repeats.gff3")
```


## Assign sRNA clusters with an annotation attribute


Each sRNA cluster contains coordinates, these can be matches with coordinates in
an annotation file (individual, or, merged). If there is a match, the cluster
will receive information on:

  - `source` : name of the program that generated this feature, or the data source
(database or project name)
- `feature` : feature type name, e.g. Gene, Variation, Similarity
- `score` : A floating point value
- `strand` : defined as + (forward) or - (reverse)
- `frame` : One of '0', '1' or '2'. '0' indicates that the first base of the
feature is the first base of a codon, '1' that the second base is the first
base of a codon, and so on
- `attribute` : A semicolon-separated list of tag-value pairs, providing
additional information about each feature.

It is important to recognise, that if you alter the chromosome names in either
of the mapping reference genomes, the chromosome names will need to match in the
annotation files(s).


```{r, eval = FALSE}

# matrix containing percentage of sRNA clusters associated to different genomic

# add matching attributes to all sRNA clusters
data_plus_attributes <- RNAattributes(data = sRNA_data_summary,
                                      annotation = "./annotation/merged annotation.gff3")

# add matching attributes to potential mobile 24nt sRNA clusters
data_plus_attributes_eggplant <- RNAattributes(data = sRNA_24_mobile_DESeq2,
                                               annotation = "./annotation/eggplant_genome.gff3")
```



# Session information
```{r}
sessionInfo()
```
