<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="The mobileRNA package allows for the identifcation of mobile small 
  RNA molecules (20-24 nt) and messenger RNAs (mRNAs) derived from one genotype 
  which have travelled across a graft junction into another genotype in a 
  hetero-graft plant system.The package includes pre-mapping steps and analysis 
  steps for sRNA-seq mRNA-seq data with the end goal of identifying mobile RNAs 
  and their potential function or implications on phenotypic changes. ">
<title>mobileRNA: Identify mobile RNA molecules in plant graft systems  • mobileRNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="mobileRNA: Identify mobile RNA molecules in plant graft systems ">
<meta property="og:description" content="The mobileRNA package allows for the identifcation of mobile small 
  RNA molecules (20-24 nt) and messenger RNAs (mRNAs) derived from one genotype 
  which have travelled across a graft junction into another genotype in a 
  hetero-graft plant system.The package includes pre-mapping steps and analysis 
  steps for sRNA-seq mRNA-seq data with the end goal of identifying mobile RNAs 
  and their potential function or implications on phenotypic changes. ">
<meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/mobileRNA.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="mobilerna-">mobileRNA 
<a class="anchor" aria-label="anchor" href="#mobilerna-"></a>
</h1>
</div>
<p><br></p>
</div>
<div class="section level1">
<h1 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h1>
<p><br> mobileRNA is an <code>R</code> package that provides a pipeline for the rapid identification of endogenous mobile RNA molecules in plant graft systems. The tool provides a pipeline for pre-processing and analysis of sRNA sequencing data, and soon mRNA sequencing data.</p>
<p>It has been established that many different substances and molecules including RNAs can travel across the graft junction. Plant heterograft systems are comprised of two genotypes joined at the graft junction; hence, molecules produces and encoded by each genotype can move across the graft junction and be exchanged. These molecules could have implications to the regulation of gene expression and trait acquisition.</p>
<p>Current methods utilise a step-wise mapping of samples to each genome within the graft system. While, here we introduce a new mapping method where we align each sample replicates to a merge genome reference comprised of both genome assemblies relating to the genotypes in the heterograft system.</p>
<div class="section level2">
<h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a>
</h2>
<p>Katie Jeynes-Cupper, University of Birmingham, <a href="mailto:kej031@student.bham.ac.uk" class="email">kej031@student.bham.ac.uk</a></p>
</div>
<div class="section level2">
<h2 id="table-of-contents">Table of Contents<a class="anchor" aria-label="anchor" href="#table-of-contents"></a>
</h2>
<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="#Loading-test-data">Loading test data</a></li>
<li><a href="#Getting-help">Getting help</a></li>
<li><a href="#Summary">Summary</a></li>
<li><a href="#Pre-mapping">Pre-mapping</a></li>
<li><a href="#Mapping">Mapping</a></li>
<li><a href="#Post-mapping-analysis">Post-mapping analysis</a></li>
<li><a href="#Output">Output</a></li>
<li><a href="#Optional-extras-and-additional-applications">Optional extras and additional applications</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The latest version of the package can be install directly from this GitHub repo:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"KJeynesCupper/mobileRNA"</span>, branch <span class="op">=</span> <span class="st">"main"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-test-data">Loading test data<a class="anchor" aria-label="anchor" href="#loading-test-data"></a>
</h2>
<p>To simulate the usage of the package, there is a grafting data set, from a tomato-eggplant grafting experiment. The data has already undertaken the pre-mapping and mapping steps (pre-processing) and can be found here <code>sRNA_data</code>. The data set has already been organised by the <code><a href="reference/RNAimport.html">RNAimport()</a></code> function available in the package.</p>
<p>To load the data set, use the following command:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sRNA_data"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h2>
<p>For additional information on each function, please read through the documentation in the <code>mobileRNA</code> package by typing the <code>?</code> help operator before any of the function names in the package or by using the <code><a href="https://rdrr.io/r/utils/help.html" class="external-link">help()</a></code> function.</p>
<p>For an in-depth step-by-step analysis, consider reading the vignette provided with this package:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"mobileRNA"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p><img src="./reference/figures/program_flow.png" width="300" height="250" align="right"></p>
<p>The workflow is shown in the figure to the right. It begin in R-Studio to merge the two genome assemblies into one, then the pre-processing moves into Linux to align each replicate to the merged reference and then back into R-Studio to undertake the analysis to identify potentially mobile RNA species.</p>
<p><br><br><br><br></p>
</div>
<div class="section level2">
<h2 id="pre-mapping">Pre-mapping<a class="anchor" aria-label="anchor" href="#pre-mapping"></a>
</h2>
<p>Raw fastq files should be trimmed to remove adapter sequences and low quality reads as per best practice. We recommend installing the <code>ShortStack</code> (<a href="https://github.com/MikeAxtell/ShortStack" class="external-link uri">https://github.com/MikeAxtell/ShortStack</a>) program to align and cluster sRNA-seq samples. The program is formulated for sRNA-seq analysis, utilising Bowtie and an custom clustering algorithm.</p>
<p>Here, we introduce a mapping method utilising a merged genome reference, comprised of both genome assemblies relating to the genotypes in the heterograft system.</p>
<p><code>mobileRNA</code> offers a function to merge two FASTA reference genomes into one. To ensure the two genomes are distinguishable within the merged file, the function added a prefix to each chromosome name. As standard, the string “A_” is added to the reference genome supplied to “genomeA” and “B_” is added to the reference genome supplied to “genomeB”. These can be customised to the users preference, see manual for more information.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">merged_reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RNAmergeGenomes.html">RNAmergeGenomes</a></span><span class="op">(</span>genomeA <span class="op">=</span> <span class="st">"./workplace/reference/ref1.fa"</span>,</span>
<span>                                    genomeB <span class="op">=</span> <span class="st">"./workplace/reference/ref2.fa"</span>,</span>
<span>                        out_dir <span class="op">=</span> <span class="st">"./workplace/reference/merge/merged_ref.fa"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level2">
<h2 id="mapping-srna-seq-data-with-shortstack">Mapping: sRNA-seq data with ShortStack<a class="anchor" aria-label="anchor" href="#mapping-srna-seq-data-with-shortstack"></a>
</h2>
<p><br></p>
<p>Here, we recommend a double-mapping process using <code>ShortStack</code>, the steps are as follow:</p>
<div class="section level4">
<h4 id="step-1---identify-loci-of-dicer-derived-srna-cluster-in-each-sample">Step 1 - Identify loci of dicer-derived sRNA cluster in each sample<a class="anchor" aria-label="anchor" href="#step-1---identify-loci-of-dicer-derived-srna-cluster-in-each-sample"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb5-2"><a href="#cb5-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb5-3"><a href="#cb5-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb5-4"><a href="#cb5-4"></a>--bowtie_cores 6 \</span>
<span id="cb5-5"><a href="#cb5-5"></a>--mmap n \</span>
<span id="cb5-6"><a href="#cb5-6"></a>--mismatches 0 \</span>
<span id="cb5-7"><a href="#cb5-7"></a>--nohp \</span>
<span id="cb5-8"><a href="#cb5-8"></a>--outdir <span class="op">&lt;</span>./output/directory<span class="op">&gt;</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="step-2---uniquely-map-samples-to-all-loci-of-identified-srna-clusters">Step 2 - Uniquely map samples to all loci of identified sRNA clusters<a class="anchor" aria-label="anchor" href="#step-2---uniquely-map-samples-to-all-loci-of-identified-srna-clusters"></a>
</h4>
<p>Using the output from Step 1, the <code><a href="reference/RNAloci.html">RNAloci()</a></code> function can collate all the loci information into a single <code>.txt</code> file which can be utilised by <code>ShortStack</code> in the finally mapping step to the <code>locifile</code> argument.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a></span>
<span id="cb6-2"><a href="#cb6-2"></a>folder &lt;-<span class="st"> </span><span class="er">&lt;</span>.<span class="op">/</span>output<span class="op">/</span>directory<span class="op">/</span>from<span class="op">/</span>step<span class="op">/</span><span class="dv">1</span><span class="op">/</span><span class="er">&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>save_folder &lt;-<span class="st"> </span><span class="er">&lt;</span>.<span class="op">/</span>output<span class="op">/</span>directory<span class="op">/</span>ClustersInfo.txt<span class="op">&gt;</span></span>
<span id="cb6-4"><a href="#cb6-4"></a>sample_names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"&lt;treatment_1&gt;"</span>, <span class="st">"&lt;treatment_2&gt;"</span>, <span class="st">"&lt;control_1&gt;"</span>,<span class="st">"&lt;control_2&gt;"</span>)</span>
<span id="cb6-5"><a href="#cb6-5"></a></span>
<span id="cb6-6"><a href="#cb6-6"></a></span>
<span id="cb6-7"><a href="#cb6-7"></a>loci_info &lt;-<span class="st"> </span><span class="kw">RNAloci</span>(<span class="dt">files =</span> folder, </span>
<span id="cb6-8"><a href="#cb6-8"></a>             <span class="dt">out =</span> save_folder,</span>
<span id="cb6-9"><a href="#cb6-9"></a>             <span class="dt">samples =</span> sample_names)</span></code></pre></div>
<p><br></p>
<p>Each sample is mapped to the merged reference genome with the annotation file containing the cluster loci to analyse.</p>
<p><br></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb7-2"><a href="#cb7-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb7-3"><a href="#cb7-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb7-4"><a href="#cb7-4"></a>--locifile <span class="op">&lt;</span>./output/directory/ClustersInfo.gff3<span class="op">&gt;</span> \</span>
<span id="cb7-5"><a href="#cb7-5"></a>--bowtie_cores 6 \</span>
<span id="cb7-6"><a href="#cb7-6"></a>--mmap n \</span>
<span id="cb7-7"><a href="#cb7-7"></a>--mismatches 0 \</span>
<span id="cb7-8"><a href="#cb7-8"></a>--nohp \</span>
<span id="cb7-9"><a href="#cb7-9"></a>--mincov 5 \</span>
<span id="cb7-10"><a href="#cb7-10"></a>--outdir <span class="op">&lt;</span>./output/directory/step2/<span class="op">&gt;</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="post-mapping-analysis">Post-mapping analysis<a class="anchor" aria-label="anchor" href="#post-mapping-analysis"></a>
</h2>
<p>The aligned data can now be analysed in R, and potential mobile sRNA can be identified.</p>
<div class="section level4">
<h4 id="step-1-import-data">Step 1: Import data<a class="anchor" aria-label="anchor" href="#step-1-import-data"></a>
</h4>
<p>State the location of the mapping results, the sample names and sRNA cluster loci annotation file.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directory containing results</span></span>
<span><span class="va">results_dir</span> <span class="op">&lt;-</span>  <span class="st">"&lt;./output/directory/step2/&gt;"</span></span>
<span></span>
<span><span class="co"># Sample names and total number of reads, in the same order. </span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"&lt;treatment_1&gt;"</span>, <span class="st">"&lt;treatment_2&gt;"</span>, <span class="st">"&lt;control_1&gt;"</span>,<span class="st">"&lt;control_2&gt;"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">sRNA_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RNAimport.html">RNAimport</a></span><span class="op">(</span>directory <span class="op">=</span> <span class="va">results_dir</span>,</span>
<span>                      samples <span class="op">=</span> <span class="va">sample_names</span><span class="op">)</span></span>
<span>                           </span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="step-2-calculate-the-consensus-of-each-srna-cluster">Step 2: Calculate the consensus of each sRNA cluster<a class="anchor" aria-label="anchor" href="#step-2-calculate-the-consensus-of-each-srna-cluster"></a>
</h4>
<p>For a given sRNA cluster, each replicate has determined the class (20-24nt) based on the most abundant small RNA size. Replicates within the same condition are expected to class a given sRNA similarly.</p>
<p>The <code><a href="reference/RNAconsensus.html">RNAconsensus()</a></code> function is used to define the class of a sRNA cluster based on the consensus across specific replicates. To identify forigen mobile sRNAs, it is recommended to base the consensus call on the heterograft samples. This will ensure that the sRNA class is more accurately defined by the genotype it originates from.</p>
<p>Below, replicates “treatment_1” and “treatment_2” represent two heterograft samples, while “control_1” and “control_2” represent self-graft samples.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"&lt;treatment_1&gt;"</span>, <span class="st">"&lt;treatment_2&gt;"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sRNA_data_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RNAconsensus.html">RNAconsensus</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data</span>, </span>
<span>                                 conditions <span class="op">=</span> <span class="va">samples</span>, </span>
<span>                                 tidy<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="step-3-identify-potential-mobile-srna">Step 3: Identify potential mobile sRNA<a class="anchor" aria-label="anchor" href="#step-3-identify-potential-mobile-srna"></a>
</h4>
<p>Finally, the <code><a href="reference/RNAmobile.html">RNAmobile()</a></code> function filters the dataset to retain the sRNA clusters mapped to the A-genome. It is important, that each genome is distinguishable by the chromosome names, hence, in this example the A-genome contains “A” before the chromosome number.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># define control samples</span></span>
<span><span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"&lt;control_1&gt;"</span>, <span class="st">"&lt;control_2&gt;"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mobile</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_summary</span>, </span>
<span>                    controls <span class="op">=</span> <span class="va">controls</span>,</span>
<span>                    id <span class="op">=</span> <span class="st">"A"</span>, </span>
<span>                    task <span class="op">=</span> <span class="st">"keep"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># output dataframe containing potentially mobile sRNAs</span></span>
<span><span class="va">output</span> <span class="op">&lt;-</span> <span class="va">mobile</span></span>
<span></span>
<span><span class="co"># save output as txt file </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">output</span>, <span class="st">"./output.txt"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h2>
<p>A dataframe where rows represent potential mobile sRNA clusters. The columns include information on the cluster, individual sample replicates, and more.</p>
<div class="section level3">
<h3 id="information-on-the-cluster">Information on the cluster:<a class="anchor" aria-label="anchor" href="#information-on-the-cluster"></a>
</h3>
<ul>
<li>
<code>chr</code>: Name of the chromosome or scaffold</li>
<li>
<code>start</code> : Start position of the cluster</li>
<li>
<code>end</code> : End position of the cluster</li>
<li>
<code>width</code> : Length of the locus (base-pairs)</li>
</ul>
</div>
<div class="section level3">
<h3 id="information-on-each-sample-replicate">Information on each sample replicate:<a class="anchor" aria-label="anchor" href="#information-on-each-sample-replicate"></a>
</h3>
<ul>
<li>
<code>cluster</code>: Cluster Name</li>
<li>
<code>Dicer-call</code> : The size of most abundant small RNA size</li>
<li>
<code>Count</code> : Number of uniquely aligned (<em>e.g.</em> not multi-mapping) reads that overlap this locus.</li>
<li>
<code>RPM</code> : Reads per Million</li>
</ul>
</div>
<div class="section level3">
<h3 id="other-information">Other information<a class="anchor" aria-label="anchor" href="#other-information"></a>
</h3>
<ul><li>
<code>sRNA_Consensus</code> : Consensus sRNA class calculated by <code><a href="reference/RNAconsensus.html">RNAconsensus()</a></code>
</li></ul>
<p>If an annotation file was imported and overlapped using <code><a href="reference/RNAattributes.html">RNAattributes()</a></code>:</p>
<ul>
<li>
<code>source</code> : name of the program that generated this feature, or the data source (database or project name)</li>
<li>
<code>feature</code> : feature type name, e.g. Gene, Variation, Similarity</li>
<li>
<code>score</code> : A floating point value</li>
<li>
<code>strand</code> : defined as + (forward) or - (reverse)</li>
<li>
<code>frame</code> : One of ‘0’, ‘1’ or ‘2’. ‘0’ indicates that the first base of the feature is the first base of a codon, ‘1’ that the second base is the first base of a codon, and so on</li>
<li>
<code>attribute</code> : A semicolon-separated list of tag-value pairs, providing additional information about each feature.</li>
</ul>
<p>If differential expression analysis was undertaken with<code><a href="reference/RNAanalysis.html">RNAanalysis()</a></code>:</p>
<ul>
<li>
<code>log2FoldChange</code> : Log2FoldChange–The effect size estimate</li>
<li>
<code>pvalue</code> : P value, the probability under the assumption of no effect or no difference, of obtaining a result equal to or more extreme than what was actually observed</li>
<li>
<code>padjusted</code> : A p-value adjustment</li>
<li>
<code>logCPM</code> : log counts per million, measure of expression level</li>
</ul>
<p>If the mean RPM and Count was calculated <code><a href="reference/RNAmean.html">RNAmean()</a></code>: - <code>mean_RPM</code> : mean RPM, based on parameters - <code>mean_Count</code> : mean counts, based on parameters</p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="optional-extras-and-additional-applications">Optional extras and additional applications<a class="anchor" aria-label="anchor" href="#optional-extras-and-additional-applications"></a>
</h2>
<p>The package also includes functions for:</p>
<ul>
<li>Exploratory and quality control analysis, such as PCA &amp; heatmap plots.</li>
<li>Summary values including RPM mean and Count mean across specific samples.</li>
<li>Plot the distribution of sRNA classes (20-24nt) across indivudal replicates or across the dataset.</li>
<li>Statistical analysis using differential methods from either DESeq2 or edgeR.</li>
<li>Functional analysis to identify potential target sequences in the destination tissue</li>
</ul>
<p>The package workflow can easily be manipulated to enable the identification of local populations of RNA species.</p>
<hr>
<p><em>Last updated:</em> 06-06-2023</p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/KJeynesCupper/mobileRNA.git/" class="external-link">Browse source code</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing mobileRNA</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Katie Jeynes-Cupper <br><small class="roles"> Maintainer, author </small>  </li>
<li>Marco Catoni <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
