<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="mobileRNA">
<title>Analyse sRNA populations using mobileRNA • mobileRNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analyse sRNA populations using mobileRNA">
<meta property="og:description" content="mobileRNA">
<meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/mobileRNA.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/mobileRNA_standard.html">Analyse sRNA populations using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_chimeric.html">Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_mRNA.html">Explore canditate mobile mRNAs &amp; population-scale mRNA changes in chimeric systems using mobileRNA</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="mobileRNA_standard_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Analyse sRNA populations using mobileRNA</h1>
                        <h4 data-toc-skip class="author">Katie Jeynes-Cupper, Marco Catoni</h4>
            
            <h4 data-toc-skip class="date">2023-09-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KJeynesCupper/mobileRNA.git/blob/HEAD/vignettes/mobileRNA_standard.Rmd" class="external-link"><code>vignettes/mobileRNA_standard.Rmd</code></a></small>
      <div class="d-none name"><code>mobileRNA_standard.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In plants, systemic signalling is an elaborated molecular system which allows to coordinate plant development at the entire organism level, integrating and transmitting the information perceived from environment to distant organs. RNA molecules play a key role in this extensive network, both across short and long distances. Of which, small RNAs (sRNA) have been shown to be involved in post-transcriptional gene silencing and RNA-directed DNA methylation. Alterations such as these have been linked to changes in gene expression and phenotypical changes or trait. Here, we introduce the use of the mobileRNA R package as a tool for sRNAseq analysis to identify sRNA molecules in a biological system, regardless of whether the system is chimeric or not.</p>
<p>While, other RNA molecules, such as messenger RNA, are also renown be play a role in systemic signalling this tool currently focuses on the identification of local sRNA populations. <br></p>
</div>
<div class="section level2">
<h2 id="method">Method<a class="anchor" aria-label="anchor" href="#method"></a>
</h2>
<p>Small RNAs (sRNAs) range from 20-24 nucleotides in length, but regardless of their short length the different groups have very specific roles in genetic regulation in plants. There are two key groups of interest; micro-RNAs (miRNAs) with a size between 21-22nt and small-interfering-RNAs (siRNAs) with a size of 24nt. miRNAs are renown to play a role in post-transcriptional regulation, while siRNAs are key components in the RNA-directed DNA methylation pathway. Therefore, understanding the population dynamics of sRNAs within your biological system could shed light on gene expression and DNA methylation patterns, bringing light to the molecular mechanisms underpinning biological processes.</p>
<p>The workflow can be utilise to explore sRNA populations in any species, and can be utilise further if you have a treatment vs. control experimental design. The workflow will aid the identification of unique sRNA populations and differences in sRNA abundances between treatment and control conditions.</p>
<p>This guide suggests mapping steps, analysis steps, and functional analysis steps.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The latest version of <code>mobileRNA</code> can be installed via GitHub using the <code>devtools</code> package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"KJeynesCupper/mobileRNA"</span>, ref <span class="op">=</span> <span class="st">"main"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KJeynesCupper/mobileRNA.git" class="external-link">mobileRNA</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pre-processing">Pre-processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h2>
<p>Assuming your sRNAseq raw samples have been cleaned to remove low quality reads and remove adapter sequences, we can align the samples to the genome assembly and detect sRNA clusters using the Linux <code>ShortStack</code> program (<a href="https://github.com/MikeAxtell/ShortStack" class="external-link uri">https://github.com/MikeAxtell/ShortStack</a>).</p>
<div class="section level3">
<h3 id="auto-detection-of-srna-cluster">Auto-Detection of sRNA Cluster<a class="anchor" aria-label="anchor" href="#auto-detection-of-srna-cluster"></a>
</h3>
<div class="section level4">
<h4 id="step-1---cluster-analysis-with-shortstack">Step 1 - Cluster analysis with ShortStack<a class="anchor" aria-label="anchor" href="#step-1---cluster-analysis-with-shortstack"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb2-2"><a href="#cb2-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb2-3"><a href="#cb2-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb2-4"><a href="#cb2-4"></a>--bowtie_cores 6 \</span>
<span id="cb2-5"><a href="#cb2-5"></a>--mmap n \</span>
<span id="cb2-6"><a href="#cb2-6"></a>--mismatches 0 \</span>
<span id="cb2-7"><a href="#cb2-7"></a>--nohp \</span>
<span id="cb2-8"><a href="#cb2-8"></a>--outdir <span class="op">&lt;</span>./output/directory<span class="op">&gt;</span></span></code></pre></div>
<div class="section level5">
<h5 id="step-2---build-srna-cluster-list">Step 2 - Build sRNA cluster list<a class="anchor" aria-label="anchor" href="#step-2---build-srna-cluster-list"></a>
</h5>
<p>Now, we collate all the sRNA loci information from each sample into a text file.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># location of step 1 output</span></span>
<span><span class="va">folder</span> <span class="op">&lt;-</span> <span class="st">"./output/directory/from/step/1/"</span></span>
<span></span>
<span><span class="co"># name and location to save output file to (must be .txt)</span></span>
<span><span class="va">save_folder</span> <span class="op">&lt;-</span> <span class="st">"./output/directory/ClustersInfo.txt"</span></span>
<span></span>
<span><span class="co"># names of samples (ie. folder names)</span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"&lt;treatment_1&gt;"</span>, <span class="st">"&lt;treatment_2&gt;"</span>, <span class="st">"&lt;control_1&gt;"</span>,<span class="st">"&lt;control_2&gt;"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">loci_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAloci.html">RNAloci</a></span><span class="op">(</span>files <span class="op">=</span> <span class="va">folder</span>, </span>
<span>             out <span class="op">=</span> <span class="va">save_folder</span>,</span>
<span>             samples <span class="op">=</span> <span class="va">sample_names</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="mapping">Mapping<a class="anchor" aria-label="anchor" href="#mapping"></a>
</h3>
<p>Each sample is mapped to the merged reference genome with the list of sRNA clusters.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb4-2"><a href="#cb4-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb4-3"><a href="#cb4-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb4-4"><a href="#cb4-4"></a>--locifile <span class="op">&lt;</span>./output/directory/ClustersInfo.gff3<span class="op">&gt;</span> \</span>
<span id="cb4-5"><a href="#cb4-5"></a>--bowtie_cores 6 \</span>
<span id="cb4-6"><a href="#cb4-6"></a>--mmap n \</span>
<span id="cb4-7"><a href="#cb4-7"></a>--mismatches 0 \</span>
<span id="cb4-8"><a href="#cb4-8"></a>--nohp \</span>
<span id="cb4-9"><a href="#cb4-9"></a>--mincov 5 \</span>
<span id="cb4-10"><a href="#cb4-10"></a>--outdir <span class="op">&lt;</span>./output/directory/step2/<span class="op">&gt;</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>The aligned data can now be analysed in R, and the native population of sRNAs can be explored. This following analysis workflow is designed to facilitate a treatment vs control experimental design.</p>
<p>For the analysis we will use the semi-synthetic data, although there are two different genotypes in this analysis it can still be utilised to show different sRNA population dynamics.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KJeynesCupper/mobileRNA.git" class="external-link">mobileRNA</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sRNA_data"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="step-1-import-data">Step 1: Import data<a class="anchor" aria-label="anchor" href="#step-1-import-data"></a>
</h3>
<p>State the type of data (which is “sRNA” here), the location of the mapping results in your directory, and the sample names.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Directory containing results</span></span>
<span><span class="va">results_dir</span> <span class="op">&lt;-</span>  <span class="st">"&lt;./output/directory/step2/&gt;"</span></span>
<span></span>
<span><span class="co"># Sample names and total number of reads, in the same order. </span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span> , <span class="st">"heterograft_3"</span>, </span>
<span>                  <span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span>, <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">sRNA_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAimport.html">RNAimport</a></span><span class="op">(</span>input <span class="op">=</span> <span class="st">"sRNA"</span>, </span>
<span>                       directory <span class="op">=</span> <span class="va">results_dir</span>,</span>
<span>                       samples <span class="op">=</span> <span class="va">sample_names</span><span class="op">)</span></span>
<span>                           </span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="step-2-quality-control">Step 2: Quality control<a class="anchor" aria-label="anchor" href="#step-2-quality-control"></a>
</h3>
<p>A handy step in the analysis is to assess the overall similarity and difference between samples within a conditions and between the conditions. This is to identify where most of the variation is introduced in the data set, as well as understanding whether the data set meets your expectations, quality-wise. It is expected that between the conditions, the sample replicates show enough variation to suggest that the replicates are from different groups. Here we will use the following methods: * Distribution of RNA classes within each sample * Principle component analysis (PCA) * Heatmap using Hierarchical clustering</p>
<p>This will show us how well samples within each condition cluster together, which may highlight outliers. Plus, to show whether our experimental conditions represent the main source of variation in the data set.</p>
<div class="section level4">
<h4 id="principle-component-analysis-pca">Principle component analysis (PCA)<a class="anchor" aria-label="anchor" href="#principle-component-analysis-pca"></a>
</h4>
<p>Principal Component Analysis (PCA) is a useful technique to illustrate sample distance as it emphasizes the variation through the reduction of dimensions in the data set. Here, we introduce the function plotSamplePCA():</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSamplePCA.html">plotSamplePCA</a></span><span class="op">(</span><span class="va">sRNA_data</span>, <span class="va">group</span><span class="op">)</span></span></code></pre></div>
<p><img src="mobileRNA_standard_files/figure-html/unnamed-chunk-4-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="hierarchical-clustered-heatmap-to-assess-sample-distance">Hierarchical clustered heatmap to assess sample distance<a class="anchor" aria-label="anchor" href="#hierarchical-clustered-heatmap-to-assess-sample-distance"></a>
</h4>
<p>Similarly, to a PCA plot, the plotSampleDistance() function undertakes hierarchical clustering with an unbiased log transformation to calculate sample distance and is plotted in the form of a heatmap.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSampleDistance.html">plotSampleDistance</a></span><span class="op">(</span><span class="va">sRNA_data</span><span class="op">)</span></span></code></pre></div>
<p><img src="mobileRNA_standard_files/figure-html/unnamed-chunk-5-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="plot-the-distribution-of-rna-classes-within-each-sample">Plot the distribution of RNA classes within each sample<a class="anchor" aria-label="anchor" href="#plot-the-distribution-of-rna-classes-within-each-sample"></a>
</h4>
<p>Explore the number of each RNA class identified within each sample using the <code>RNAdistribution</code> function which produces a table and plots the results. The results can be plotted as either a bar chart (style = “bar”) or a line graph (style = “line”) for each sample and can be shown in a bar chart facet (facet = TRUE) or in as a single line graph, where each line represents a sample (together=TRUE). Alternatively, the results can be plotted individually for each sample in either format.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_distribution</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span><span class="va">sRNA_data</span>,</span>
<span>                                            style <span class="op">=</span> <span class="st">"line"</span>,</span>
<span>                                            together <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Lets look at the plot and the data:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># view plot </span></span>
<span><span class="va">sample_distribution</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="mobileRNA_standard_files/figure-html/unnamed-chunk-7-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># view data </span></span>
<span><span class="va">sample_distribution</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="co">#&gt;    Class heterograft_1 heterograft_2 heterograft_3 selfgraft_1 selfgraft_2</span></span>
<span><span class="co">#&gt; 1:    20             9            14             7           8          13</span></span>
<span><span class="co">#&gt; 2:    21            27            28            24          25          26</span></span>
<span><span class="co">#&gt; 3:    22            19            27            26          19          27</span></span>
<span><span class="co">#&gt; 4:    23            25            38            42          18          31</span></span>
<span><span class="co">#&gt; 5:    24           597           491           575         459         353</span></span>
<span><span class="co">#&gt;    selfgraft_3</span></span>
<span><span class="co">#&gt; 1:           6</span></span>
<span><span class="co">#&gt; 2:          22</span></span>
<span><span class="co">#&gt; 3:          26</span></span>
<span><span class="co">#&gt; 4:          35</span></span>
<span><span class="co">#&gt; 5:         437</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="step-3-calculate-the-consensus-of-each-srna-cluster">Step 3: Calculate the consensus of each sRNA cluster<a class="anchor" aria-label="anchor" href="#step-3-calculate-the-consensus-of-each-srna-cluster"></a>
</h3>
<p>For a given sRNA cluster, each replicate has determined the class (20-24nt) based on the most abundant small RNA size. Replicates within the same condition are expected to class a given sRNA similarly.</p>
<p>The <code><a href="../reference/RNAdicercall.html">RNAdicercall()</a></code> function is used to define the dicer class of a sRNA cluster based on the consensus across (specific) replicates.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate dicercall consensus </span></span>
<span><span class="va">sRNA_data_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdicercall.html">RNAdicercall</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data</span>,  </span>
<span>                                 tidy<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating consensus dicercall based on information from all replicates... </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The consensus dicercall will be excluded in the case of a tie... </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Removing small RNA clusters with no consensus dicercall...</span></span>
<span><span class="co"># plot results </span></span>
<span><span class="va">plot_distribution</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>, </span>
<span>                        consensus <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co"># View figure</span></span>
<span><span class="va">plot_distribution</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="mobileRNA_standard_files/figure-html/unnamed-chunk-9-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="step-4-statistical-analysis-using-deseq2-or-edger">Step 4: Statistical analysis using DESeq2 or edgeR<a class="anchor" aria-label="anchor" href="#step-4-statistical-analysis-using-deseq2-or-edger"></a>
</h3>
<p>Statistical analysis can be helpful to understand the differential abundances of sRNA clusters found within your system. Here we can utilize either the DESeq2 or edgeR algorithms to support your analysis, and can be used in the next steps.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sample conditions in order within dataframe</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Differential analysis using the DEseq2 method </span></span>
<span><span class="va">sRNA_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_summary</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"DESeq2"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Differential analysis using the edgeR method </span></span>
<span></span>
<span><span class="va">sRNA_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_summary</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"edgeR"</span><span class="op">)</span></span></code></pre></div>
<p>This will add differential analysis values for each sRNA cluster to the working dataframe. We can then easily manipulate this data and plot it as heatmap.</p>
<div class="section level4">
<h4 id="plot-a-volcano-plot">Plot a volcano plot<a class="anchor" aria-label="anchor" href="#plot-a-volcano-plot"></a>
</h4>
<p>A Volcano plots can be plotted to show the relationship between fold change and the adjusted p-value (i.e. FDR) across the dataset. Each point represents an individual sRNA cluster. Pointes are coloured according to their significance level thresholds (0.05, 0.01, and 0.001 or unchanged).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot with the top 10 clusters labelled</span></span>
<span><span class="va">vol_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotVolcano.html">plotVolcano</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_DESeq2</span>, labels <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">vol_plot</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="mobileRNA_standard_files/figure-html/unnamed-chunk-11-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="step-5-explore-difference-in-srna-abundances">Step 5: Explore difference in sRNA abundances<a class="anchor" aria-label="anchor" href="#step-5-explore-difference-in-srna-abundances"></a>
</h3>
<p>When comparing treatment to control conditions, it might be the case that the same sRNA clusters are found within both yet there could be difference in the total abundance of the shared clusters. For instance, for a given sRNA cluster the samples in the treatment condition might have a greater abundance than the samples in the control condition. These difference could attribute to observed differences.</p>
<p>The statistical analysis calculated the log2FC values for each sRNA cluster by comparing the normalised counts between treatment and control. Here, a positive log2FC indicates an increased abundance of transcripts for a given sRNA cluster in the treatment compared to the control, while negative log2FC indicates decreased abundance of transcripts for a given sRNA cluster. The statistical significance of the log2FC is determined by the (adjusted) pvalue also calculated in the previous step.</p>
<p>Here we will filter the data to select sRNA clusters which are statistically significant, and then plot the results as a heatmap to compare the conditions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select only signifcant sRNAs</span></span>
<span><span class="va">significant_sRNAs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsignificant.html">RNAsignificant</a></span><span class="op">(</span><span class="va">sRNA_DESeq2</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#plot</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">significant_sRNAs</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Lets look at the plot:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span><span class="op">$</span><span class="va">plot</span> </span></code></pre></div>
<p><img src="mobileRNA_standard_files/figure-html/unnamed-chunk-13-1.png" width="700" style="display: block; margin: auto;"><br></p>
</div>
<div class="section level3">
<h3 id="step-6-identify-absolute-srna-population-difference">Step 6: Identify absolute sRNA population difference<a class="anchor" aria-label="anchor" href="#step-6-identify-absolute-srna-population-difference"></a>
</h3>
<p>Finally, the <code><a href="../reference/RNApopulation.html">RNApopulation()</a></code> function can be utilised to identify unique sRNA populations found in the treatment or control condition.</p>
<p>First lets look at the sRNA clusters unique to the treatment condition:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select sRNA clusters only found in treatment &amp; not in the control samples</span></span>
<span><span class="va">treatment_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span> , <span class="st">"heterograft_3"</span><span class="op">)</span></span>
<span><span class="va">unique_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNApopulation.html">RNApopulation</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>, </span>
<span>                                   conditions <span class="op">=</span> <span class="va">treatment_reps</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># look at number of sRNA cluster only found in treatment </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">unique_treatment</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 148</span></span></code></pre></div>
<p>Now, the sRNA clusters unique to the control condition:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select sRNA clusters only found in control &amp; not in the treatment samples</span></span>
<span><span class="va">control_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span> , <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span><span class="va">unique_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNApopulation.html">RNApopulation</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>, </span>
<span>                                   conditions <span class="op">=</span> <span class="va">control_reps</span><span class="op">)</span></span>
<span><span class="co"># look at number of sRNA cluster only found in control  </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">unique_control</span><span class="op">)</span> </span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="plotting-heterograft-only-srna-populations">Plotting heterograft-only sRNA populations<a class="anchor" aria-label="anchor" href="#plotting-heterograft-only-srna-populations"></a>
</h4>
<p>Here we located 148 sRNA clusters which are uniquely present to the heterografts, and not found in the self-graft controls. These results can be plotted as a distribution plot and heatmap.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot distribution of sRNA classes only found in treatment </span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">unique_treatment</span>, style <span class="op">=</span> <span class="st">"line"</span>, together <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot heatmap of treatment-only sRNA clusters</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">unique_treatment</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Lets look at the plots:</p>
<p>Here we can see the distribution of sRNA classes only found in treatment</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="mobileRNA_standard_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Here we can compare the difference and similarities between the treatment and controls using a heatmap:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p3</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="mobileRNA_standard_files/figure-html/unnamed-chunk-18-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="step-6-functional-analysis">Step 6: Functional analysis<a class="anchor" aria-label="anchor" href="#step-6-functional-analysis"></a>
</h3>
<div class="section level4">
<h4 id="target-sequences">Target sequences<a class="anchor" aria-label="anchor" href="#target-sequences"></a>
</h4>
<p>sRNAs target complementary sequences in the genome to mediate silencing via post-transcriptional or pre-transcriptional mechanisms. Here we will identify the complementary sequences matching the most abundant sRNA within a cluster.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">target_sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsequences.html">RNAsequences</a></span><span class="op">(</span><span class="va">unique_treatment</span><span class="op">)</span></span>
<span><span class="co">#&gt; The minimum number of matches required to form a consensus sRNA sequence is...  1 </span></span>
<span><span class="co">#&gt; The consensus sRNA sequences will be choose at random in the case of a tie</span></span></code></pre></div>
<p>This dataframe can be exported the results loaded into BLAST to identify regions within the genome that these complementary sequences are present. There are tools, such as <code>TargetFinder</code> or <code>sRNATargetDigger</code>, which use a predictive algorithm to identify the most likely locations in the genome where the sRNAs might target.</p>
<p>Lets export the data frame :</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">target_sequences</span>, <span class="st">"./data/mobile_sequences.txt"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="overlapping-regions">Overlapping regions<a class="anchor" aria-label="anchor" href="#overlapping-regions"></a>
</h4>
<p>Finally, we can identify regions in the genome annotation where the sRNA cluster overlaps with based on the chromosome and the start/end coordinated. This can be explored using two functions: [RNAattributes()] and [RNAfeatures()].</p>
<p>The <code>RNAattributes</code> function compared a genome annotation with the sRNA clusters and, if an overlap is present, the information is joined. This is accomplished by adding the the information from the annotation file to the sRNA working dataframe as additional columns.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">attributes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAattributes.html">RNAattributes</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">unique_treatment</span>,</span>
<span>                            annotation <span class="op">=</span> <span class="st">"./annotation/origin_annotation.gff3"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>RNAfeatures</code> functions is similar to the <code>RNAattributes</code> function, but instead it returns either the absolute number or percentage of sRNA clusters which overlap with specific regions in the genome such as repeats, introns, exons, UTR and promoter regions. It could be helpful to subset your dataset to only explore a specific sRNA class, see [mobileRNA::RNAsubset()] for more information.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAfeatures.html">RNAfeatures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">unique_treatment</span>,</span>
<span>                            annotation <span class="op">=</span> <span class="st">"./annotation/origin_annotation.gff3"</span>, </span>
<span>                      repeats <span class="op">=</span> <span class="st">"./annotation/origin_annotation_repeats.gff3"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.0.2 (2020-06-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS  10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] mobileRNA_0.99.8 BiocStyle_2.16.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] colorspace_2.0-3            rprojroot_2.0.2            </span></span>
<span><span class="co">#&gt;   [3] DNAcopy_1.62.0              XVector_0.28.0             </span></span>
<span><span class="co">#&gt;   [5] GenomicRanges_1.40.0        fs_1.5.2                   </span></span>
<span><span class="co">#&gt;   [7] rstudioapi_0.14             farver_2.1.0               </span></span>
<span><span class="co">#&gt;   [9] listenv_0.8.0               audio_0.1-10               </span></span>
<span><span class="co">#&gt;  [11] affyio_1.58.0               ggrepel_0.9.1              </span></span>
<span><span class="co">#&gt;  [13] bit64_4.0.5                 AnnotationDbi_1.50.3       </span></span>
<span><span class="co">#&gt;  [15] fansi_1.0.2                 codetools_0.2-18           </span></span>
<span><span class="co">#&gt;  [17] splines_4.0.2               Repitools_1.34.0           </span></span>
<span><span class="co">#&gt;  [19] doParallel_1.0.17           cachem_1.0.6               </span></span>
<span><span class="co">#&gt;  [21] geneplotter_1.66.0          knitr_1.42                 </span></span>
<span><span class="co">#&gt;  [23] jsonlite_1.8.4              Rsamtools_2.4.0            </span></span>
<span><span class="co">#&gt;  [25] annotate_1.66.0             cluster_2.1.2              </span></span>
<span><span class="co">#&gt;  [27] vsn_3.56.0                  pheatmap_1.0.12            </span></span>
<span><span class="co">#&gt;  [29] BiocManager_1.30.16         compiler_4.0.2             </span></span>
<span><span class="co">#&gt;  [31] Matrix_1.4-0                fastmap_1.1.0              </span></span>
<span><span class="co">#&gt;  [33] limma_3.44.3                cli_3.6.0                  </span></span>
<span><span class="co">#&gt;  [35] beepr_1.3                   htmltools_0.5.5            </span></span>
<span><span class="co">#&gt;  [37] tools_4.0.2                 gtable_0.3.0               </span></span>
<span><span class="co">#&gt;  [39] glue_1.6.2                  GenomeInfoDbData_1.2.3     </span></span>
<span><span class="co">#&gt;  [41] affy_1.66.0                 dplyr_1.1.2                </span></span>
<span><span class="co">#&gt;  [43] Rcpp_1.0.8.3                Biobase_2.48.0             </span></span>
<span><span class="co">#&gt;  [45] jquerylib_0.1.4             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt;  [47] vctrs_0.6.2                 Biostrings_2.56.0          </span></span>
<span><span class="co">#&gt;  [49] preprocessCore_1.50.0       progressr_0.10.0           </span></span>
<span><span class="co">#&gt;  [51] rtracklayer_1.48.0          iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [53] xfun_0.37                   stringr_1.5.0              </span></span>
<span><span class="co">#&gt;  [55] globals_0.14.0              lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt;  [57] gtools_3.9.2                XML_3.99-0.9               </span></span>
<span><span class="co">#&gt;  [59] future_1.24.0               edgeR_3.30.3               </span></span>
<span><span class="co">#&gt;  [61] zlibbioc_1.34.0             MASS_7.3-55                </span></span>
<span><span class="co">#&gt;  [63] scales_1.2.1                BSgenome_1.56.0            </span></span>
<span><span class="co">#&gt;  [65] ragg_1.2.2                  parallel_4.0.2             </span></span>
<span><span class="co">#&gt;  [67] SummarizedExperiment_1.18.2 RColorBrewer_1.1-2         </span></span>
<span><span class="co">#&gt;  [69] yaml_2.3.5                  curl_4.3.2                 </span></span>
<span><span class="co">#&gt;  [71] pbapply_1.7-0               memoise_2.0.1              </span></span>
<span><span class="co">#&gt;  [73] ggplot2_3.4.2               sass_0.4.0                 </span></span>
<span><span class="co">#&gt;  [75] stringi_1.7.6               RSQLite_2.2.10             </span></span>
<span><span class="co">#&gt;  [77] highr_0.9                   genefilter_1.70.0          </span></span>
<span><span class="co">#&gt;  [79] S4Vectors_0.26.1            desc_1.4.1                 </span></span>
<span><span class="co">#&gt;  [81] foreach_1.5.2               Ringo_1.52.0               </span></span>
<span><span class="co">#&gt;  [83] caTools_1.18.2              BiocGenerics_0.34.0        </span></span>
<span><span class="co">#&gt;  [85] BiocParallel_1.22.0         truncnorm_1.0-8            </span></span>
<span><span class="co">#&gt;  [87] GenomeInfoDb_1.24.2         rlang_1.1.0                </span></span>
<span><span class="co">#&gt;  [89] pkgconfig_2.0.3             systemfonts_1.0.4          </span></span>
<span><span class="co">#&gt;  [91] matrixStats_0.61.0          bitops_1.0-7               </span></span>
<span><span class="co">#&gt;  [93] Rsolnp_1.16                 evaluate_0.15              </span></span>
<span><span class="co">#&gt;  [95] lattice_0.20-45             purrr_1.0.1                </span></span>
<span><span class="co">#&gt;  [97] labeling_0.4.2              GenomicAlignments_1.24.0   </span></span>
<span><span class="co">#&gt;  [99] bit_4.0.4                   tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt; [101] parallelly_1.30.0           magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [103] bookdown_0.32               DESeq2_1.28.1              </span></span>
<span><span class="co">#&gt; [105] R6_2.5.1                    IRanges_2.22.2             </span></span>
<span><span class="co">#&gt; [107] gplots_3.1.1                generics_0.1.2             </span></span>
<span><span class="co">#&gt; [109] DelayedArray_0.14.1         DBI_1.1.2                  </span></span>
<span><span class="co">#&gt; [111] withr_2.5.0                 pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [113] gsmoothr_0.1.7              RPushbullet_0.3.4          </span></span>
<span><span class="co">#&gt; [115] survival_3.3-1              RCurl_1.98-1.6             </span></span>
<span><span class="co">#&gt; [117] future.apply_1.8.1          tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [119] crayon_1.5.0                KernSmooth_2.23-20         </span></span>
<span><span class="co">#&gt; [121] utf8_1.2.2                  rmarkdown_2.20             </span></span>
<span><span class="co">#&gt; [123] locfit_1.5-9.4              grid_4.0.2                 </span></span>
<span><span class="co">#&gt; [125] data.table_1.14.2           blob_1.2.2                 </span></span>
<span><span class="co">#&gt; [127] SimDesign_2.10.1            digest_0.6.29              </span></span>
<span><span class="co">#&gt; [129] xtable_1.8-4                tidyr_1.3.0                </span></span>
<span><span class="co">#&gt; [131] textshaping_0.3.6           stats4_4.0.2               </span></span>
<span><span class="co">#&gt; [133] munsell_0.5.0               viridisLite_0.4.0          </span></span>
<span><span class="co">#&gt; [135] bslib_0.3.1                 sessioninfo_1.2.2</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
