<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="mobileRNA">
<title>Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA • mobileRNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA">
<meta property="og:description" content="mobileRNA">
<meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/mobileRNA.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/mobileRNA_standard.html">Analyse sRNA populations using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_chimeric.html">Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_mRNA.html">Explore canditate mobile mRNAs &amp; population-scale mRNA changes in chimeric systems using mobileRNA</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="mobileRNA_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA</h1>
                        <h4 data-toc-skip class="author">Katie Jeynes-Cupper, Marco Catoni</h4>
            
            <h4 data-toc-skip class="date">2023-09-14</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KJeynesCupper/mobileRNA.git/blob/HEAD/vignettes/mobileRNA.Rmd" class="external-link"><code>vignettes/mobileRNA.Rmd</code></a></small>
      <div class="d-none name"><code>mobileRNA.Rmd</code></div>
    </div>

    
    
<p><br></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In plants, systemic signalling is an elaborated molecular system which coordinates plant development, integrating and transmitting the information perceived from environment to distant organs. An important role in long-distance signalling is played by small RNA molecules (sRNAs). These endogenous sRNAs can be produced in a tissue and then transported systemically across the vascular system into recipient organs, where they can induce a molecular response and coordinate physiological changes. Plant grafting can be utilised to create chimeric plant systems composed of two genotypes. For example, grafting can join tomato and eggplant or different species of grapevine. Grafting of horticultural crops can lead to favorable traits, such as improved plant vigour or architecture, and resistance to stress. sRNAs have been known to move across a graft junction, hence, understanding the complex regulatory networks controlled by sRNAs traveling from different genomes could assist the development of novel methods for transferring traits into crops, as alternatives to traditional breeding or transgenic approaches.</p>
<p>To identify sRNA population which are mobile or native, analysts currently adopt and adapt a general RNA sequencing pipeline. Several genomic studies have used grafting to explore and characterise mobile sRNAs in chimeric systems, however, it is clear that there is not a standardised approach.</p>
<p>Here we introduce the R Package mobileRNA, a recommend pipeline and analysis workflow for the exploration and identification of mobile small RNAs for chimeric systems and small RNA population changes between treatment and condition of non-chimeric systems. The pipeline offers an optimised mapping/alignment of samples to genome assemblies to improve sRNA origin, and offers analysis steps to remove mapping errors, visualization of the results, and supports output for functional analysis. <br></p>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>This manual offers a pipeline for the analysis of sRNAseq data taken from chimeric systems. For example, plant grafting experiments between different cultivars and species (with distinct genome assemblies). This pipeline involves pre-processing and analysis steps for the identification of candidate mobile sRNA molecules and the exploration of population level changes in chimeric systems.</p>
<p>There are two clear area which must be optimised to make better predicts of the<br>
sRNA mobility and population dynamics in chimeric systems: * Optimisation of the alignment steps * Optimisation of analysis: sRNA classification and mapping error removal.</p>
<p>The first involves optimising the use of the mapping algorithm to best place sRNA to each genome. Here, we supply the mapping algorithms with as much information as possible to make its predictions. This includes supplying both genome assemblies at the same time, through the use of a merged genome. A merged genome is a single FASTA file containing multiple genome assemblies. In many databases, the chromosome names have been named in the same patterns. This means that when you merge multiple genome assemblies together it is very challenging to distinguish the individual genomes in the merged file. Here, we offer a function to merge two genome assemblies into a single file and is able to maintain the distinguishability by adding a distinct pre-fix to chromosome names in each genome. The other bit of extra information to supply to the aligner, is a list of de novo sRNA loci identified in each sample in the analysis. This ensures consistency across the analysis and helps prevent false assumptions. sRNA molecules are expected to have a very low abundance; It might be the case that in one replicate the read abundance of a given sRNA locus was not significant enough to record as a sRNA cluster, while in another replicate it was recorded. Therefore, it would be wrong to assume there are zero reads at the locus. Collectively, considering the extra information should optimise the alignment of sRNAs to the best locations within the merged genome.</p>
<p>The second area to optimise is the analysis in R. The analysis workflow can be completed in a simple 3-step process to identify potentially mobile sRNAs and similarly to explore the population-scale dynamics this can be achieved in a simple 4-step process. Across the analysis, additional features are available to advance the analysis. This includes plotting principal component analysis results, heatmaps and distribution plots, as well as including statistical analysis through the deployment of differential analysis methods (edgeR or DESeq2), and steps to assist functional analysis. At many of the core steps, there is the ability to remove mapping errors by comparing controls to the treatment. This is a step not obviously described by other workflows but adds huge value to the analysis. Similarly, an important core step to understand is consensus dicer classification (see function ). For a given sRNA cluster, each replicate at the mapping and clustering pre-processing step will have been predicted a sRNA classification. This analysis step uses this information to define the classification of the cluster based on the consensus across replicates. There is a range of parameters to control how these are drawn, making it either stringent or loose, which will alter the downstream analysis. Finally, the step to identify the candidate mobile molecules (see function ) also has a parameter which can either result in stringent or looser outcomes. These parameter utilises the number of replicates which contributed to the consensus dicer classification as a threshold value. For instance, it would be more stringent to select sRNA clusters which were unambiguously classified across condition replicates.</p>
<p>Here we introduce the use of statistical analysis to identify mobile sRNA molecules in chimeric systems, and it should be emphasized that this is an optional step. However, statistical analysis for the exploration of population dynamics should be a mandatory step. Here we include a single function which can be instructed to undertake differential analysis using either the edgeR or DESeq2 method (see function ). The function outputs additional columns to your working dataset, including the raw count mean, log2FoldChange, p-value and adjusted p-value.</p>
<div class="section level3">
<h3 id="workflow-summary">Workflow Summary<a class="anchor" aria-label="anchor" href="#workflow-summary"></a>
</h3>
<p><img align="right" width="300" src="../reference/figures/program_flow.png"></p>
<p>The summarised workflow is shown below, where it starts in R-Studio to merge the two genome assemblies into one, then the pre-processing moves into Linux to align each replicate to the merged reference and then back into R-Studio to undertake the analysis. <br><br><br></p>
<p>The analysis pipeline, splits into two tracks when trying to identify candidate mobile sRNAs and population dynamics. This can be seen in the diagram below: <br><br><br><img src="../reference/figures/complete_workflow_mobile_native3.png" width="700" height="750"><br><br></p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The latest version of <code>mobileRNA</code> can be installed via GitHub using the <code>devtools</code> package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"KJeynesCupper/mobileRNA"</span>, ref <span class="op">=</span> <span class="st">"main"</span><span class="op">)</span></span></code></pre></div>
<p>Load package to library:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KJeynesCupper/mobileRNA.git" class="external-link">mobileRNA</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="an-overview-of-the-data-used">An overview of the data used<a class="anchor" aria-label="anchor" href="#an-overview-of-the-data-used"></a>
</h3>
<p>For the following examples, a semi-synthetic small RNA-seq data set has been utilised to simulate the movement of sRNA molecules from an Tomato (<em>Solanium lycopersicum</em>) rootstock to a Eggplant (<em>Solanium melongena</em>) scion, a grafting system known to be compatible.</p>
<p>There are three heterograft replicate, where each is an individual tomato replicate spiked with the same random set of 150 tomato sRNA clusters. These are known as: * <code>heterograft_1</code> * <code>heterograft_2</code> * <code>heterograft_3</code></p>
<p>There are three self-graft replicates, where each is an individual tomato replicate without the spiked tomato sRNA clusters. These are:</p>
<ul>
<li><code>selfgraft_1</code></li>
<li><code>selfgraft_2</code></li>
<li><code>selfgraft_3</code></li>
</ul>
<p>The replicates mirror each other where, for instance, <code>heterograft_1</code> and <code>selfgraft_1</code> are the same replicate, either with or without the spiked clusters. These replicates serve for the aim to analyse sRNA mobility.</p>
<p>The data set, called <code>sRNA_data</code>, stores a matrix containing the pre-processed data from the experiment. As a user, this allows you to see what a full data set might look like and how you might expect it to change across the analysis.</p>
<p>These can be loaded in the R environment by using the following command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sRNA_data"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-organisation">Data organisation<a class="anchor" aria-label="anchor" href="#data-organisation"></a>
</h3>
<p>There are two key elements required for the pipeline analysis: * sRNA-seq sample replicates (.fastq/.fq) * Two reference genomes (.fasta/.fa)</p>
<p>And one additional element; to improve functional analysis: * Reference genome annotations (.gff/.gff3) <br><br></p>
<p>It is recommended to rename your files to names you wish for them to be represented as within the analysis and shown as labels in plots. Plus, it makes the analysis easier!</p>
<p>For example, instead of names such as:</p>
<ul>
<li>1: <code>sample_1.fq</code>
</li>
<li>2: <code>sample_2.fq</code>
</li>
<li>3: <code>sample_3.fq</code>
</li>
<li>4: <code>sample_4.fq</code>
</li>
<li>5: <code>sample_5.fq</code>
</li>
<li>6: <code>sample_6.fq</code>
</li>
</ul>
<p>For the example data set included in the package, here we have renamed the files based on the condition (treatment or control).For the hetero-grafts, where the is a eggplant scion and an tomato rootstock:</p>
<ul>
<li>1: <code>heterograft_1.fq</code>
</li>
<li>2: <code>heterograft_2.fq</code>
</li>
<li>3: <code>heterograft_3.fq</code>
</li>
</ul>
<p>and for the eggplant self-grafts:</p>
<ul>
<li>4: <code>selfgraft_1.fq</code>
</li>
<li>5: <code>selfgraft_2.fq</code>
</li>
<li>6: <code>selfgraft_3.fq</code> <br>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="pre-processing">Pre-Processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h2>
<p>The pre-processing steps involve cleaning raw data and aligning it to the merged genome. Going forward, the pipeline has assumed that the sRNA-seq samples have met quality control standards.</p>
<p>We recommend using <code>ShortStack</code> (<a href="https://github.com/MikeAxtell/ShortStack" class="external-link uri">https://github.com/MikeAxtell/ShortStack</a>) to aligns samples to the merged genome and undertake cluster analysis. This program was formulated for sRNAseq analysis, utilising Bowtie (Version 1) to map samples along with an algorithm to cluster sRNAs.</p>
<p>Here, we introduce an alternative mapping method for the analysis of plant heterograft samples. The heterograft system involves two genotypes; here the two genome references are merged into a single reference to which samples are aligned to. <br></p>
<div class="section level3">
<h3 id="pre-mapping">Pre-mapping<a class="anchor" aria-label="anchor" href="#pre-mapping"></a>
</h3>
<div class="section level4">
<h4 id="merging-genome-assemblies">Merging Genome Assemblies<a class="anchor" aria-label="anchor" href="#merging-genome-assemblies"></a>
</h4>
<p>Here we merge the two reference genomes into a single merged genome using the <code>RNAgenomeMerge()</code> function. This removes punctuation from chromosome names which could cause issues when mapping with ShortStack and it also adds a prefix to each chromosome name which is unique to each genome assembly. By default, the string <code>A_</code> is added to the reference genome supplied to <code>genomeA</code> and <code>B_</code> is added to the reference genome supplied to <code>genomeB.</code> These can be customized to the users preference, see the manual for more information.</p>
<p>Depending on where you attained your genome assembly, the same chromosome naming pattern might be the same for each assembly. Therefore, without altering the chromosome names there would be no way to distinguish the two genomes apart within the merged file. Moreover, it makes the analysis through this pipeline much smoother is chromosomes have distinguishable character differences.</p>
<p>In the example below, demonstrates how to use the function. For the example dataset, we would merge the assemblies of tomato (<em>Solanum lycopersicum</em>) and eggplant (<em>Solanum melongena</em>). Here, we have decided to customize the prefix where we will add the <code>SL</code> prefix to the tomato chromosomes and <code>SM</code> prefix to the eggplant chromosomes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/RNAmergeGenomes.html">RNAmergeGenomes</a></span><span class="op">(</span>genomeA <span class="op">=</span><span class="st">"/Users/user1/projectname/workplace/reference/ref1.fa"</span>,</span>
<span>               genomeB <span class="op">=</span><span class="st">"/Users/user1/projectname/workplace/reference/ref2.fa"</span>,</span>
<span>               abbreviationGenomeA <span class="op">=</span> <span class="st">"SL"</span>,</span>
<span>               abbreviationGenomeB <span class="op">=</span> <span class="st">"SM"</span>,</span>
<span>               out_dir <span class="op">=</span> <span class="st">"/Users/user1/projectname/workplace/reference/merge/</span></span>
<span><span class="st">               ref_merged.fa"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="merging-genome-annotations">Merging Genome Annotations<a class="anchor" aria-label="anchor" href="#merging-genome-annotations"></a>
</h3>
<p>To identify genomic features which overlap the sRNA producing loci, we must overlap the data with the genome annotation files (GFF). However, if you are exploring a dataset which contains sRNA from both genomes, you will need to alter the chromosome names of the genome annotations in the same way we have for the genome assemblies. This can be undertaken using the <code><a href="../reference/RNAmergeAnnotations.html">RNAmergeAnnotations()</a></code> function.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="co">## merge the annotation files into a single annotation file</span></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="kw">RNAmergeAnnotation</span>(<span class="dt">annotationA =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="st">                   annotation/annotation_1.gff3"</span>,</span>
<span id="cb5-5"><a href="#cb5-5"></a>                   <span class="dt">annotationB =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-6"><a href="#cb5-6"></a><span class="st">                   annotation/annotation_1.gff3"</span>,</span>
<span id="cb5-7"><a href="#cb5-7"></a>                   <span class="dt">abbreviationAnnoA =</span> <span class="st">"SL"</span>,</span>
<span id="cb5-8"><a href="#cb5-8"></a>                   <span class="dt">abbreviationAnnoB =</span> <span class="st">"SM"</span></span>
<span id="cb5-9"><a href="#cb5-9"></a>                  <span class="dt">out_dir =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-10"><a href="#cb5-10"></a><span class="st">                  annotation/merge/anno_merged.gff"</span>)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="auto-detection-of-srna-cluster">Auto-Detection of sRNA Cluster<a class="anchor" aria-label="anchor" href="#auto-detection-of-srna-cluster"></a>
</h3>
<p>Here we identify and build a list sRNA clusters within each sample to assist the mapping step later on to ensure consistency across the analysis. This step is very important to consider in the analysis as it prevents false assumptions. This step aligns a replicate to the given genome assembly and the outputs contains loci where sRNA clusters were found. Across your replicates, there will be difference in the cluster identified. In some instance, it might be the case that the abundance of reads at a given sRNA locus was not significant enough to record as a sRNA cluster, while in another replicate it was recorded. Therefore, it would be wrong to assume there are zero reads at the loci. Here, we identify all the potential sRNA cluster loci and use this information to improve the mapping.</p>
<div class="section level4">
<h4 id="step-1---cluster-analysis-with-shortstack">Step 1 - Cluster analysis with ShortStack<a class="anchor" aria-label="anchor" href="#step-1---cluster-analysis-with-shortstack"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb6-2"><a href="#cb6-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb6-3"><a href="#cb6-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb6-4"><a href="#cb6-4"></a>--bowtie_cores 6 \</span>
<span id="cb6-5"><a href="#cb6-5"></a>--mmap n \</span>
<span id="cb6-6"><a href="#cb6-6"></a>--mismatches 0 \</span>
<span id="cb6-7"><a href="#cb6-7"></a>--nohp \</span>
<span id="cb6-8"><a href="#cb6-8"></a>--outdir <span class="op">&lt;</span>./output/directory<span class="op">&gt;</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-2---build-srna-cluster-list">Step 2 - Build sRNA cluster list<a class="anchor" aria-label="anchor" href="#step-2---build-srna-cluster-list"></a>
</h4>
<p>Now, we collate all the sRNA loci information from each sample into a text file.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># location of step 1 output</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>folder &lt;-<span class="st"> </span><span class="er">&lt;</span>.<span class="op">/</span>output<span class="op">/</span>directory<span class="op">/</span>from<span class="op">/</span>step<span class="op">/</span><span class="dv">1</span><span class="op">/</span><span class="er">&gt;</span></span>
<span id="cb7-3"><a href="#cb7-3"></a></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co"># name and location to save output file to (must be .txt)</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>save_folder &lt;-<span class="st"> </span><span class="er">&lt;</span>.<span class="op">/</span>output<span class="op">/</span>directory<span class="op">/</span>ClustersInfo.txt<span class="op">&gt;</span></span>
<span id="cb7-6"><a href="#cb7-6"></a></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co"># names of samples (ie. folder names)</span></span>
<span id="cb7-8"><a href="#cb7-8"></a>sample_names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"&lt;treatment_1&gt;"</span>, <span class="st">"&lt;treatment_2&gt;"</span>, <span class="st">"&lt;control_1&gt;"</span>,<span class="st">"&lt;control_2&gt;"</span>)</span>
<span id="cb7-9"><a href="#cb7-9"></a></span>
<span id="cb7-10"><a href="#cb7-10"></a></span>
<span id="cb7-11"><a href="#cb7-11"></a>loci_info &lt;-<span class="st"> </span><span class="kw">RNAloci</span>(<span class="dt">files =</span> folder, </span>
<span id="cb7-12"><a href="#cb7-12"></a>             <span class="dt">out =</span> save_folder,</span>
<span id="cb7-13"><a href="#cb7-13"></a>             <span class="dt">samples =</span> sample_names)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="mapping">Mapping<a class="anchor" aria-label="anchor" href="#mapping"></a>
</h3>
<p>Each sample is mapped to the merged reference genome with the list of sRNA clusters.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb8-2"><a href="#cb8-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb8-3"><a href="#cb8-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb8-4"><a href="#cb8-4"></a>--locifile <span class="op">&lt;</span>./output/directory/ClustersInfo.gff3<span class="op">&gt;</span> \</span>
<span id="cb8-5"><a href="#cb8-5"></a>--bowtie_cores 6 \</span>
<span id="cb8-6"><a href="#cb8-6"></a>--mmap n \</span>
<span id="cb8-7"><a href="#cb8-7"></a>--mismatches 0 \</span>
<span id="cb8-8"><a href="#cb8-8"></a>--nohp \</span>
<span id="cb8-9"><a href="#cb8-9"></a>--mincov 5 \</span>
<span id="cb8-10"><a href="#cb8-10"></a>--outdir <span class="op">&lt;</span>./output/directory/step2/<span class="op">&gt;</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>Here, we will analyses the output of the pre-processed data to identify candidate mobile small RNA molecules and native small RNA population changes. <br></p>
<div class="section level3">
<h3 id="import-data">Import Data<a class="anchor" aria-label="anchor" href="#import-data"></a>
</h3>
<p>In the pre-processing steps, the data was cleaned and aligned to the merged reference genome. During the mapping step, a folder for each sample is created storing the results and we are interested in the file <code>Results.txt</code>. The <code><a href="../reference/RNAimport.html">RNAimport()</a></code> function imports the <code>Results.txt</code> file for each sample, extracting specific pieces of information and storing it all in a single working dataframe. To do so, the function requires the path to the directory path which stores the output folders produced by ShortStack. Along with a vector containing the names of each sample - this must correspond to the folder names previously mentioned to which the directory path leads to.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Import &amp; organise data.</span></span>
<span><span class="va">results_dir</span> <span class="op">&lt;-</span>  <span class="st">"./analysis/alignment_unique_two/"</span></span>
<span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>,<span class="st">"heterograft_2"</span>, <span class="st">"heterograft_3"</span>,</span>
<span>                  <span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span>, <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">sRNA_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAimport.html">RNAimport</a></span><span class="op">(</span>input <span class="op">=</span> <span class="st">"sRNA"</span>,</span>
<span>                       directory <span class="op">=</span> <span class="va">results_dir</span>,</span>
<span>                       samples <span class="op">=</span> <span class="va">sample_names</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="sample-level-quality-control">Sample-level quality control<a class="anchor" aria-label="anchor" href="#sample-level-quality-control"></a>
</h3>
<p>A handy step in the analysis is to assess the overall similarity between sample replicates to understand which samples are the similar and/or different. As well as understand where the most variation is introduced and whether the data meets your expectations. It is expected that between the conditions, the sample replicates show enough variation to suggest that the replicates are from different groups.</p>
<p>To investigate the sample similarity/difference, we will undertake sample-level quality control using three different methods:</p>
<ul>
<li>Distribution of RNA classes within each sample</li>
<li>Principle component analysis (PCA)</li>
<li>Heatmap using hierarchical clustering</li>
</ul>
<p>This will show us how well samples within each condition cluster together, which may highlight outliers. Plus, to show whether our experimental conditions represent the main source of variation in the data set.</p>
<p>Here we will be employing an unsupervised clustering methods for the PCA and hierarchical clustering Heatmap. This involves an unbiased log2-transformation of the raw counts which will emphasis the the sample clustering to improve visualization. The DESeq2 package contains a particularly useful function to undertake regularized log transformation (rlog) which controls the variance across the mean, and in this package we have utilized this for the quality control steps. <br></p>
<div class="section level4">
<h4 id="plot-the-distribution-of-rna-classes-within-each-sample">Plot the distribution of RNA classes within each sample<a class="anchor" aria-label="anchor" href="#plot-the-distribution-of-rna-classes-within-each-sample"></a>
</h4>
<p>Here, we can generate a number of different customized plots which represent the number of each small RNA class identified within a sample. The function, <code><a href="../reference/RNAdistribution.html">RNAdistribution()</a></code> outputs a table and the plots.</p>
<p>The results can be plotted as either a bar chart (<code>style = "bar"</code>) or a line graph (<code>style = "line"</code>) for each sample and can be shown in a facet (<code>facet = TRUE</code>) or in as a single line graph, where each line represents a sample (<code>together=TRUE</code>). Alternatively, the results can be plotted individually for each sample in either format (<code>together = FALSE</code>).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># plot each replicate as a line, separately, and facet </span></span>
<span><span class="va">sample_distribution_line</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span><span class="va">sRNA_data</span>,</span>
<span>                                            style <span class="op">=</span> <span class="st">"line"</span>,</span>
<span>                                            together <span class="op">=</span> <span class="cn">FALSE</span>,facet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>                                            </span>
<span> <span class="co"># plot each replicate as a bar, separately, and facet </span></span>
<span><span class="va">sample_distribution_bar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span><span class="va">sRNA_data</span>,</span>
<span>                                           style <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>                                           facet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot each replicate on a single plot</span></span>
<span><span class="va">sample_distribution</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span><span class="va">sRNA_data</span>,</span>
<span>                                           style <span class="op">=</span> <span class="st">"line"</span>,</span>
<span>                                           together <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Lets view these plots:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View plot (only)</span></span>
<span><span class="va">sample_distribution_line</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-9-1.png" alt="An example facet line graph to show the distribution of RNA classes within each sample." width="700"><p class="caption">
An example facet line graph to show the distribution of RNA classes within each sample.
</p>
</div>
<p><br></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View plot (only)</span></span>
<span><span class="va">sample_distribution_bar</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-10-1.png" alt="An example facet bar graph to show the distribution of RNA classes within each sample." width="700"><p class="caption">
An example facet bar graph to show the distribution of RNA classes within each sample.
</p>
</div>
<p><br></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View plot</span></span>
<span><span class="va">sample_distribution</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-11-1.png" alt="An example line graph showing all the replicates plotted on a single plot." width="700"><p class="caption">
An example line graph showing all the replicates plotted on a single plot.
</p>
</div>
</div>
<div class="section level4">
<h4 id="principle-component-analysis-to-assess-sample-distance">Principle component analysis to assess sample distance<a class="anchor" aria-label="anchor" href="#principle-component-analysis-to-assess-sample-distance"></a>
</h4>
<p>Principal Component Analysis (PCA) is a useful technique to illustrate sample distance as it emphasizes the variation through the reduction of dimensions in the data set. Here, we introduce the function <code><a href="../reference/plotSamplePCA.html">plotSamplePCA()</a></code></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSamplePCA.html">plotSamplePCA</a></span><span class="op">(</span><span class="va">sRNA_data</span>, <span class="va">group</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking and organising data </span></span>
<span><span class="co">#&gt; Transforming the count data to the log2 scale </span></span>
<span><span class="co">#&gt; Organising principal component analysis</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-13-1.png" alt="An example of a PCA, illustracting the sRNA data set sample similarity" width="700"><p class="caption">
An example of a PCA, illustracting the sRNA data set sample similarity
</p>
</div>
<p>This PCA is somewhat not what we would expect to see in a real dataset. This is because PC1 is showing the variation between the non-spiked and spiked version of the replicate while PC2 shows the natural variation between the replicates.</p>
<p>It is more likely that in a real small RNA dataset, you might expect more or less variation between the conditions. For example, analysis of a self-graft vs heterograft may show little variation between the conditions. This might be expected because there may only be a very small number of endogenous mobile/foreign small RNAs that are present in the whole population while the population of small RNAs present in the destination tissue is expected to be largely similar. Although, there is likely to be some difference in this too due to the endogenous changes. <br></p>
</div>
<div class="section level4">
<h4 id="hierarchical-clustered-heatmap-to-assess-sample-distance">Hierarchical clustered heatmap to assess sample distance<a class="anchor" aria-label="anchor" href="#hierarchical-clustered-heatmap-to-assess-sample-distance"></a>
</h4>
<p>Similarly, to a PCA plot, the <code><a href="../reference/plotSampleDistance.html">plotSampleDistance()</a></code> function undertakes hierarchical clustering with an unbiased log transformation to calculate sample distance and is plotted in the form of a heatmap.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plotSampleDistance.html">plotSampleDistance</a></span><span class="op">(</span><span class="va">sRNA_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking data </span></span>
<span><span class="co">#&gt; Transforming the count data to the log2 scale </span></span>
<span><span class="co">#&gt; Calculating distance matrix </span></span>
<span><span class="co">#&gt; Creating sample distance plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-15-1.png" alt="An example of a heatmap,illustrating the sRNA data set sample similarity" width="700"><p class="caption">
An example of a heatmap,illustrating the sRNA data set sample similarity
</p>
</div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="define-the-consensus-rna-class">Define the consensus RNA class<a class="anchor" aria-label="anchor" href="#define-the-consensus-rna-class"></a>
</h3>
<p>Have a look at the <code>sRNA_data</code> object, you will see that for each sample the sRNA class for a given cluster has been determined (see columns with names containing with “DicerCall”) which will state a number from 20-24 in the data which represents the nucleotide length of the small RNA, or, state NA/N which represents that the small RNA was unclassified in the sample.</p>
<p>The <code><a href="../reference/RNAdicercall.html">RNAdicercall()</a></code> function is used to calculate the consensus dicer classification for each small RNA cluster based of the classification by each sample replicate.</p>
<p>When working with samples from a chimeric system, the function contains a specialised parameter which should be utilized. This is <code>chimeric=TRUE</code>, along with the <code>genome.ID</code> and <code>controls</code> parameters. This helps optimise classification by removal of potential mapping errors.</p>
<p>The used can alter the functions decision making by telling it to base the consensus on specific samples or across all samples in the analysis (default). Plus, if the user utilises the <code>tidy=TRUE</code> parameter, this will remove small RNA cluster which have an unclassified (N) consensus dicer classification - reducing noise in the data.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># define consensus, store as a data summary file.</span></span>
<span><span class="va">sRNA_data_consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdicercall.html">RNAdicercall</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data</span>, </span>
<span>                                     chimeric <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                     genome.ID <span class="op">=</span> <span class="st">"SL"</span>, </span>
<span>                                     controls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, </span>
<span>                                                  <span class="st">"selfgraft_2"</span>, </span>
<span>                                                  <span class="st">"selfgraft_3"</span><span class="op">)</span>,</span>
<span>                                     tidy<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating consensus dicercall based on information from all replicates... </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; The consensus dicercall will be excluded in the case of a tie... </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Removing small RNA clusters with no consensus dicercall...</span></span></code></pre></div>
<p>For the downstream analysis, it can be useful to define the distinct 24-nt and 21/22-nt populations of small RNAs, To subset the data, use the <code><a href="../reference/RNAsubset.html">RNAsubset()</a></code> function to choose which small RNA populations to include: 20-nt, 21-nt, 22-nt, 23-nt, or 24-nt in this data. For example:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Subset data for analysis: 24-nt sRNAs</span></span>
<span><span class="va">sRNA_24</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsubset.html">RNAsubset</a></span><span class="op">(</span><span class="va">sRNA_data_consensus</span>, type <span class="op">=</span> <span class="fl">24</span><span class="op">)</span></span>
<span><span class="co"># Subset data for analysis: 24 21/22-nt sRNAs</span></span>
<span><span class="va">sRNA_2122</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsubset.html">RNAsubset</a></span><span class="op">(</span><span class="va">sRNA_data_consensus</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">21</span>, <span class="fl">22</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="plot-consensus-classification-of-srna">Plot consensus classification of sRNA<a class="anchor" aria-label="anchor" href="#plot-consensus-classification-of-srna"></a>
</h4>
<p>Similar to before, the <code>RNAdistriution()</code> function can be used to visualize the distribution of the consensus sRNA classes.</p>
<p>The function can plot the absolute or relative frequency, use argument <code>relative</code>. It can be beneficial to view the absolute frequency as well as the relative to observe whether the data meets your expectations.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">consensus_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>,</span>
<span>                                  consensus <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                  style <span class="op">=</span> <span class="st">"consensus"</span><span class="op">)</span></span></code></pre></div>
<p>Now, view the plot:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># view </span></span>
<span><span class="va">consensus_plot</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-20-1.png" alt="An example of the distribution of small RNA consensus dicer classifications. " width="700"><p class="caption">
An example of the distribution of small RNA consensus dicer classifications.
</p>
</div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis-with-deseq2-or-edger">Differential Expression analysis with DESeq2 or edgeR<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-with-deseq2-or-edger"></a>
</h3>
<p>Differential expression (DE) analysis is undertaken to identify small RNA which are statistically significant to discover quantitative changes in the abundance levels between the treatment (hetero-grafting) and the control (self-grafting) groups. This technique can be undertaken with a variety of tools, in <code>mobileRNA</code> users have the option to use the <code>DESeq2</code> or <code>edgeR</code> analytical method.</p>
<p>Based on your data, one analytical method might be preferable over another. For instance, the applied <code>DESeq2</code> method is not appropriate when the experiment does not contain replicate (ie. one sample replicate per condition). While, <code>edgeR</code> can be used. Here, we have included the recommend practice for edgeR when the data does not contain replicates. This option can be employed by setting a custom dispersion value, see argument <code>dispersionValue</code>.</p>
<p>Please note that this step is optional and not a required part of the analysis.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># sample conditions in order within dataframe</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Differential analysis of whole dataset: DESeq22 method </span></span>
<span><span class="va">sRNA_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"DESeq2"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="identify-candiate-mobile-small-rnas">Identify candiate mobile small RNAs<a class="anchor" aria-label="anchor" href="#identify-candiate-mobile-small-rnas"></a>
</h2>
<p>It is essential to remove the noise from the data to isolate potential mobile molecules which include falsely mapped clusters and, in this case, the clusters which map the the scion genotype when the aim is to identify molecules traveling from the rootstock. In the data, the aim is to explore whether there are tomato sRNA traveling up through the graft junction and exiting into the eggplant leaf tissues. We can identify potential mobile molecule by using the <code><a href="../reference/RNAmobile.html">RNAmobile()</a></code> function. It selects clusters mapping to the rootstock genome in the hetero-graft. Therefore, we will remove clusters mapped to chromosome in the genotype used in the self-graft controls.</p>
<p>In the example, the tomato genome has chromosome labeled with “SL40” whilst the the eggplant genome has no labels. To remove clusters associated with the eggplant genome and keep only clusters associated with the tomato genome, set the <code>"task"</code> argument as <code>"keep"</code>.</p>
<p>The <code><a href="../reference/RNAmobile.html">RNAmobile()</a></code> function can take into account the statistically significance. If set to consider this, the function will filter sRNA based on an adjusted p-value threshold = 0.05. The threshold can be changed to make it less stringent by using the argument <code>padj</code>.Similarly, if you would prefer to extract the mobile RNA based on the p-value,rather than the adjusted p-values, a numeric threshold can be set for the argument <code>p.value</code>. As default, the function does not consider the statistical values and properties of the sRNAs. To consider, use <code>statistical=TRUE</code></p>
<p>Below we have used the datasets which contains the additional columns with the statistical values. As stated above, we do not need to use the statistical values to identify the mobile molecules but it can be explored and utilised to plot a heatmap based on the logFC.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># vector of control names</span></span>
<span><span class="va">control_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span>, <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Identify potential tomato mobile molecules</span></span>
<span><span class="va">mobile_sRNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         genome.ID <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span>, </span>
<span>                         statistical <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Identify potential tomato mobile molecules which are statistically significant</span></span>
<span><span class="va">mobile_sRNA_significant</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         genome.ID <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span>, </span>
<span>                         statistical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="co"># change this parameter</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="heatmap-plots-to-represent-mobile-molecules">Heatmap plots to represent mobile molecules<a class="anchor" aria-label="anchor" href="#heatmap-plots-to-represent-mobile-molecules"></a>
</h4>
<p>We can plot our results as a heatmap - this heatmap represents the normalised reads-per-million (RPM) values which have been log transformed.<br><br></p>
<p>Here we will plot all potential mobile molecules and those which are statistically significant:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p10</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">mobile_sRNA</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-23-1.png" alt="An example heatmap of candidate mobile small RNAs.Where the columns represent the sample replicates and the rows represent the small RNA cluster." width="700"><p class="caption">
An example heatmap of candidate mobile small RNAs.Where the columns represent the sample replicates and the rows represent the small RNA cluster.
</p>
</div>
<p><br></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p11</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">mobile_sRNA_significant</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-24-1.png" alt="An example heatmap of candidate mobile small RNAs, which are statistically signifcant (adjusted pvalue &lt; 0.05). Where the columns represent the sample replicates and the rows represent the small RNA cluster." width="700"><p class="caption">
An example heatmap of candidate mobile small RNAs, which are statistically signifcant (adjusted pvalue &lt; 0.05). Where the columns represent the sample replicates and the rows represent the small RNA cluster.
</p>
</div>
</div>
<div class="section level4">
<h4 id="save-output">Save output<a class="anchor" aria-label="anchor" href="#save-output"></a>
</h4>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">mobile_sRNA</span>, <span class="st">"./candidate_mobile_sRNAs.txt"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="functional-analysis">Functional Analysis<a class="anchor" aria-label="anchor" href="#functional-analysis"></a>
</h3>
<p>At this point in the analysis, a dataframe contain candidate mobile small RNAs has been produced. Now, we can extrapolate information to assist the identification of their potential effect and role in the biological system.</p>
<p>We offer three different tools to assist the functional analysis. The first, can be used to identify genomic features associated with the small RNA producing locus associated to each cluster. This function is known as <code><a href="../reference/RNAattributes.html">RNAattributes()</a></code>. The second tool does something very similar, it identifies the same overlaps but returns a value stating either the absolute or relative number of sRNA cluster which over overlap with the different genomic features. This function is known as <code><a href="../reference/RNAfeatures.html">RNAfeatures()</a></code>. Finally, the last tools, called <code><a href="../reference/RNAsequences.html">RNAsequences()</a></code> calculates the consensus small RNA sequence for each candidate and predicts the potential target sequences.</p>
<p>The consensus RNA sequences for the candidates can be used to predict targets in a given genome using tools such as <code>psRNATarget</code> (Xinbin Dai, Zhaohong Zhuang and Patrick X. Zhao (2018). psRNATarget: a plant small RNA target analysis server (2017 release). Nucleic Acids Research. doi: 10.1093/nar/gky316).</p>
<p>IMPORTANT: If chromosome names were altered in either of genome assemblies used in the mapping steps, the chromosome names in the genome annotation files must match the same pattern. If you used the [mobileRNA::RNAmergeGenomes()] function, the chromosome names will by default be altered. The [mobileRNA::RNAmergeAnnotations()] function can be utilised to alter both genome annotation files in the same way. This functions alters the names, saves each altered genome annotation and then merges them into one. <br></p>
<div class="section level4">
<h4 id="add-genomic-attributes-to-srna-clusters">Add genomic attributes to sRNA clusters<a class="anchor" aria-label="anchor" href="#add-genomic-attributes-to-srna-clusters"></a>
</h4>
<p>Each sRNA cluster contains coordinates, these can be matches with coordinates in an annotation file. A match occurs when the cluster is found within the coordinates of a feature. If there is a match, the function returns the input data with additional fields of information from the annotation file.</p>
<p>Using the genomic locations of the mobile sRNA clusters, we can identify overlaps in the genome annotation of the origin tissue to predict the origin of the mobile small RNA. Here we will only be overlapping the data with genes, and adding a buffer region of 1 kb upstream and downstream of the each gene.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mobile_df_attributes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAattributes.html">RNAattributes</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mobile_sRNA</span>, match <span class="op">=</span><span class="st">"genes"</span>,</span>
<span>                            annotation <span class="op">=</span> <span class="st">"./annotation/origin_annotation.gff3"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="summarise-overlaps-with-genomic-features">Summarise overlaps with genomic features<a class="anchor" aria-label="anchor" href="#summarise-overlaps-with-genomic-features"></a>
</h4>
<p>Very similar to before, we can find overlaps between our candidate small RNA clusters and genomic features. However, this time we can calculate the relative or absolute number of small RNA clusters which are associated to each type of feature. These include promoter regions, exon, introns, untranslated regions and repeat regions.The results can either be displayed in the matrix as an absolute value or as a percentage of the total:</p>
<p>NOTE: This function look for the exact matches in the locus (chromosome, start and end regions) between the sRNA cluster and the annotation file.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mobile_df_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAfeatures.html">RNAfeatures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mobile_sRNA</span>,</span>
<span>                            annotation <span class="op">=</span> <span class="st">"./annotation/origin_annotation.gff3"</span>, </span>
<span>                      repeats <span class="op">=</span> <span class="st">"./annotation/origin_annotation_repeats.gff3"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="retrieve-srna-sequence-complementary-sequences">Retrieve sRNA sequence &amp; complementary sequences<a class="anchor" aria-label="anchor" href="#retrieve-srna-sequence-complementary-sequences"></a>
</h4>
<p>We can speculate the potential targets of the mobile small RNAs using target prediction tools such as <code>psRNATarget</code> (Xinbin Dai, Zhaohong Zhuang and Patrick X. Zhao (2018). psRNATarget: a plant small RNA target analysis server (2017 release). Nucleic Acids Research. doi: 10.1093/nar/gky316).</p>
<p>The <code><a href="../reference/RNAsequences.html">RNAsequences()</a></code> function can be used to extrapolate the consensus RNA sequence and also calculate the complementary sequences.</p>
<p>The <code>RNAsequences</code> function does so by identifying whether the most abundant small RNA is consistent across the replicates, and if so, it extracts the sRNA nucleotide sequence and calculates the RNA and DNA complementary sequences, as well as stating the length of the sequence. However, if there is a tie between the he most abundant small RNAs within a cluster the user can utilise the <code>duplicates</code> parameter to choose whether to set the consensus sequence as NA or to pick one at random. Similarly, the make the analysis more stringent there is a threshold parameter which can be used to set the minimum number of replicates required to share the sRNA sequence to count as a match.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mobile_sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsequences.html">RNAsequences</a></span><span class="op">(</span><span class="va">mobile_sRNA</span><span class="op">)</span></span>
<span><span class="co">#&gt; The minimum number of matches required to form a consensus sRNA sequence is...  1 </span></span>
<span><span class="co">#&gt; The consensus sRNA sequences will be choose at random in the case of a tie</span></span></code></pre></div>
<p><br></p>
<p>We can extract the consensus small RNA sequences using the following code:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co"># select the cluster and sequence columns </span></span>
<span><span class="va">sequences</span> <span class="op">&lt;-</span><span class="va">mobile_sequences</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Cluster</span>, <span class="va">Sequence</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add prefix, remove row with NA</span></span>
<span><span class="va">prefix</span> <span class="op">&lt;-</span> <span class="st">"&gt;"</span></span>
<span><span class="va">sequences</span><span class="op">$</span><span class="va">Cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">prefix</span>, <span class="va">sequences</span><span class="op">$</span><span class="va">Cluster</span><span class="op">)</span></span>
<span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># convert</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">)</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">sequences</span><span class="op">[</span><span class="va">i</span>, <span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># save output </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">res</span>, file <span class="op">=</span><span class="st">"./candidate_mobile_sRNAs_sequences.txt"</span> ,</span>
<span>            row.names <span class="op">=</span> <span class="cn">FALSE</span>, col.names <span class="op">=</span> <span class="cn">FALSE</span>, quote <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The output can be used as input for tools such as <code>psRNATarget</code> (Xinbin Dai, Zhaohong Zhuang and Patrick X. Zhao (2018). psRNATarget: a plant small RNA target analysis server (2017 release). Nucleic Acids Research. doi: 10.1093/nar/gky316). <br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="explore-small-rna-population-difference-between-treatment-and-control">Explore small RNA population difference between treatment and control<a class="anchor" aria-label="anchor" href="#explore-small-rna-population-difference-between-treatment-and-control"></a>
</h2>
<p>Here, we will look at comparing the native population of sRNAs and how they might have changes in the tissue due to the chimeric system compared to the control.</p>
<p>As a reminder, the data used here will not yield any results as the treatment and control samples contain the exact same population of eggplant sRNAs, the only difference in the treatment samples are the spiked tomato sRNA clusters.</p>
<p>We have already undertaken the statistical analysis in the previous steps, we will be utilising this dataset going forward.</p>
<div class="section level4">
<h4 id="srna-abundance">sRNA abundance<a class="anchor" aria-label="anchor" href="#srna-abundance"></a>
</h4>
<p>When comparing treatment to control conditions, it might be the case that the same sRNA clusters are found within both, yet, there could be difference in the total abundance of the shared clusters. For instance, for a given sRNA cluster the samples in the treatment condition might have a greater abundance than the samples in the control condition. These difference could attribute to observed differences.</p>
<p>The statistical analysis calculated the log2FC values for each sRNA cluster by comparing the normalised counts between treatment and control. Here, a positive log2FC indicates an increased abundance of transcripts for a given sRNA cluster in the treatment compared to the control, while negative log2FC indicates decreased abundance of transcripts for a given sRNA cluster. The statistical significance of the log2FC is determined by the (adjusted) pvalue also calculated in the previous step.</p>
<p>Here we will filter the data to select sRNA clusters which are statistically significant, and then plot the results as a heatmap to compare the conditions.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select only significant sRNAs</span></span>
<span><span class="va">significant_sRNAs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsignificant.html">RNAsignificant</a></span><span class="op">(</span><span class="va">sRNA_DESeq2</span><span class="op">)</span></span></code></pre></div>
<p>Plot the results: <br></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#plot</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">significant_sRNAs</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="mobileRNA_files/figure-html/unnamed-chunk-31-1.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="step-6-identify-srna-population-difference">Step 6: Identify sRNA population difference<a class="anchor" aria-label="anchor" href="#step-6-identify-srna-population-difference"></a>
</h3>
<p>Finally, the <code><a href="../reference/RNApopulation.html">RNApopulation()</a></code> function can be utilised to identify unique sRNA populations found in the treatment or control conditions.</p>
<p>First lets look at the sRNA clusters unique to the treatment condition. In the chimeric heterografts, we expect that the foreign sRNAs will also be selected in this pick-up, therefore, we can use the parameter <code>genome.ID</code> to remove sRNA cluster related to the foreign genome.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># select sRNA clusters only found in treatment &amp; not in the control samples</span></span>
<span><span class="va">treatment_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span> , <span class="st">"heterograft_3"</span><span class="op">)</span></span>
<span><span class="va">unique_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNApopulation.html">RNApopulation</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_edgeR</span>, </span>
<span>                                  conditions <span class="op">=</span> <span class="va">treatment_reps</span>,</span>
<span>                                  chimeric <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                  genome.ID <span class="op">=</span> <span class="st">"SL"</span>, </span>
<span>                                  controls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>,</span>
<span>                                               <span class="st">"selfgraft_2"</span>, </span>
<span>                                               <span class="st">"selfgraft_3"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># look at number of sRNA cluster only found in treatment </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">unique_treatment</span><span class="op">)</span></span></code></pre></div>
<p>Now, the sRNA clusters unique to the control condition:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># select sRNA clusters only found in control &amp; not in the treatment samples</span></span>
<span><span class="va">control_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span> , <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span><span class="va">unique_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNApopulation.html">RNApopulation</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_edgeR</span>, </span>
<span>                                   conditions <span class="op">=</span> <span class="va">control_reps</span><span class="op">)</span></span>
<span><span class="co"># look at number of sRNA cluster only found in control  </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">unique_control</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p>The data here can be manipulated in many ways, you may wish to extract statistically significant small RNA clusters within a population to explore further. For this you can utilise the function.</p>
<p>The plotting functions within the package can be used to display these results, specifically and . Similarly, the functions to assist with functional analysis can also be utilised to pull the consensus sRNA sequence for each cluster in the population, or find overlaps with known annotations in the native genome. <br></p>
<p>Similarly if you are comparing several different treatments, a venn diagram could be plotted to observe whether any unique populations of small RNA are shared.</p>
</div>
</div>
<div class="section level2">
<h2 id="additional-features">Additional features<a class="anchor" aria-label="anchor" href="#additional-features"></a>
</h2>
<p>The <em>mobileRNA</em> package offers several optional extra-helpful functions to aid analysis. This includes a function to calculate the mean RPM and counts across specific samples and also calculate the overlap between genomic features and sRNA clusters.</p>
<p><br></p>
<div class="section level3">
<h3 id="calculate-rpm-and-count-means-for-specific-samples">Calculate RPM and Count means for specific samples<a class="anchor" aria-label="anchor" href="#calculate-rpm-and-count-means-for-specific-samples"></a>
</h3>
<p>This function calculates the RPM mean and Count mean across specific samples.The output of the calculation is added as an additional column to the input dataframe.</p>
<p>This can be manipulated to be used before or after identified mobile RNAs. When calculating means after identifying mobile molecules, it is recommend to base this on the heterograft samples rather than the full dataset (self-grafts + heterograft replicates) as all potential mobile sRNAs should have zero counts or RPM value in the self-grafts control replicates. By including the full dataset in these calculations, the mean values for the mobile molecules will be skewed.</p>
<p>Therefore, depending on your intended output this function could be approached in multiple ways.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># calculate mean RPM and mean Counts for heterograft samples</span></span>
<span><span class="va">selected_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span>, <span class="st">"heterograft_3"</span><span class="op">)</span></span>
<span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmean.html">RNAmean</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>, conditions <span class="op">=</span> <span class="va">selected_samples</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.0.2 (2020-06-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS  10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] dplyr_1.1.2      mobileRNA_0.99.8 magrittr_2.0.3   BiocStyle_2.16.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] colorspace_2.0-3            rprojroot_2.0.2            </span></span>
<span><span class="co">#&gt;   [3] DNAcopy_1.62.0              XVector_0.28.0             </span></span>
<span><span class="co">#&gt;   [5] GenomicRanges_1.40.0        fs_1.5.2                   </span></span>
<span><span class="co">#&gt;   [7] rstudioapi_0.14             farver_2.1.0               </span></span>
<span><span class="co">#&gt;   [9] listenv_0.8.0               audio_0.1-10               </span></span>
<span><span class="co">#&gt;  [11] affyio_1.58.0               ggrepel_0.9.1              </span></span>
<span><span class="co">#&gt;  [13] bit64_4.0.5                 AnnotationDbi_1.50.3       </span></span>
<span><span class="co">#&gt;  [15] fansi_1.0.2                 codetools_0.2-18           </span></span>
<span><span class="co">#&gt;  [17] splines_4.0.2               Repitools_1.34.0           </span></span>
<span><span class="co">#&gt;  [19] doParallel_1.0.17           cachem_1.0.6               </span></span>
<span><span class="co">#&gt;  [21] geneplotter_1.66.0          knitr_1.42                 </span></span>
<span><span class="co">#&gt;  [23] jsonlite_1.8.4              Rsamtools_2.4.0            </span></span>
<span><span class="co">#&gt;  [25] annotate_1.66.0             cluster_2.1.2              </span></span>
<span><span class="co">#&gt;  [27] vsn_3.56.0                  pheatmap_1.0.12            </span></span>
<span><span class="co">#&gt;  [29] BiocManager_1.30.16         compiler_4.0.2             </span></span>
<span><span class="co">#&gt;  [31] Matrix_1.4-0                fastmap_1.1.0              </span></span>
<span><span class="co">#&gt;  [33] limma_3.44.3                cli_3.6.0                  </span></span>
<span><span class="co">#&gt;  [35] beepr_1.3                   htmltools_0.5.5            </span></span>
<span><span class="co">#&gt;  [37] tools_4.0.2                 gtable_0.3.0               </span></span>
<span><span class="co">#&gt;  [39] glue_1.6.2                  GenomeInfoDbData_1.2.3     </span></span>
<span><span class="co">#&gt;  [41] affy_1.66.0                 Rcpp_1.0.8.3               </span></span>
<span><span class="co">#&gt;  [43] Biobase_2.48.0              jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [45] pkgdown_2.0.7               vctrs_0.6.2                </span></span>
<span><span class="co">#&gt;  [47] Biostrings_2.56.0           preprocessCore_1.50.0      </span></span>
<span><span class="co">#&gt;  [49] progressr_0.10.0            rtracklayer_1.48.0         </span></span>
<span><span class="co">#&gt;  [51] iterators_1.0.14            xfun_0.37                  </span></span>
<span><span class="co">#&gt;  [53] stringr_1.5.0               globals_0.14.0             </span></span>
<span><span class="co">#&gt;  [55] lifecycle_1.0.3             gtools_3.9.2               </span></span>
<span><span class="co">#&gt;  [57] XML_3.99-0.9                future_1.24.0              </span></span>
<span><span class="co">#&gt;  [59] edgeR_3.30.3                zlibbioc_1.34.0            </span></span>
<span><span class="co">#&gt;  [61] MASS_7.3-55                 scales_1.2.1               </span></span>
<span><span class="co">#&gt;  [63] BSgenome_1.56.0             ragg_1.2.2                 </span></span>
<span><span class="co">#&gt;  [65] parallel_4.0.2              SummarizedExperiment_1.18.2</span></span>
<span><span class="co">#&gt;  [67] RColorBrewer_1.1-2          curl_4.3.2                 </span></span>
<span><span class="co">#&gt;  [69] yaml_2.3.5                  pbapply_1.7-0              </span></span>
<span><span class="co">#&gt;  [71] memoise_2.0.1               ggplot2_3.4.2              </span></span>
<span><span class="co">#&gt;  [73] sass_0.4.0                  stringi_1.7.6              </span></span>
<span><span class="co">#&gt;  [75] RSQLite_2.2.10              highr_0.9                  </span></span>
<span><span class="co">#&gt;  [77] genefilter_1.70.0           S4Vectors_0.26.1           </span></span>
<span><span class="co">#&gt;  [79] desc_1.4.1                  foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [81] Ringo_1.52.0                caTools_1.18.2             </span></span>
<span><span class="co">#&gt;  [83] BiocGenerics_0.34.0         BiocParallel_1.22.0        </span></span>
<span><span class="co">#&gt;  [85] truncnorm_1.0-8             GenomeInfoDb_1.24.2        </span></span>
<span><span class="co">#&gt;  [87] rlang_1.1.0                 pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [89] systemfonts_1.0.4           matrixStats_0.61.0         </span></span>
<span><span class="co">#&gt;  [91] bitops_1.0-7                Rsolnp_1.16                </span></span>
<span><span class="co">#&gt;  [93] evaluate_0.15               lattice_0.20-45            </span></span>
<span><span class="co">#&gt;  [95] purrr_1.0.1                 labeling_0.4.2             </span></span>
<span><span class="co">#&gt;  [97] GenomicAlignments_1.24.0    bit_4.0.4                  </span></span>
<span><span class="co">#&gt;  [99] tidyselect_1.2.0            parallelly_1.30.0          </span></span>
<span><span class="co">#&gt; [101] bookdown_0.32               DESeq2_1.28.1              </span></span>
<span><span class="co">#&gt; [103] R6_2.5.1                    IRanges_2.22.2             </span></span>
<span><span class="co">#&gt; [105] gplots_3.1.1                generics_0.1.2             </span></span>
<span><span class="co">#&gt; [107] DelayedArray_0.14.1         DBI_1.1.2                  </span></span>
<span><span class="co">#&gt; [109] withr_2.5.0                 pillar_1.9.0               </span></span>
<span><span class="co">#&gt; [111] gsmoothr_0.1.7              RPushbullet_0.3.4          </span></span>
<span><span class="co">#&gt; [113] survival_3.3-1              RCurl_1.98-1.6             </span></span>
<span><span class="co">#&gt; [115] future.apply_1.8.1          tibble_3.2.1               </span></span>
<span><span class="co">#&gt; [117] crayon_1.5.0                KernSmooth_2.23-20         </span></span>
<span><span class="co">#&gt; [119] utf8_1.2.2                  rmarkdown_2.20             </span></span>
<span><span class="co">#&gt; [121] locfit_1.5-9.4              grid_4.0.2                 </span></span>
<span><span class="co">#&gt; [123] data.table_1.14.2           blob_1.2.2                 </span></span>
<span><span class="co">#&gt; [125] SimDesign_2.10.1            digest_0.6.29              </span></span>
<span><span class="co">#&gt; [127] xtable_1.8-4                tidyr_1.3.0                </span></span>
<span><span class="co">#&gt; [129] textshaping_0.3.6           stats4_4.0.2               </span></span>
<span><span class="co">#&gt; [131] munsell_0.5.0               bslib_0.3.1                </span></span>
<span><span class="co">#&gt; [133] sessioninfo_1.2.2</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
