<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="mobileRNA">
<title>Overview of the mobileRNA R package for mobile sRNA identification between genotypes • mobileRNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the mobileRNA R package for mobile sRNA identification between genotypes">
<meta property="og:description" content="mobileRNA">
<meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/mobileRNA.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mobileRNA_local.html">Analyse local sRNA populations using mobileRNA</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="mobileRNA_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Overview of the mobileRNA R package for mobile sRNA identification between genotypes</h1>
                        <h4 data-toc-skip class="author">Katie Jeynes-Cupper, Marco Catoni</h4>
            
            <h4 data-toc-skip class="date">2023-06-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KJeynesCupper/mobileRNA.git/blob/HEAD/vignettes/mobileRNA.Rmd" class="external-link"><code>vignettes/mobileRNA.Rmd</code></a></small>
      <div class="d-none name"><code>mobileRNA.Rmd</code></div>
    </div>

    
    
<p><br></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In plants, systemic signalling is an elaborated molecular system which allows to coordinate plant development at the entire organism level, integrating and transmitting the information perceived from environment to distant organs. An important role in long-distance signalling is played by small RNA molecules (sRNAs). These sRNAs can be produced in a tissue and then being transported systemically across the vascular system into recipient organs, where they can induce a molecular response and coordinate physiological changes. Being mobile sRNA able to move across a graft junction, their study can bring to novel methods for transferring traits into crops, for example by combining genotypes by grafting, rather than using breeding or transgenic approaches. Several genomics studies used grafting to explore and characterise mobile sRNAs, however, the approach used to distinguish these mobile molecules from the sequencing noise is not standardised, and for each study it is necessary to adapt general RNA sequencing pipelines to the specific study of mobile molecules.<br>
Here we introduce the R Package , to aid the identification of mobile RNA molecules in a chimeric system, enable the visualization of the results, and supports functional analysis. The package generates mobile molecules through a consensus classification system, in hand with the optional use of differential expression analysis and can quickly achieve results. <br></p>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>This manual offers a pipeline for the analysis of sRNAseq data from plant grafting experiments, specifically to identify the movement of sRNA molecules across a graft junction from one genotype to another.</p>
<p>The pre-processing method is different to others, as it involves mapping the samples to a merged reference genome, comprising of the two genotypes in the heterograft system. In addition, all samples are analysed against a list of all identified sRNA cluster loci to ensure that all sRNA clusters are fairly recognised in the analysis</p>
<p>The analysis method takes the results produced by ShortStack mapping and cluster analysis and identified potential mobile sRNA by removing those mapped to the destination tissue genotype and removing additional falsely mapped sRNA. Previous methods have not employed statistical analysis to select mobile sRNA, here we have incorporated this into the analysis options. This adds greater flexibility and control to the user. By the end of the analysis using mobileRNA, the user could have produced a number of figures to represent the data. This includes plots to represent the:</p>
<ul>
<li><p>Distribution of sRNA sizes across samples, total dataset consensus sRNA,or across the potential mobile sRNA.</p></li>
<li><p>Distribution of sRNA across genomic features</p></li>
<li><p>Sample distance, including a PCA and heatmap</p></li>
<li><p>Heatmap of mobile sRNA expression <br></p></li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The latest version of <code>mobileRNA</code> can be installed via GitHub using the <code>devtools</code> package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"KJeynesCupper/mobileRNA"</span>, ref <span class="op">=</span> <span class="st">"main"</span><span class="op">)</span></span></code></pre></div>
<p>Load into Library:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KJeynesCupper/mobileRNA.git" class="external-link">mobileRNA</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sRNA_data"</span><span class="op">)</span></span></code></pre></div>
<p>Once installed, you will have access to the following functions: <br></p>
<p><strong>Pre-processing</strong></p>
<ul>
<li>1: <code><a href="../reference/RNAmergeGenomes.html">RNAmergeGenomes()</a></code>
</li>
<li>2: <code><a href="../reference/RNAmergeAnnotations.html">RNAmergeAnnotations()</a></code>
</li>
<li>3: <code><a href="../reference/RNAloci.html">RNAloci()</a></code>
</li>
</ul>
<p><strong>Analysis</strong></p>
<ul>
<li>1: <code><a href="../reference/RNAimport.html">RNAimport()</a></code>
</li>
<li>2: <code><a href="../reference/RNAconsensus.html">RNAconsensus()</a></code>
</li>
<li>3: <code><a href="../reference/RNAsubset.html">RNAsubset()</a></code>
</li>
<li>4: <code><a href="../reference/RNAdistribution.html">RNAdistribution()</a></code>
</li>
<li>5: <code><a href="../reference/plotSamplePCA.html">plotSamplePCA()</a></code>
</li>
<li>6: <code>plotSampleDistribution()</code>
</li>
<li>7: <code><a href="../reference/RNAanalysis.html">RNAanalysis()</a></code>
</li>
<li>8: <code><a href="../reference/RNAmobile.html">RNAmobile()</a></code>
</li>
<li>9: <code><a href="../reference/plotHeatmap.html">plotHeatmap()</a></code>
</li>
<li>10: <code><a href="../reference/RNAmean.html">RNAmean()</a></code>
</li>
<li>11: <code><a href="../reference/RNAfeatures.html">RNAfeatures()</a></code>
</li>
<li>12: <code><a href="../reference/RNAattributes.html">RNAattributes()</a></code>
</li>
<li>13: <code><a href="../reference/RNAsequences.html">RNAsequences()</a></code>
</li>
</ul>
<p>Further details on how each function operates will be provided later on in the documentation. For some functions there are recurring arguments. In the pre-processing steps, the <code>files</code> argument occurs frequently and tells the function a local directory path to a folder holding a particular file(s). Similarly, the argument <code>out</code> instructs the function of the location to save the output file(s). In the analysis steps, some important arguments include <code>samples</code> which refers to the a character vector contains the names of all samples in the experiment, <code>conditions</code> which refers to a character vector containing a select few samples depending on your analysis preference, <code>groups</code> which refers to a character vector of the two experimental conditions and lastly <code>controls</code>which refers to the control samples in the analysis. A recurring argument in the analysis pipeline is the <code>method</code> argument which tells the function how to process the data from the choice of two analytical methods (i.e.<code>"DESeq"</code>, or <code>"edgeR"</code>). All other arguments are discussed in further detail by looking at the a functions help file.</p>
<p><br> The workflow is shown below. It begin in R-Studio to merge the two genome assemblies into one, then the pre-processing moves into Linux to align each replicate to the merged reference and then back into R-Studio to undertake the analysis to identify potentially mobile RNA species.</p>
<p><img src="../reference/figures/program_flow.png" width="400" height="350"></p>
<p>Going forward, we assume standard quality control steps on raw samples has been completed (i.e. trimming of adapters and low quality reads) <br></p>
<div class="section level3">
<h3 id="an-overview-of-the-data-used">An overview of the data used<a class="anchor" aria-label="anchor" href="#an-overview-of-the-data-used"></a>
</h3>
<p>For the following examples, a semi-synthetic small RNA-seq data set has been utilised to simulate the movement of sRNA molecules from an Tomato (<strong>Solanium lycopersicum</strong>) rootstock to a Eggplant (<strong>Solanium melongena</strong>) scion, a grafting system known to be compatible. The three heterograft replicates, known as <code>heterograft_1</code>, <code>heterograft_2</code> and <code>heterograft_3</code> are individual tomato replicates spiked with 150 tomato sRNA clusters. While the selfgraft replicates, known as <code>selfgraft_1</code>, <code>selfgraft_2</code> and <code>selfgraft_3</code>, are the individual tomato replicates without the spiked tomato sRNA clusters. Moreover, the replicates mirror each other where, for instance, <code>heterograft_1</code> and <code>selfgraft_1</code> are the same replicate, either with or without the spiked clusters. These replicates serve for the aim to analyse sRNA mobility.</p>
<p>The data set, called <code>sRNA_data</code>, stores a matrix containing the pre-processed data from the experiment. As a user, this allows you to see what a full data set might look like and how you might expect it to change across the analysis.</p>
<p>These can be loaded in these R workspace by using the following command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sRNA_data"</span><span class="op">)</span></span></code></pre></div>
<p>In addition, there are two further datasets called “sRNA_data_consensus” and “sRNA_data_mobile” which are utilised in the function documentations. Both build on the “sRNA_data” dataset where “sRNA_data_consensus” contains the RNA consensus results formed by the function [RNAconsensus()]. While, the “sRNA_data_mobile” contains only the likely mobile sRNAs selected by the function [RNAmobile()] under the default parameters. <br></p>
</div>
<div class="section level3">
<h3 id="data-organisation">Data organisation<a class="anchor" aria-label="anchor" href="#data-organisation"></a>
</h3>
<p>There are Two key elements required for the pipeline analysis: * sRNA-seq sample replicates (.fastq/.fq) * Two reference genomes (.fasta/.fa)</p>
<p>And one additional element; to improve functional analysis: * Two reference genome annotations (.gff/.gff3) <br><br></p>
<div class="section level4">
<h4 id="naming-data-files">Naming data files<a class="anchor" aria-label="anchor" href="#naming-data-files"></a>
</h4>
<p>It is recommended to rename your files to names you wish for them to be represented as within the analysis and shown as labels in plots. Plus, it makes the analysis easier!</p>
<p>For example, instead of names such as:</p>
<ul>
<li>1: <code>sample_1.fq</code>
</li>
<li>2: <code>sample_2.fq</code>
</li>
<li>3: <code>sample_3.fq</code>
</li>
<li>4: <code>sample_4.fq</code>
</li>
<li>5: <code>sample_5.fq</code>
</li>
<li>6: <code>sample_6.fq</code>
</li>
</ul>
<p>For the example data set included in the package, here we have renamed the files based on the condition (treatment or control).For the hetero-grafts, where the is a eggplant scion and an tomato rootstock:</p>
<ul>
<li>1: <code>heterograft_1.fq</code>
</li>
<li>2: <code>heterograft_2.fq</code>
</li>
<li>3: <code>heterograft_3.fq</code>
</li>
</ul>
<p>and for the eggplant self-grafts:</p>
<ul>
<li>4: <code>selfgraft_1.fq</code>
</li>
<li>5: <code>selfgraft_2.fq</code>
</li>
<li>6: <code>selfgraft_3.fq</code> <br>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="the-reference-genomes-and-annotation-files--fasta-fa">The reference genomes and annotation files (.fasta/.fa)<a class="anchor" aria-label="anchor" href="#the-reference-genomes-and-annotation-files--fasta-fa"></a>
</h4>
<p>The pipeline is designed to analyse grafting systems with two distinct genomes, here tomato and eggplant. <br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="pre-processing">Pre-Processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h2>
<p>The pre-processing step involves cleaning raw data and aligning data to the merged genome. We recommend installing and using the <code>ShortStack</code> (<a href="https://github.com/MikeAxtell/ShortStack" class="external-link uri">https://github.com/MikeAxtell/ShortStack</a>) program to aligns samples to the merged genome and undertake cluster analysis. This program is specifically formulated for sRNAseq analysis, utilising Bowtie (Version 1) to map samples and a specifically formulated algorithm to cluster sRNAs.</p>
<p>Here, we introduce an alternative mapping method for the analysis of plant heterograft samples. The heterograft system involves two genotypes; here the two genome references are merged into a single reference to which samples are aligned to.</p>
<p><code>mobileRNA</code> offers a function to merge two FASTA reference genomes into one. To distinguish between the reference genomes in a merged file, it is important to make sure the chromosome names between the genomes are different and distinguishable. The function below added a particular character string to the start of each chromosome name in each reference genome. As standard, the string <code>A_</code> is added to the reference genome supplied to <code>genomeA</code> and <code>B_</code> is added to the reference genome supplied to <code>genomeB.</code> These can be customised to the users preference, see manual for more information.</p>
<p>Going forward, the pipeline has assumed that the sRNA-seq samples have met quality control standards. <br></p>
<div class="section level3">
<h3 id="pre-mapping">Pre-mapping<a class="anchor" aria-label="anchor" href="#pre-mapping"></a>
</h3>
<div class="section level4">
<h4 id="merging-genome-assemblies">Merging Genome Assemblies<a class="anchor" aria-label="anchor" href="#merging-genome-assemblies"></a>
</h4>
<p>Here we merge the two reference genomes into a single merged genome using the [mobileRNA::RNAgenomeMerge()] function. As default, this function changes the chromosome names of each genome to ensure they are distinguishable. To do so, the function requires an input of the initials of each organism’s Latin name. But why do this:</p>
<ul>
<li><p>If the two genomes use the same pattern to name the chromosomes, the user will not be able to differentiate the chromosomes from one another in the merged genome. This could be solves by adding letters to the chromosomes of one of the genomes, for example, <code>SM</code> to represent the Latin name of eggplant.</p></li>
<li><p>If a chromosome naming pattern contains punctuation, the mapping step will not work.</p></li>
</ul>
<p>In the example, the <em>Solanum lycopersicum</em> (version 4) genome contains a full-stop/period within each chromosome name which needs to be removed as well. Here we rename the chromosomes of the tomato genome to <code>SL</code> and the chromosomes related to the eggplant genome (<em>Solanum melongena</em>) to <code>SM</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RNAmergeGenomes.html">RNAmergeGenomes</a></span><span class="op">(</span>genomeA <span class="op">=</span><span class="st">"/Users/user1/projectname/workplace/reference/ref1.fa"</span>,</span>
<span>               genomeB <span class="op">=</span><span class="st">"/Users/user1/projectname/workplace/reference/ref2.fa"</span>,</span>
<span>               abbreviationGenomeA <span class="op">=</span> <span class="st">"SL"</span>,</span>
<span>               abbreviationGenomeB <span class="op">=</span> <span class="st">"SM"</span>,</span>
<span>               out_dir <span class="op">=</span> <span class="st">"/Users/user1/projectname/workplace/reference/merge/</span></span>
<span><span class="st">               ref_merged.fa"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="merging-genome-annotations">Merging Genome Annotations<a class="anchor" aria-label="anchor" href="#merging-genome-annotations"></a>
</h3>
<p>To identify associates between sRNA clusters and genomic features, the genomic annotation files (GFF) of each reference genome must be merged in the same way the reference genomes were. This includes following the same naming patterns for the chromosomes in each genome as undertaken earlier with the <code><a href="../reference/RNAmergeGenomes.html">RNAmergeGenomes()</a></code> function. To undertake a similar merging between the two GFF files, use the <code>RNAmergeAnnotations</code> function.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co">## merge the annotation files into a single annotation file</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="kw">RNAmergeAnnotation</span>(<span class="dt">annotationA =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="st">                   annotation/annotation_1.gff3"</span>,</span>
<span id="cb5-4"><a href="#cb5-4"></a>                   <span class="dt">annotationB =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="st">                   annotation/annotation_1.gff3"</span>,</span>
<span id="cb5-6"><a href="#cb5-6"></a>                   <span class="dt">abbreviationAnnoA =</span> <span class="st">"SL"</span>,</span>
<span id="cb5-7"><a href="#cb5-7"></a>                   <span class="dt">abbreviationAnnoB =</span> <span class="st">"SM"</span></span>
<span id="cb5-8"><a href="#cb5-8"></a>                  <span class="dt">out_dir =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="st">                  annotation/merge/anno_merged.gff"</span>)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="auto-detection-of-srna-cluster">Auto-Detection of sRNA Cluster<a class="anchor" aria-label="anchor" href="#auto-detection-of-srna-cluster"></a>
</h3>
<p>Here we identify and build a list sRNA clusters within each sample to assist the mapping step later on to ensure consistency across the analysis.</p>
<p>We recommend installing the <code>ShortStack</code> (<a href="https://github.com/MikeAxtell/ShortStack" class="external-link uri">https://github.com/MikeAxtell/ShortStack</a>) program to detect sRNA clusters and align sRNAseq samples. <br></p>
<div class="section level5">
<h5 id="step-1---cluster-analysis-with-shortstack">Step 1 - Cluster analysis with ShortStack<a class="anchor" aria-label="anchor" href="#step-1---cluster-analysis-with-shortstack"></a>
</h5>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb6-2"><a href="#cb6-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb6-3"><a href="#cb6-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb6-4"><a href="#cb6-4"></a>--bowtie_cores 6 \</span>
<span id="cb6-5"><a href="#cb6-5"></a>--mmap n \</span>
<span id="cb6-6"><a href="#cb6-6"></a>--mismatches 0 \</span>
<span id="cb6-7"><a href="#cb6-7"></a>--nohp \</span>
<span id="cb6-8"><a href="#cb6-8"></a>--outdir <span class="op">&lt;</span>./output/directory<span class="op">&gt;</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="step-2---build-srna-cluster-list">Step 2 - Build sRNA cluster list<a class="anchor" aria-label="anchor" href="#step-2---build-srna-cluster-list"></a>
</h5>
<p>Now, we collate all the sRNA loci information from each sample into a text file.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># location of step 1 output</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>folder &lt;-<span class="st"> </span><span class="er">&lt;</span>.<span class="op">/</span>output<span class="op">/</span>directory<span class="op">/</span>from<span class="op">/</span>step<span class="op">/</span><span class="dv">1</span><span class="op">/</span><span class="er">&gt;</span></span>
<span id="cb7-3"><a href="#cb7-3"></a></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co"># name and location to save output file to (must be .txt)</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>save_folder &lt;-<span class="st"> </span><span class="er">&lt;</span>.<span class="op">/</span>output<span class="op">/</span>directory<span class="op">/</span>ClustersInfo.txt<span class="op">&gt;</span></span>
<span id="cb7-6"><a href="#cb7-6"></a></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co"># names of samples (ie. folder names)</span></span>
<span id="cb7-8"><a href="#cb7-8"></a>sample_names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"&lt;treatment_1&gt;"</span>, <span class="st">"&lt;treatment_2&gt;"</span>, <span class="st">"&lt;control_1&gt;"</span>,<span class="st">"&lt;control_2&gt;"</span>)</span>
<span id="cb7-9"><a href="#cb7-9"></a></span>
<span id="cb7-10"><a href="#cb7-10"></a></span>
<span id="cb7-11"><a href="#cb7-11"></a>loci_info &lt;-<span class="st"> </span><span class="kw">RNAloci</span>(<span class="dt">files =</span> folder, </span>
<span id="cb7-12"><a href="#cb7-12"></a>             <span class="dt">out =</span> save_folder,</span>
<span id="cb7-13"><a href="#cb7-13"></a>             <span class="dt">samples =</span> sample_names)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="mapping">Mapping<a class="anchor" aria-label="anchor" href="#mapping"></a>
</h3>
<p>Each sample is mapped to the merged reference genome with the list of sRNA clusters.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb8-2"><a href="#cb8-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb8-3"><a href="#cb8-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb8-4"><a href="#cb8-4"></a>--locifile <span class="op">&lt;</span>./output/directory/ClustersInfo.gff3<span class="op">&gt;</span> \</span>
<span id="cb8-5"><a href="#cb8-5"></a>--bowtie_cores 6 \</span>
<span id="cb8-6"><a href="#cb8-6"></a>--mmap n \</span>
<span id="cb8-7"><a href="#cb8-7"></a>--mismatches 0 \</span>
<span id="cb8-8"><a href="#cb8-8"></a>--nohp \</span>
<span id="cb8-9"><a href="#cb8-9"></a>--mincov 5 \</span>
<span id="cb8-10"><a href="#cb8-10"></a>--outdir <span class="op">&lt;</span>./output/directory/step2/<span class="op">&gt;</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>Here, the analysis of the pre-processed data occurs, with the aim to identify if any mobile sRNA molecules are traveling from one genotype to the other in your grafting system. <br></p>
<div class="section level3">
<h3 id="import-data">Import Data<a class="anchor" aria-label="anchor" href="#import-data"></a>
</h3>
<p>In the pre-processing steps, the data was cleaned and aligned to the merged reference genome. During the mapping step, a folder for each sample is created which stores all the results in. The analysis steps requires the information which is stored in the <code>Results.txt</code>.</p>
<p>The <code><a href="../reference/RNAimport.html">RNAimport()</a></code> function imports the data from the folder containing all sample folders. The function extracts the required information, stores it in a matrix.</p>
<p>The function requires some information to coordinate the importation and calculations. It requires a directory path to your processed samples; this is the path to the folder containing all the individual sample folders which is stores in argument <code>directory</code>. The second directory path is to the loci file containing information on the loci of the sRNA clusters across the analysis, stored in the argument <code>loci</code>. The last pieces of information the function requires a vector containing the sample names, both treatment and control replicates which is stored in the argument <code>samples</code>. The sample names must match the names of the folders produced in the mapping, and stored in the directory supplied to the argument <code>directory</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Import &amp; organise data.</span></span>
<span><span class="va">results_dir</span> <span class="op">&lt;-</span>  <span class="st">"./analysis/alignment_unique_two/"</span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>,<span class="st">"heterograft_2"</span>, <span class="st">"heterograft_3"</span>,</span>
<span>                  <span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span>, <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">sRNA_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAimport.html">RNAimport</a></span><span class="op">(</span>input <span class="op">=</span> <span class="st">"sRNA"</span>,</span>
<span>                       directory <span class="op">=</span> <span class="va">results_dir</span>,</span>
<span>                           samples <span class="op">=</span> <span class="va">sample_names</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># example dataframe </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sRNA_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   Locus chr     start       end   Cluster</span></span>
<span><span class="co">#&gt; 1 1:100000956-100001076   1 100000956 100001076 cluster_1</span></span>
<span><span class="co">#&gt; 2 1:100001486-100001554   1 100001486 100001554 cluster_2</span></span>
<span><span class="co">#&gt; 3 1:100001646-100001731   1 100001646 100001731 cluster_3</span></span>
<span><span class="co">#&gt; 4   1:10001034-10001073   1  10001034  10001073 cluster_4</span></span>
<span><span class="co">#&gt; 5 1:100011795-100011930   1 100011795 100011930 cluster_5</span></span>
<span><span class="co">#&gt; 6   1:10002000-10002065   1  10002000  10002065 cluster_6</span></span>
<span><span class="co">#&gt;   DicerCall_heterograft_1 Count_heterograft_1 RPM_heterograft_1</span></span>
<span><span class="co">#&gt; 1                      24                  19             1.908</span></span>
<span><span class="co">#&gt; 2                      24                   3             0.301</span></span>
<span><span class="co">#&gt; 3                       N                   3             0.301</span></span>
<span><span class="co">#&gt; 4                      24                   1             0.100</span></span>
<span><span class="co">#&gt; 5                      24                  11             1.105</span></span>
<span><span class="co">#&gt; 6                      24                   9             0.904</span></span>
<span><span class="co">#&gt;      MajorRNA_heterograft_1 DicerCall_heterograft_2 Count_heterograft_2</span></span>
<span><span class="co">#&gt; 1  AGUGACGGUCUCAUGUAACAUAGA                      24                  14</span></span>
<span><span class="co">#&gt; 2  AAAGAGUAGACAUAUAUUAAAUGG                      20                   5</span></span>
<span><span class="co">#&gt; 3 CUGUCUCAUGUGACACUCUUUCACU                      23                   6</span></span>
<span><span class="co">#&gt; 4  AAAUUGUCAUUUAGUCAGAGGAAU                       N                   0</span></span>
<span><span class="co">#&gt; 5  AUAGACAAAUCGCACUCGCUCUCG                      24                   2</span></span>
<span><span class="co">#&gt; 6  UUAGAUUUCUGGAGUAACAACUUU                       N                   2</span></span>
<span><span class="co">#&gt;   RPM_heterograft_2   MajorRNA_heterograft_2 DicerCall_heterograft_3</span></span>
<span><span class="co">#&gt; 1             1.803 AGUGACGGUCUCAUGUAACAUAGA                      24</span></span>
<span><span class="co">#&gt; 2             0.644     ACAUAUUUUUUUGGACGAAC                      24</span></span>
<span><span class="co">#&gt; 3             0.773  UCUGUCUCAUGUGACACUCUUUC                      24</span></span>
<span><span class="co">#&gt; 4             0.000                        N                      24</span></span>
<span><span class="co">#&gt; 5             0.258 AAGUGUCUACCCAUUGUGAUUCGC                      24</span></span>
<span><span class="co">#&gt; 6             0.258      UUAUUGACAAACUAUAACU                      22</span></span>
<span><span class="co">#&gt;   Count_heterograft_3 RPM_heterograft_3   MajorRNA_heterograft_3</span></span>
<span><span class="co">#&gt; 1                  30             2.953 AGUGACGGUCUCAUGUAACAUAGA</span></span>
<span><span class="co">#&gt; 2                   1             0.098 UAUGUCUACUCUUUAUUACUCUCC</span></span>
<span><span class="co">#&gt; 3                  11             1.083 AUCUUGUCAAGUGAAAGAGUGUCA</span></span>
<span><span class="co">#&gt; 4                  11             1.083 AUUCCUCUGACUAAAUGACAAUUU</span></span>
<span><span class="co">#&gt; 5                   5             0.492 AAUCACAAUGGGUAGACACUUAGA</span></span>
<span><span class="co">#&gt; 6                   3             0.295 UUAGAUUUCUGGAGUAACAACUUU</span></span>
<span><span class="co">#&gt;   DicerCall_selfgraft_1 Count_selfgraft_1 RPM_selfgraft_1</span></span>
<span><span class="co">#&gt; 1                    24                19           1.909</span></span>
<span><span class="co">#&gt; 2                    24                 3           0.301</span></span>
<span><span class="co">#&gt; 3                     N                 3           0.301</span></span>
<span><span class="co">#&gt; 4                    24                 1           0.100</span></span>
<span><span class="co">#&gt; 5                    24                11           1.105</span></span>
<span><span class="co">#&gt; 6                    24                 9           0.904</span></span>
<span><span class="co">#&gt;        MajorRNA_selfgraft_1 DicerCall_selfgraft_2 Count_selfgraft_2</span></span>
<span><span class="co">#&gt; 1  AGUGACGGUCUCAUGUAACAUAGA                    24                14</span></span>
<span><span class="co">#&gt; 2  AAAGAGUAGACAUAUAUUAAAUGG                    20                 5</span></span>
<span><span class="co">#&gt; 3 CUGUCUCAUGUGACACUCUUUCACU                    23                 6</span></span>
<span><span class="co">#&gt; 4  AAAUUGUCAUUUAGUCAGAGGAAU                     N                 0</span></span>
<span><span class="co">#&gt; 5  AAUGGGUAGACACUUAGAUACAUG                    24                 2</span></span>
<span><span class="co">#&gt; 6  UUAGAUUUCUGGAGUAACAACUUU                     N                 2</span></span>
<span><span class="co">#&gt;   RPM_selfgraft_2     MajorRNA_selfgraft_2 DicerCall_selfgraft_3</span></span>
<span><span class="co">#&gt; 1           1.804 AGUGACGGUCUCAUGUAACAUAGA                    24</span></span>
<span><span class="co">#&gt; 2           0.644     ACAUAUUUUUUUGGACGAAC                    24</span></span>
<span><span class="co">#&gt; 3           0.773  UCUGUCUCAUGUGACACUCUUUC                    24</span></span>
<span><span class="co">#&gt; 4           0.000                        N                    24</span></span>
<span><span class="co">#&gt; 5           0.258 AUAGACAAAUCGCACUCGCUCUCG                    24</span></span>
<span><span class="co">#&gt; 6           0.258  AGAUUUCUGGAGUAACAACUUUG                    22</span></span>
<span><span class="co">#&gt;   Count_selfgraft_3 RPM_selfgraft_3     MajorRNA_selfgraft_3</span></span>
<span><span class="co">#&gt; 1                30           2.954 AGUGACGGUCUCAUGUAACAUAGA</span></span>
<span><span class="co">#&gt; 2                 1           0.098 UAUGUCUACUCUUUAUUACUCUCC</span></span>
<span><span class="co">#&gt; 3                11           1.083 AUCUUGUCAAGUGAAAGAGUGUCA</span></span>
<span><span class="co">#&gt; 4                11           1.083 AUUCCUCUGACUAAAUGACAAUUU</span></span>
<span><span class="co">#&gt; 5                 5           0.492 AAUCACAAUGGGUAGACACUUAGA</span></span>
<span><span class="co">#&gt; 6                 3           0.295   UUGACAAACUAUAACUUAAAUU</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="sample-level-quality-control">Sample-level quality control<a class="anchor" aria-label="anchor" href="#sample-level-quality-control"></a>
</h3>
<p>A handy step in the analysis is to assess the overall similarity between sample replicates to understand which samples are the most similar and which are different, and where the most variation is introduced in the data set. As well as understanding whether the data set meets your expectations. It is expected that between the conditions, the sample replicates show enough variation to suggest that the replicates are from different groups.</p>
<p>To investigate the sample similarity/difference, we will undertake sample-level quality control using three different methods:</p>
<ul>
<li>Distribution of RNA classes within each sample</li>
<li>Principle component analysis (PCA)</li>
<li>hierarchical clustering Heatmap</li>
</ul>
<p>This will show us how well samples within each condition cluster together, which may highlight outliers. Plus, to show whether our experimental conditions represent the main source of variation in the data set.</p>
<p>Here we will be employing an unsupervised clustering methods for the PCA and hierarchical clustering Heatmap. This involves an unbiased log2-transformation of the counts which will emphasis the the sample clustering to improve visualization. The DESeq2 package contains a particularly useful function to undertake regularized log transformation (rlog) which controls the variance across the mean, and in this package we have utilized this for the quality control steps.</p>
<p><br></p>
<div class="section level4">
<h4 id="plot-the-distribution-of-rna-classes-within-each-sample">Plot the distribution of RNA classes within each sample<a class="anchor" aria-label="anchor" href="#plot-the-distribution-of-rna-classes-within-each-sample"></a>
</h4>
<p>Explore the number of each RNA class identified within each sample using the <code>RNAdistribution</code> function which produces a table and plots the results. The results can be plotted as either a bar chart (<code>style = "bar"</code>) or a line graph (<code>style = "line"</code>) for each sample and can be shown in a bar chart facet (<code>facet = TRUE</code>) or in as a single line graph, where each line represents a sample (<code>together=TRUE</code>). Alternatively, the results can be plotted individually for each sample in either format.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">sample_distribution_line</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span><span class="va">sRNA_data</span>,</span>
<span>                                            style <span class="op">=</span> <span class="st">"line"</span>,</span>
<span>                                            together <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                          facet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">sample_distribution_bar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span><span class="va">sRNA_data</span>,</span>
<span>                                           style <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>                                           facet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Class heterograft_1 heterograft_2 heterograft_3 selfgraft_1 selfgraft_2</span></span>
<span><span class="co">#&gt; 1:    20          1720          2102          1306        1719        2101</span></span>
<span><span class="co">#&gt; 2:    21          4940          5655          4278        4938        5653</span></span>
<span><span class="co">#&gt; 3:    22          3655          4883          3422        3655        4883</span></span>
<span><span class="co">#&gt; 4:    23          6657          6909          7469        6650        6902</span></span>
<span><span class="co">#&gt; 5:    24         99330         85514        101771       99191       85374</span></span>
<span><span class="co">#&gt;    selfgraft_3</span></span>
<span><span class="co">#&gt; 1:        1305</span></span>
<span><span class="co">#&gt; 2:        4276</span></span>
<span><span class="co">#&gt; 3:        3422</span></span>
<span><span class="co">#&gt; 4:        7462</span></span>
<span><span class="co">#&gt; 5:      101630</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-8-1.png" alt="An example line graph (A) and bar graph (B), to show the distribution of
RNA classes within each sample" width="700"><img src="mobileRNA_files/figure-html/unnamed-chunk-8-2.png" alt="An example line graph (A) and bar graph (B), to show the distribution of
RNA classes within each sample" width="700"><p class="caption">
An example line graph (A) and bar graph (B), to show the distribution of RNA classes within each sample
</p>
</div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="principle-component-analysis-to-assess-sample-distance">Principle component analysis to assess sample distance<a class="anchor" aria-label="anchor" href="#principle-component-analysis-to-assess-sample-distance"></a>
</h4>
<p>Principal Component Analysis (PCA) is a useful technique to illustrate sample distance as it emphasizes the variation through the reduction of dimensions in the data set. Here, we introduce the function <code><a href="../reference/plotSamplePCA.html">plotSamplePCA()</a></code></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSamplePCA.html">plotSamplePCA</a></span><span class="op">(</span><span class="va">sRNA_data</span>, <span class="va">group</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-10-1.png" alt="An example of a PCA, illustracting the sRNA data set sample similarity" width="50%"><p class="caption">
An example of a PCA, illustracting the sRNA data set sample similarity
</p>
</div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="hierarchical-clustered-heatmap-to-assess-sample-distance">Hierarchical clustered heatmap to assess sample distance<a class="anchor" aria-label="anchor" href="#hierarchical-clustered-heatmap-to-assess-sample-distance"></a>
</h4>
<p>Similarly, to a PCA plot, the <code><a href="../reference/plotSampleDistance.html">plotSampleDistance()</a></code> function undertakes hierarchical clustering with an unbiased log transformation to calculate sample distance and is plotted in the form of a heatmap.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSampleDistance.html">plotSampleDistance</a></span><span class="op">(</span><span class="va">sRNA_data</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-12-1.png" alt="An example of a heatmap,illustracting the sRNA data set sample similarity" width="50%"><p class="caption">
An example of a heatmap,illustracting the sRNA data set sample similarity
</p>
</div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="define-the-consensus-rna-class">Define the consensus RNA class<a class="anchor" aria-label="anchor" href="#define-the-consensus-rna-class"></a>
</h3>
<p>Have a look at the <code>sRNA_data</code> object, you will see that for each sample the class of sRNA has been determined (see columns with names starting with DicerCall). The result will either state a number from 20-24 which represents the nucleotide length of the sRNA, or, state NA/N which represents that the sRNA was unclassified in the sample. There is usually some discrepancies between the classification of a sRNA, therefore, a consensus class can be assigned based off the information from the treatment samples. The <code><a href="../reference/RNAconsensus.html">RNAconsensus()</a></code> function is used to compute this by counting the number of occurrences of classifications for a sRNA. The consensus is determined based on the class with the most counts.</p>
<p>The function requires the user to state which samples to base the consensus calculation on (see argument <code>conditions</code>). It is recommend to state the treatment samples only, because these are the ones of interest. However, if you have less than two treatment replicate, calculating the consensus class this way could dramatically shrink your data - although this could be a benefit as it’ll make the analysis more stringent. Therefore, in this particular circumstance, it is suggested that the user tries defining the consensus class with all samples in parallel to defining the consensus with the treatment samples.</p>
<p>The function also works to remove noise from your data. If the user wishes to keep the noise, set <code>tidy=FALSE</code> to prevent the removal of unclassified clusters.</p>
<p>For the downstream analysis, it can be useful to define the distinct 24-nt and 21/22-nt populations of sRNA molecules at the beginning of the analysis. Hence, the data should be subset. To subset the data, use the <code><a href="../reference/RNAsubset.html">RNAsubset()</a></code> function to choose which sRNA populations to include: 20-nt, 21-nt, 22-nt, 23-nt, or 24-nt. It is recommended to subset the data into two dataframes: one which includes only 24-nt sRNA and the other to include both 21-nt and 22-nt sRNAs. To instruct the <code><a href="../reference/RNAsubset.html">RNAsubset()</a></code> function to select specific groups, use the <code>type</code> argument.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##  define consensus sRNA classes based of treatment/hetero-graft samples</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>,<span class="st">"heterograft_2"</span>, <span class="st">"heterograft_3"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define consensus, store as a data summary file.</span></span>
<span><span class="va">sRNA_data_consensus</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAconsensus.html">RNAconsensus</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data</span>,</span>
<span>                                  conditions <span class="op">=</span> <span class="va">samples</span>,</span>
<span>                                  tidy<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sRNA_data_consensus</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   Locus chr     start       end   Cluster</span></span>
<span><span class="co">#&gt; 1 1:100000956-100001076   1 100000956 100001076 cluster_1</span></span>
<span><span class="co">#&gt; 2 1:100001486-100001554   1 100001486 100001554 cluster_2</span></span>
<span><span class="co">#&gt; 3 1:100001646-100001731   1 100001646 100001731 cluster_3</span></span>
<span><span class="co">#&gt; 4   1:10001034-10001073   1  10001034  10001073 cluster_4</span></span>
<span><span class="co">#&gt; 5 1:100011795-100011930   1 100011795 100011930 cluster_5</span></span>
<span><span class="co">#&gt; 6   1:10002000-10002065   1  10002000  10002065 cluster_6</span></span>
<span><span class="co">#&gt;   DicerCall_heterograft_1 Count_heterograft_1 RPM_heterograft_1</span></span>
<span><span class="co">#&gt; 1                      24                  19             1.908</span></span>
<span><span class="co">#&gt; 2                      24                   3             0.301</span></span>
<span><span class="co">#&gt; 3                       N                   3             0.301</span></span>
<span><span class="co">#&gt; 4                      24                   1             0.100</span></span>
<span><span class="co">#&gt; 5                      24                  11             1.105</span></span>
<span><span class="co">#&gt; 6                      24                   9             0.904</span></span>
<span><span class="co">#&gt;      MajorRNA_heterograft_1 DicerCall_heterograft_2 Count_heterograft_2</span></span>
<span><span class="co">#&gt; 1  AGUGACGGUCUCAUGUAACAUAGA                      24                  14</span></span>
<span><span class="co">#&gt; 2  AAAGAGUAGACAUAUAUUAAAUGG                      20                   5</span></span>
<span><span class="co">#&gt; 3 CUGUCUCAUGUGACACUCUUUCACU                      23                   6</span></span>
<span><span class="co">#&gt; 4  AAAUUGUCAUUUAGUCAGAGGAAU                       N                   0</span></span>
<span><span class="co">#&gt; 5  AUAGACAAAUCGCACUCGCUCUCG                      24                   2</span></span>
<span><span class="co">#&gt; 6  UUAGAUUUCUGGAGUAACAACUUU                       N                   2</span></span>
<span><span class="co">#&gt;   RPM_heterograft_2   MajorRNA_heterograft_2 DicerCall_heterograft_3</span></span>
<span><span class="co">#&gt; 1             1.803 AGUGACGGUCUCAUGUAACAUAGA                      24</span></span>
<span><span class="co">#&gt; 2             0.644     ACAUAUUUUUUUGGACGAAC                      24</span></span>
<span><span class="co">#&gt; 3             0.773  UCUGUCUCAUGUGACACUCUUUC                      24</span></span>
<span><span class="co">#&gt; 4             0.000                        N                      24</span></span>
<span><span class="co">#&gt; 5             0.258 AAGUGUCUACCCAUUGUGAUUCGC                      24</span></span>
<span><span class="co">#&gt; 6             0.258      UUAUUGACAAACUAUAACU                      22</span></span>
<span><span class="co">#&gt;   Count_heterograft_3 RPM_heterograft_3   MajorRNA_heterograft_3</span></span>
<span><span class="co">#&gt; 1                  30             2.953 AGUGACGGUCUCAUGUAACAUAGA</span></span>
<span><span class="co">#&gt; 2                   1             0.098 UAUGUCUACUCUUUAUUACUCUCC</span></span>
<span><span class="co">#&gt; 3                  11             1.083 AUCUUGUCAAGUGAAAGAGUGUCA</span></span>
<span><span class="co">#&gt; 4                  11             1.083 AUUCCUCUGACUAAAUGACAAUUU</span></span>
<span><span class="co">#&gt; 5                   5             0.492 AAUCACAAUGGGUAGACACUUAGA</span></span>
<span><span class="co">#&gt; 6                   3             0.295 UUAGAUUUCUGGAGUAACAACUUU</span></span>
<span><span class="co">#&gt;   DicerCall_selfgraft_1 Count_selfgraft_1 RPM_selfgraft_1</span></span>
<span><span class="co">#&gt; 1                    24                19           1.909</span></span>
<span><span class="co">#&gt; 2                    24                 3           0.301</span></span>
<span><span class="co">#&gt; 3                     N                 3           0.301</span></span>
<span><span class="co">#&gt; 4                    24                 1           0.100</span></span>
<span><span class="co">#&gt; 5                    24                11           1.105</span></span>
<span><span class="co">#&gt; 6                    24                 9           0.904</span></span>
<span><span class="co">#&gt;        MajorRNA_selfgraft_1 DicerCall_selfgraft_2 Count_selfgraft_2</span></span>
<span><span class="co">#&gt; 1  AGUGACGGUCUCAUGUAACAUAGA                    24                14</span></span>
<span><span class="co">#&gt; 2  AAAGAGUAGACAUAUAUUAAAUGG                    20                 5</span></span>
<span><span class="co">#&gt; 3 CUGUCUCAUGUGACACUCUUUCACU                    23                 6</span></span>
<span><span class="co">#&gt; 4  AAAUUGUCAUUUAGUCAGAGGAAU                     N                 0</span></span>
<span><span class="co">#&gt; 5  AAUGGGUAGACACUUAGAUACAUG                    24                 2</span></span>
<span><span class="co">#&gt; 6  UUAGAUUUCUGGAGUAACAACUUU                     N                 2</span></span>
<span><span class="co">#&gt;   RPM_selfgraft_2     MajorRNA_selfgraft_2 DicerCall_selfgraft_3</span></span>
<span><span class="co">#&gt; 1           1.804 AGUGACGGUCUCAUGUAACAUAGA                    24</span></span>
<span><span class="co">#&gt; 2           0.644     ACAUAUUUUUUUGGACGAAC                    24</span></span>
<span><span class="co">#&gt; 3           0.773  UCUGUCUCAUGUGACACUCUUUC                    24</span></span>
<span><span class="co">#&gt; 4           0.000                        N                    24</span></span>
<span><span class="co">#&gt; 5           0.258 AUAGACAAAUCGCACUCGCUCUCG                    24</span></span>
<span><span class="co">#&gt; 6           0.258  AGAUUUCUGGAGUAACAACUUUG                    22</span></span>
<span><span class="co">#&gt;   Count_selfgraft_3 RPM_selfgraft_3     MajorRNA_selfgraft_3 sRNA_Consensus</span></span>
<span><span class="co">#&gt; 1                30           2.954 AGUGACGGUCUCAUGUAACAUAGA          nt_24</span></span>
<span><span class="co">#&gt; 2                 1           0.098 UAUGUCUACUCUUUAUUACUCUCC          nt_24</span></span>
<span><span class="co">#&gt; 3                11           1.083 AUCUUGUCAAGUGAAAGAGUGUCA          nt_23</span></span>
<span><span class="co">#&gt; 4                11           1.083 AUUCCUCUGACUAAAUGACAAUUU          nt_24</span></span>
<span><span class="co">#&gt; 5                 5           0.492 AAUCACAAUGGGUAGACACUUAGA          nt_24</span></span>
<span><span class="co">#&gt; 6                 3           0.295   UUGACAAACUAUAACUUAAAUU          nt_22</span></span>
<span></span>
<span><span class="co"># Subset data for analysis: 24-nt sRNAs</span></span>
<span><span class="va">sRNA_24</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsubset.html">RNAsubset</a></span><span class="op">(</span><span class="va">sRNA_data_consensus</span>, type <span class="op">=</span> <span class="fl">24</span><span class="op">)</span></span>
<span><span class="co"># Subset data for analysis: 24 21/22-nt sRNAs</span></span>
<span><span class="va">sRNA_2122</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsubset.html">RNAsubset</a></span><span class="op">(</span><span class="va">sRNA_data_consensus</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">21</span>, <span class="fl">22</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="plot-consensus-classification-of-srna">Plot consensus classification of sRNA<a class="anchor" aria-label="anchor" href="#plot-consensus-classification-of-srna"></a>
</h4>
<p>Use <code>RNAdistriution()</code> to visualize the distribution of the consensus sRNA classes within the data. This step can be used as a quality control step to illustrate the distribution of each sRNA class within the data. The function can plot the absolute or relative frequency, use argument <code>relative</code>. It can be beneficial to view the absolute frequency as well as the relative to observe whether the data meets your expectations.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>,</span>
<span>                                  consensus <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                  style <span class="op">=</span> <span class="st">"consensus"</span><span class="op">)</span></span>
<span><span class="co"># view </span></span>
<span><span class="va">consensus_plot</span></span>
<span><span class="co">#&gt; $plot</span></span></code></pre></div>
<p><img src="mobileRNA_files/figure-html/unnamed-chunk-14-1.png" width="700" style="display: block; margin: auto;"></p>
<pre><code><span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $data</span></span>
<span><span class="co">#&gt;   sRNA_Consensus      n</span></span>
<span><span class="co">#&gt; 1             20   2545</span></span>
<span><span class="co">#&gt; 2             21   6116</span></span>
<span><span class="co">#&gt; 3             22   4320</span></span>
<span><span class="co">#&gt; 4             23   6510</span></span>
<span><span class="co">#&gt; 5             24 101647</span></span></code></pre>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis-with-deseq2-or-edger">Differential Expression analysis with DESeq2 or edgeR<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-with-deseq2-or-edger"></a>
</h3>
<p>Differential expression (DE) analysis is undertaken to identify sRNA which are statistically significant to discover quantitative changes in the expression levels between the treatment (hetero-grafting) and the control (self-grafting) groups. This technique can be undertaken with a variety of tools, in <code>mobileRNA</code> users have the option to use the <code>DESeq2</code> or <code>edgeR</code> analytical method.</p>
<p>Although both methods will produce different results, it is typically they case that that at the step where we isolate the mobile molecules, the same results are produce. What is meant by this, is that the same mobile molecules are typically identified at the end of the analysis regardless of the analytical method. However, in particular case, one method will be preferred over the other. For instance, the <code>DESeq2</code> method is not appropriate when the experiment does not contain replicate (ie. one sample replicate per condition). On the other hand, edgeR can be used. Here, we have included the recommend practice for edgeR when the data set does not contain replicates. This option can be employed by setting a custom dispersion value, see argument <code>dispersionValue</code>.</p>
<p>Please note that this step is optional and not a required part of the analysis.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sample conditions in order within dataframe</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Differential analysis of whole dataset: DEseq2 method </span></span>
<span><span class="va">sRNA_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"DESeq2"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">## Differential analysis: DEseq2 method</span></span>
<span><span class="co"># 24-nt sRNA data-set</span></span>
<span><span class="va">sRNA_24_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_24</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"DESeq2"</span> <span class="op">)</span></span>
<span><span class="co"># 2122-nt sRNA data-set</span></span>
<span><span class="va">sRNA_2122_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_2122</span>,</span>
<span>                                group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                                method <span class="op">=</span> <span class="st">"DESeq2"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Differential analysis of whole dataset: edgeR method </span></span>
<span><span class="va">sRNA_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"edgeR"</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Design matrix not provided. Switch to the classic mode.</span></span>
<span></span>
<span></span>
<span><span class="co">## Differential analysis: edgeR method</span></span>
<span><span class="va">sRNA_24_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_24</span>,</span>
<span>                             group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                             method <span class="op">=</span> <span class="st">"edgeR"</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Design matrix not provided. Switch to the classic mode.</span></span>
<span></span>
<span><span class="va">sRNA_2122_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_2122</span> ,</span>
<span>                               group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                               method <span class="op">=</span> <span class="st">"edgeR"</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Design matrix not provided. Switch to the classic mode.</span></span>
<span></span>
<span><span class="co"># view example output:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sRNA_edgeR</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   Locus chr     start       end   Cluster</span></span>
<span><span class="co">#&gt; 1 1:100000956-100001076   1 100000956 100001076 cluster_1</span></span>
<span><span class="co">#&gt; 2 1:100001486-100001554   1 100001486 100001554 cluster_2</span></span>
<span><span class="co">#&gt; 3 1:100001646-100001731   1 100001646 100001731 cluster_3</span></span>
<span><span class="co">#&gt; 4   1:10001034-10001073   1  10001034  10001073 cluster_4</span></span>
<span><span class="co">#&gt; 5 1:100011795-100011930   1 100011795 100011930 cluster_5</span></span>
<span><span class="co">#&gt; 6   1:10002000-10002065   1  10002000  10002065 cluster_6</span></span>
<span><span class="co">#&gt;   DicerCall_heterograft_1 Count_heterograft_1 RPM_heterograft_1</span></span>
<span><span class="co">#&gt; 1                      24                  19             1.908</span></span>
<span><span class="co">#&gt; 2                      24                   3             0.301</span></span>
<span><span class="co">#&gt; 3                       N                   3             0.301</span></span>
<span><span class="co">#&gt; 4                      24                   1             0.100</span></span>
<span><span class="co">#&gt; 5                      24                  11             1.105</span></span>
<span><span class="co">#&gt; 6                      24                   9             0.904</span></span>
<span><span class="co">#&gt;      MajorRNA_heterograft_1 DicerCall_heterograft_2 Count_heterograft_2</span></span>
<span><span class="co">#&gt; 1  AGUGACGGUCUCAUGUAACAUAGA                      24                  14</span></span>
<span><span class="co">#&gt; 2  AAAGAGUAGACAUAUAUUAAAUGG                      20                   5</span></span>
<span><span class="co">#&gt; 3 CUGUCUCAUGUGACACUCUUUCACU                      23                   6</span></span>
<span><span class="co">#&gt; 4  AAAUUGUCAUUUAGUCAGAGGAAU                       N                   0</span></span>
<span><span class="co">#&gt; 5  AUAGACAAAUCGCACUCGCUCUCG                      24                   2</span></span>
<span><span class="co">#&gt; 6  UUAGAUUUCUGGAGUAACAACUUU                       N                   2</span></span>
<span><span class="co">#&gt;   RPM_heterograft_2   MajorRNA_heterograft_2 DicerCall_heterograft_3</span></span>
<span><span class="co">#&gt; 1             1.803 AGUGACGGUCUCAUGUAACAUAGA                      24</span></span>
<span><span class="co">#&gt; 2             0.644     ACAUAUUUUUUUGGACGAAC                      24</span></span>
<span><span class="co">#&gt; 3             0.773  UCUGUCUCAUGUGACACUCUUUC                      24</span></span>
<span><span class="co">#&gt; 4             0.000                        N                      24</span></span>
<span><span class="co">#&gt; 5             0.258 AAGUGUCUACCCAUUGUGAUUCGC                      24</span></span>
<span><span class="co">#&gt; 6             0.258      UUAUUGACAAACUAUAACU                      22</span></span>
<span><span class="co">#&gt;   Count_heterograft_3 RPM_heterograft_3   MajorRNA_heterograft_3</span></span>
<span><span class="co">#&gt; 1                  30             2.953 AGUGACGGUCUCAUGUAACAUAGA</span></span>
<span><span class="co">#&gt; 2                   1             0.098 UAUGUCUACUCUUUAUUACUCUCC</span></span>
<span><span class="co">#&gt; 3                  11             1.083 AUCUUGUCAAGUGAAAGAGUGUCA</span></span>
<span><span class="co">#&gt; 4                  11             1.083 AUUCCUCUGACUAAAUGACAAUUU</span></span>
<span><span class="co">#&gt; 5                   5             0.492 AAUCACAAUGGGUAGACACUUAGA</span></span>
<span><span class="co">#&gt; 6                   3             0.295 UUAGAUUUCUGGAGUAACAACUUU</span></span>
<span><span class="co">#&gt;   DicerCall_selfgraft_1 Count_selfgraft_1 RPM_selfgraft_1</span></span>
<span><span class="co">#&gt; 1                    24                19           1.909</span></span>
<span><span class="co">#&gt; 2                    24                 3           0.301</span></span>
<span><span class="co">#&gt; 3                     N                 3           0.301</span></span>
<span><span class="co">#&gt; 4                    24                 1           0.100</span></span>
<span><span class="co">#&gt; 5                    24                11           1.105</span></span>
<span><span class="co">#&gt; 6                    24                 9           0.904</span></span>
<span><span class="co">#&gt;        MajorRNA_selfgraft_1 DicerCall_selfgraft_2 Count_selfgraft_2</span></span>
<span><span class="co">#&gt; 1  AGUGACGGUCUCAUGUAACAUAGA                    24                14</span></span>
<span><span class="co">#&gt; 2  AAAGAGUAGACAUAUAUUAAAUGG                    20                 5</span></span>
<span><span class="co">#&gt; 3 CUGUCUCAUGUGACACUCUUUCACU                    23                 6</span></span>
<span><span class="co">#&gt; 4  AAAUUGUCAUUUAGUCAGAGGAAU                     N                 0</span></span>
<span><span class="co">#&gt; 5  AAUGGGUAGACACUUAGAUACAUG                    24                 2</span></span>
<span><span class="co">#&gt; 6  UUAGAUUUCUGGAGUAACAACUUU                     N                 2</span></span>
<span><span class="co">#&gt;   RPM_selfgraft_2     MajorRNA_selfgraft_2 DicerCall_selfgraft_3</span></span>
<span><span class="co">#&gt; 1           1.804 AGUGACGGUCUCAUGUAACAUAGA                    24</span></span>
<span><span class="co">#&gt; 2           0.644     ACAUAUUUUUUUGGACGAAC                    24</span></span>
<span><span class="co">#&gt; 3           0.773  UCUGUCUCAUGUGACACUCUUUC                    24</span></span>
<span><span class="co">#&gt; 4           0.000                        N                    24</span></span>
<span><span class="co">#&gt; 5           0.258 AUAGACAAAUCGCACUCGCUCUCG                    24</span></span>
<span><span class="co">#&gt; 6           0.258  AGAUUUCUGGAGUAACAACUUUG                    22</span></span>
<span><span class="co">#&gt;   Count_selfgraft_3 RPM_selfgraft_3     MajorRNA_selfgraft_3 sRNA_Consensus</span></span>
<span><span class="co">#&gt; 1                30           2.954 AGUGACGGUCUCAUGUAACAUAGA          nt_24</span></span>
<span><span class="co">#&gt; 2                 1           0.098 UAUGUCUACUCUUUAUUACUCUCC          nt_24</span></span>
<span><span class="co">#&gt; 3                11           1.083 AUCUUGUCAAGUGAAAGAGUGUCA          nt_23</span></span>
<span><span class="co">#&gt; 4                11           1.083 AUUCCUCUGACUAAAUGACAAUUU          nt_24</span></span>
<span><span class="co">#&gt; 5                 5           0.492 AAUCACAAUGGGUAGACACUUAGA          nt_24</span></span>
<span><span class="co">#&gt; 6                 3           0.295   UUGACAAACUAUAACUUAAAUU          nt_22</span></span>
<span><span class="co">#&gt;   CountMean log2FoldChange pvalue padjusted    logCPM</span></span>
<span><span class="co">#&gt; 1 21.000000  -6.349360e-04      1         1 3.1418030</span></span>
<span><span class="co">#&gt; 2  3.000000  -3.297740e-03      1         1 0.9823983</span></span>
<span><span class="co">#&gt; 3  6.666667  -1.671208e-03      1         1 1.7404592</span></span>
<span><span class="co">#&gt; 4  4.000000   4.340157e-05      1         1 1.1317553</span></span>
<span><span class="co">#&gt; 5  6.000000  -2.400719e-04      1         1 1.5852815</span></span>
<span><span class="co">#&gt; 6  4.666667  -3.249351e-04      1         1 1.3337822</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="identify-the-mobile-molecules">Identify the mobile molecules<a class="anchor" aria-label="anchor" href="#identify-the-mobile-molecules"></a>
</h3>
<p>It is essential to remove the noise from the data to isolate potential mobile molecules which include falsely mapped clusters and, in this case, the clusters which map the the scion genotype when the aim is to identify molecules traveling from the rootstock. In the data, the aim is to explore whether there are tomato sRNA traveling up through the graft junction and exiting into the eggplant leaf tissues. We can identify potential mobile molecule by using the <code><a href="../reference/RNAmobile.html">RNAmobile()</a></code> function. It selects clusters mapping to the rootstock genome in the hetero-graft. Therefore, we will remove clusters mapped to chromosome in the genotype used in the self-graft controls.</p>
<p>In the example, the tomato genome has chromosome labeled with “SL40” whilst the the eggplant genome has no labels. To remove clusters associated with the eggplant genome and keep only clusters associated with the tomato genome, set the <code>"task"</code> argument as <code>"keep"</code>.</p>
<p>The <code><a href="../reference/RNAmobile.html">RNAmobile()</a></code> function can take into account the statistically significance. If set to consider this, the function will filter sRNA based on an adjusted p-value threshold = 0.05. The threshold can be changed to make it less stringent by using the argument <code>padj</code>.Similarly, if you would prefer to extract the mobile RNA based on the p-value,rather than the adjusted p-values, a numeric threshold can be set for the argument <code>p.value</code>. As default, the function does not consider the statistical values and properties of the sRNAs. To consider, use <code>statistical=TRUE</code></p>
<p>Below we have used the datasets which contains the additional columns with the statistical values. As stated above, we do not need to use the statistical values to identify the mobile molecules but it can be explored and utilised to plot a heatmap based on the logFC.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># vector of control names</span></span>
<span><span class="va">control_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span>, <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Identify potential tomato mobile molecules, before subsetting classes</span></span>
<span><span class="va">mobile_sRNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         id <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or identify within specific classes </span></span>
<span><span class="co"># within 24-nt sRNA datset </span></span>
<span><span class="va">mobile_sRNA_24</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_24_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         id <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span><span class="op">)</span></span>
<span><span class="co"># within 21/22-nt sRNA datset</span></span>
<span><span class="va">mobile_sRNA_2122</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_2122_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         id <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># If you wish to consider the statistical values from DESeq2 analysis </span></span>
<span></span>
<span><span class="va">mobile_sRNA_sig24</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_24_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         id <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span>, </span>
<span>                         statistical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mobile_sRNA_sig2122</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_2122_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         id <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span>, </span>
<span>                         statistical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># view part of the total data set</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mobile_sRNA</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         Locus      chr    start      end        Cluster</span></span>
<span><span class="co">#&gt; 16   SL40ch01:2354374-2354700 SL40ch01  2354374  2354700 cluster_133430</span></span>
<span><span class="co">#&gt; 20     SL40ch01:281585-281652 SL40ch01   281585   281652 cluster_133435</span></span>
<span><span class="co">#&gt; 24     SL40ch01:618187-618375 SL40ch01   618187   618375 cluster_133450</span></span>
<span><span class="co">#&gt; 25 SL40ch01:64072258-64072464 SL40ch01 64072258 64072464 cluster_133451</span></span>
<span><span class="co">#&gt; 26 SL40ch01:65062584-65062714 SL40ch01 65062584 65062714 cluster_133452</span></span>
<span><span class="co">#&gt; 28 SL40ch01:66152385-66152584 SL40ch01 66152385 66152584 cluster_133455</span></span>
<span><span class="co">#&gt;    DicerCall_heterograft_1 Count_heterograft_1 RPM_heterograft_1</span></span>
<span><span class="co">#&gt; 16                      24                  14             1.406</span></span>
<span><span class="co">#&gt; 20                      24                  24             2.410</span></span>
<span><span class="co">#&gt; 24                      24                  22             2.209</span></span>
<span><span class="co">#&gt; 25                      24                  12             1.205</span></span>
<span><span class="co">#&gt; 26                      24                  18             1.808</span></span>
<span><span class="co">#&gt; 28                      24                  35             3.515</span></span>
<span><span class="co">#&gt;      MajorRNA_heterograft_1 DicerCall_heterograft_2 Count_heterograft_2</span></span>
<span><span class="co">#&gt; 16 UCCGCUAUAUAAAGCCAUUGCCAA                      24                  14</span></span>
<span><span class="co">#&gt; 20 AUGUAACUGAUUGAGUAUAUAUUU                      24                  24</span></span>
<span><span class="co">#&gt; 24  CCCGGACUGUCGGACACUACACA                      24                  22</span></span>
<span><span class="co">#&gt; 25  ACGGCCUAUUGCACACCUGAAUU                      24                  12</span></span>
<span><span class="co">#&gt; 26 AGGAUUACUUACCACUCCACCAUA                      24                  18</span></span>
<span><span class="co">#&gt; 28 ACACUUGUAGGUCUGAGUGAAACU                      24                  35</span></span>
<span><span class="co">#&gt;    RPM_heterograft_2   MajorRNA_heterograft_2 DicerCall_heterograft_3</span></span>
<span><span class="co">#&gt; 16             1.803 AGAACUCUCUCGAUUAUACUGUGU                      24</span></span>
<span><span class="co">#&gt; 20             3.090 AUGUAACUGAUUGAGUAUAUAUUU                      24</span></span>
<span><span class="co">#&gt; 24             2.833  CCCGGACUGUCGGACACUACACA                      24</span></span>
<span><span class="co">#&gt; 25             1.545  ACGGCCUAUUGCACACCUGAAUU                      24</span></span>
<span><span class="co">#&gt; 26             2.318 AGGAUUACUUACCACUCCACCAUA                      24</span></span>
<span><span class="co">#&gt; 28             4.507 ACACUUGUAGGUCUGAGUGAAACU                      24</span></span>
<span><span class="co">#&gt;    Count_heterograft_3 RPM_heterograft_3   MajorRNA_heterograft_3</span></span>
<span><span class="co">#&gt; 16                  14             1.378       CGAAGAGUAGGAUGUAUG</span></span>
<span><span class="co">#&gt; 20                  24             2.362 AUGUAACUGAUUGAGUAUAUAUUU</span></span>
<span><span class="co">#&gt; 24                  22             2.166  CCCGGACUGUCGGACACUACACA</span></span>
<span><span class="co">#&gt; 25                  12             1.181  ACGGCCUAUUGCACACCUGAAUU</span></span>
<span><span class="co">#&gt; 26                  18             1.772 AGGAUUACUUACCACUCCACCAUA</span></span>
<span><span class="co">#&gt; 28                  35             3.445 ACACUUGUAGGUCUGAGUGAAACU</span></span>
<span><span class="co">#&gt;    DicerCall_selfgraft_1 Count_selfgraft_1 RPM_selfgraft_1 MajorRNA_selfgraft_1</span></span>
<span><span class="co">#&gt; 16                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 20                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 24                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 25                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 26                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 28                     N                 0               0                    N</span></span>
<span><span class="co">#&gt;    DicerCall_selfgraft_2 Count_selfgraft_2 RPM_selfgraft_2 MajorRNA_selfgraft_2</span></span>
<span><span class="co">#&gt; 16                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 20                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 24                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 25                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 26                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 28                     N                 0               0                    N</span></span>
<span><span class="co">#&gt;    DicerCall_selfgraft_3 Count_selfgraft_3 RPM_selfgraft_3 MajorRNA_selfgraft_3</span></span>
<span><span class="co">#&gt; 16                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 20                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 24                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 25                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 26                     N                 0               0                    N</span></span>
<span><span class="co">#&gt; 28                     N                 0               0                    N</span></span>
<span><span class="co">#&gt;    sRNA_Consensus  baseMean log2FoldChange       pvalue    padjusted</span></span>
<span><span class="co">#&gt; 16          nt_24  7.538054      -6.327241 5.181896e-06 5.145198e-03</span></span>
<span><span class="co">#&gt; 20          nt_24 12.922378      -7.108850 6.879236e-08 1.207715e-04</span></span>
<span><span class="co">#&gt; 24          nt_24 11.845513      -6.982577 1.426551e-07 2.304089e-04</span></span>
<span><span class="co">#&gt; 25          nt_24  6.461189      -6.103991 1.622151e-05 1.383809e-02</span></span>
<span><span class="co">#&gt; 26          nt_24  9.691784      -6.691479 7.325720e-07 9.645744e-04</span></span>
<span><span class="co">#&gt; 28          nt_24 18.845135      -7.656902 2.604860e-09 7.012053e-06</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="heatmap-plots-to-represent-mobile-molecules">Heatmap plots to represent mobile molecules<a class="anchor" aria-label="anchor" href="#heatmap-plots-to-represent-mobile-molecules"></a>
</h4>
<p>Here we can use a hierarchical clustering heatmap to analysis the potential statistically significant mobile sRNAs we have identified. It is expected that the log2 fold changes should be opposite between your treatment and controls where the treatment samples shows up-regulation and the control samples show down-regulation.</p>
<p><br></p>
</div>
<div class="section level4">
<h4 id="mobile-molecules-from-the-24-nt-srna-dataset">Mobile molecules from the 24-nt sRNA dataset<a class="anchor" aria-label="anchor" href="#mobile-molecules-from-the-24-nt-srna-dataset"></a>
</h4>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p10</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">mobile_sRNA_sig24</span><span class="op">)</span></span>
<span><span class="va">p11</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">mobile_sRNA_sig2122</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-17-1.png" alt="An example heatmap of the potential mobile RNA molecules Identifed by the `DESeq2` method Class of 24-nt sRNAs (left) and Class of 21/22-nt sRNAs (right). Where the rows represents the sRNA cluster number and the columns represents replicates." width="50%"><img src="mobileRNA_files/figure-html/unnamed-chunk-17-2.png" alt="An example heatmap of the potential mobile RNA molecules Identifed by the `DESeq2` method Class of 24-nt sRNAs (left) and Class of 21/22-nt sRNAs (right). Where the rows represents the sRNA cluster number and the columns represents replicates." width="50%"><p class="caption">
An example heatmap of the potential mobile RNA molecules Identifed by the <code>DESeq2</code> method Class of 24-nt sRNAs (left) and Class of 21/22-nt sRNAs (right). Where the rows represents the sRNA cluster number and the columns represents replicates.
</p>
</div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="save-output">Save output<a class="anchor" aria-label="anchor" href="#save-output"></a>
</h4>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">mobile_sRNA</span>, <span class="st">"./output.txt"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="functional-analysis">Functional Analysis<a class="anchor" aria-label="anchor" href="#functional-analysis"></a>
</h3>
<p>By now you will have retrieved a list of potentially mobile sRNA molecules and you will want to identify whether they play a role in the biological system.</p>
<p>Small interfering RNAs (siRNAs) have a length of 24-nucleotides, and are known to play a role in the RNA-directed DNA methylation pathway via targeting complementary sequencing in the DNA. While, miRNAs have a length of 21/22-nucleotides, and are known to instigate post-transcriptional gene silencing via mRNA degradation or repression of translation.</p>
<p>To predict the function of a mobile sRNA in destination tissues, we can overlap the information with the genome annotation of the origin tissue to determine the local functionality. As well as, retrieving the nucleotide sequence of the most abundant sRNA within a cluster and locating complementary sequences in both the genome of origin and genome of destination using BLAST.</p>
<p>IMPORTANT: If chromosome names were altered in either of genome assemblies used in the mapping steps, the chromosome names in the genome annotation files must match the same pattern. If you used the [mobileRNA::RNAmergeGenomes()] function, the chromosome names will by default be altered. The [mobileRNA::RNAmergeAnnotations()] function can be utilsed to alter both genome annotation files in the same way. This functions alters the names, saves each altered genome annotation and then merges them into one.</p>
<p><br></p>
<div class="section level4">
<h4 id="add-genomic-attributes-to-srna-clusters">Add genomic attributes to sRNA clusters<a class="anchor" aria-label="anchor" href="#add-genomic-attributes-to-srna-clusters"></a>
</h4>
<p>Each sRNA cluster contains coordinates, these can be matches with coordinates in an annotation file. If there is a match, the function returns the input data with additional fields, where sRNA cluster will receive information on:</p>
<ul>
<li>
<code>source</code> : name of the program that generated this feature, or the data source (database or project name)</li>
<li>
<code>feature</code> : feature type name, e.g. Gene, Variation, Similarity</li>
<li>
<code>score</code> : A floating point value</li>
<li>
<code>strand</code> : defined as + (forward) or - (reverse)</li>
<li>
<code>frame</code> : One of ‘0’, ‘1’ or ‘2’. ‘0’ indicates that the first base of the feature is the first base of a codon, ‘1’ that the second base is the first base of a codon, and so on</li>
<li>
<code>attribute</code> : A semicolon-separated list of tag-value pairs, providing additional information about each feature.</li>
</ul>
<p>Using the genomic locations of the mobile sRNA clusters, we can identify overlaps in the genome annotation of the origin tissue to predict the implication of the mobile sRNA.</p>
<p>This function can be utilsed at the beginning of the analysis steps where all sRNA clusters from both genotypes can be annotated with the genomic attributes associated to their location. This can be completed in a single step using a merge annotation file using the function [mobileRNA::RNAmergeAnnotations()].</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mobile_df_attributes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAattributes.html">RNAattributes</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mobile_sRNA</span>,</span>
<span>                            annotation <span class="op">=</span> <span class="st">"./annotation/origin_annotation.gff3"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="summarise-overlaps-with-genomic-features">Summarise overlaps with genomic features<a class="anchor" aria-label="anchor" href="#summarise-overlaps-with-genomic-features"></a>
</h4>
<p>We can also locate specific genomic features which overlap with the the sRNA cluster locus including promoter regions, exon, introns, untranslated regions and repeat regions.The results can either be displayed in the matrix as an absolute value or as a percentage of the total:</p>
<p>NOTE: This function look for the exact matches in the locus (chromosome, start and end regions) between the sRNA cluster and the annotation file.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mobile_df_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAfeatures.html">RNAfeatures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mobile_sRNA</span>,</span>
<span>                            annotation <span class="op">=</span> <span class="st">"./annotation/origin_annotation.gff3"</span>, </span>
<span>                      repeats <span class="op">=</span> <span class="st">"./annotation/origin_annotation_repeats.gff3"</span><span class="op">)</span></span></code></pre></div>
<p>The distribution of sRNA clusters across genomic features is strongly related to the class of sRNA based on their nucleotide length. It is expected that 24-nt sRNA are associated with repeat regions and promoter regions, while 21/22-nt sRNA are likely to be associated with coding sequence.</p>
<p>The resulting dataframe can be passed into a plot of your choice. Similarly, this features can be utilsed across different data sets. For example, across the total data set or specific mobile sRNA subsets. <br></p>
</div>
<div class="section level4">
<h4 id="retrieve-srna-sequence-complementary-sequences">Retrieve sRNA sequence &amp; complementary sequences<a class="anchor" aria-label="anchor" href="#retrieve-srna-sequence-complementary-sequences"></a>
</h4>
<p>Lastly, lets extract the sRNA cluster sequences.These can utilised with <code>BLAST</code>, or plant sRNA target prediction tools such as <code>TargetFinder</code> (<a href="https://github.com/carringtonlab/TargetFinder" class="external-link uri">https://github.com/carringtonlab/TargetFinder</a>) or overlapped with <code>mRNAseq</code> data to identify complementary sequences to further elucidate the potential function of sRNAs.</p>
<p>The <code>RNAsequences</code> function identifies whether the most abundant sRNA is consistent across the replicates, and if so, it extracts the sRNA nucleotide sequence and calculates the RNA and DNA complementary sequences, as well as stating the length of the sequence.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mobile_sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsequences.html">RNAsequences</a></span><span class="op">(</span><span class="va">mobile_sRNA</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mobile_sequences</span><span class="op">)</span></span>
<span><span class="co">#&gt;           Cluster Match                 Sequence width        Complementary_RNA</span></span>
<span><span class="co">#&gt; 16 cluster_133430 FALSE                     &lt;NA&gt;    NA                       NA</span></span>
<span><span class="co">#&gt; 20 cluster_133435  TRUE AUGUAACUGAUUGAGUAUAUAUUU    24 UACAUUGACUAACUCAUAUAUAAA</span></span>
<span><span class="co">#&gt; 24 cluster_133450  TRUE  CCCGGACUGUCGGACACUACACA    23  GGGCCUGACAGCCUGUGAUGUGU</span></span>
<span><span class="co">#&gt; 25 cluster_133451  TRUE  ACGGCCUAUUGCACACCUGAAUU    23  UGCCGGAUAACGUGUGGACUUAA</span></span>
<span><span class="co">#&gt; 26 cluster_133452  TRUE AGGAUUACUUACCACUCCACCAUA    24 UCCUAAUGAAUGGUGAGGUGGUAU</span></span>
<span><span class="co">#&gt; 28 cluster_133455  TRUE ACACUUGUAGGUCUGAGUGAAACU    24 UGUGAACAUCCAGACUCACUUUGA</span></span>
<span><span class="co">#&gt;           Complementary_DNA</span></span>
<span><span class="co">#&gt; 16                       NA</span></span>
<span><span class="co">#&gt; 20 TACATTGACTAACTCATATATAAA</span></span>
<span><span class="co">#&gt; 24  GGGCCTGACAGCCTGTGATGTGT</span></span>
<span><span class="co">#&gt; 25  TGCCGGATAACGTGTGGACTTAA</span></span>
<span><span class="co">#&gt; 26 TCCTAATGAATGGTGAGGTGGTAT</span></span>
<span><span class="co">#&gt; 28 TGTGAACATCCAGACTCACTTTGA</span></span></code></pre></div>
<p><br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="additional-features">Additional features<a class="anchor" aria-label="anchor" href="#additional-features"></a>
</h2>
<p>The <em>mobileRNA</em> package offers several optional extra-helpful functions to aid analysis. This includes a function to calculate the mean RPM and counts across specific samples and also calculate the overlap between genomic features and sRNA clusters.</p>
<p><br></p>
<div class="section level3">
<h3 id="calculate-rpm-and-count-means-for-specific-samples">Calculate RPM and Count means for specific samples<a class="anchor" aria-label="anchor" href="#calculate-rpm-and-count-means-for-specific-samples"></a>
</h3>
<p>This function calculates the RPM mean and Count mean across specific samples.The output of the calculation is added as an additional column to the input dataframe.</p>
<p>This can be manipulated to be used before or after identified mobile RNAs. When calculating means after identifying mobile molecules, it is recommend to base this on the heterograft samples rather than the full dataset (selfgrafts + heterograft replicates) as all potential mobile sRNAs should have zero counts or RPM value in the selfgraft control replicates. By including the full dataset in these calculations, the mean values for the mobile molecules will be skewed.</p>
<p>Therefore, depending on your intended output this function could be approached in multiple ways.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate mean RPM and mean Counts for heterograft samples</span></span>
<span><span class="va">selected_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span>, <span class="st">"heterograft_3"</span><span class="op">)</span></span>
<span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmean.html">RNAmean</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>, conditions <span class="op">=</span> <span class="va">selected_samples</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.0.2 (2020-06-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS  10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] mobileRNA_1.0.0  BiocStyle_2.16.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] colorspace_2.0-3            rprojroot_2.0.2            </span></span>
<span><span class="co">#&gt;   [3] DNAcopy_1.62.0              XVector_0.28.0             </span></span>
<span><span class="co">#&gt;   [5] GenomicRanges_1.40.0        fs_1.5.2                   </span></span>
<span><span class="co">#&gt;   [7] rstudioapi_0.14             farver_2.1.0               </span></span>
<span><span class="co">#&gt;   [9] listenv_0.8.0               audio_0.1-10               </span></span>
<span><span class="co">#&gt;  [11] affyio_1.58.0               ggrepel_0.9.1              </span></span>
<span><span class="co">#&gt;  [13] bit64_4.0.5                 AnnotationDbi_1.50.3       </span></span>
<span><span class="co">#&gt;  [15] fansi_1.0.2                 codetools_0.2-18           </span></span>
<span><span class="co">#&gt;  [17] splines_4.0.2               Repitools_1.34.0           </span></span>
<span><span class="co">#&gt;  [19] cachem_1.0.6                geneplotter_1.66.0         </span></span>
<span><span class="co">#&gt;  [21] knitr_1.42                  jsonlite_1.8.4             </span></span>
<span><span class="co">#&gt;  [23] Rsamtools_2.4.0             annotate_1.66.0            </span></span>
<span><span class="co">#&gt;  [25] cluster_2.1.2               vsn_3.56.0                 </span></span>
<span><span class="co">#&gt;  [27] pheatmap_1.0.12             BiocManager_1.30.16        </span></span>
<span><span class="co">#&gt;  [29] compiler_4.0.2              Matrix_1.4-0               </span></span>
<span><span class="co">#&gt;  [31] fastmap_1.1.0               limma_3.44.3               </span></span>
<span><span class="co">#&gt;  [33] cli_3.6.0                   beepr_1.3                  </span></span>
<span><span class="co">#&gt;  [35] htmltools_0.5.5             tools_4.0.2                </span></span>
<span><span class="co">#&gt;  [37] gtable_0.3.0                glue_1.6.2                 </span></span>
<span><span class="co">#&gt;  [39] GenomeInfoDbData_1.2.3      affy_1.66.0                </span></span>
<span><span class="co">#&gt;  [41] dplyr_1.1.2                 Rcpp_1.0.8.3               </span></span>
<span><span class="co">#&gt;  [43] Biobase_2.48.0              jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt;  [45] pkgdown_2.0.7               vctrs_0.6.2                </span></span>
<span><span class="co">#&gt;  [47] Biostrings_2.56.0           preprocessCore_1.50.0      </span></span>
<span><span class="co">#&gt;  [49] progressr_0.10.0            rtracklayer_1.48.0         </span></span>
<span><span class="co">#&gt;  [51] iterators_1.0.14            xfun_0.37                  </span></span>
<span><span class="co">#&gt;  [53] stringr_1.5.0               globals_0.14.0             </span></span>
<span><span class="co">#&gt;  [55] lifecycle_1.0.3             gtools_3.9.2               </span></span>
<span><span class="co">#&gt;  [57] XML_3.99-0.9                future_1.24.0              </span></span>
<span><span class="co">#&gt;  [59] edgeR_3.30.3                zlibbioc_1.34.0            </span></span>
<span><span class="co">#&gt;  [61] MASS_7.3-55                 scales_1.2.1               </span></span>
<span><span class="co">#&gt;  [63] BSgenome_1.56.0             ragg_1.2.2                 </span></span>
<span><span class="co">#&gt;  [65] parallel_4.0.2              SummarizedExperiment_1.18.2</span></span>
<span><span class="co">#&gt;  [67] RColorBrewer_1.1-2          yaml_2.3.5                 </span></span>
<span><span class="co">#&gt;  [69] curl_4.3.2                  pbapply_1.7-0              </span></span>
<span><span class="co">#&gt;  [71] memoise_2.0.1               ggplot2_3.4.2              </span></span>
<span><span class="co">#&gt;  [73] sass_0.4.0                  stringi_1.7.6              </span></span>
<span><span class="co">#&gt;  [75] RSQLite_2.2.10              highr_0.9                  </span></span>
<span><span class="co">#&gt;  [77] genefilter_1.70.0           S4Vectors_0.26.1           </span></span>
<span><span class="co">#&gt;  [79] desc_1.4.1                  foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [81] Ringo_1.52.0                caTools_1.18.2             </span></span>
<span><span class="co">#&gt;  [83] BiocGenerics_0.34.0         BiocParallel_1.22.0        </span></span>
<span><span class="co">#&gt;  [85] truncnorm_1.0-8             GenomeInfoDb_1.24.2        </span></span>
<span><span class="co">#&gt;  [87] rlang_1.1.0                 pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [89] systemfonts_1.0.4           matrixStats_0.61.0         </span></span>
<span><span class="co">#&gt;  [91] bitops_1.0-7                Rsolnp_1.16                </span></span>
<span><span class="co">#&gt;  [93] evaluate_0.15               lattice_0.20-45            </span></span>
<span><span class="co">#&gt;  [95] purrr_1.0.1                 labeling_0.4.2             </span></span>
<span><span class="co">#&gt;  [97] GenomicAlignments_1.24.0    bit_4.0.4                  </span></span>
<span><span class="co">#&gt;  [99] tidyselect_1.2.0            parallelly_1.30.0          </span></span>
<span><span class="co">#&gt; [101] magrittr_2.0.3              bookdown_0.32              </span></span>
<span><span class="co">#&gt; [103] DESeq2_1.28.1               R6_2.5.1                   </span></span>
<span><span class="co">#&gt; [105] IRanges_2.22.2              gplots_3.1.1               </span></span>
<span><span class="co">#&gt; [107] generics_0.1.2              DelayedArray_0.14.1        </span></span>
<span><span class="co">#&gt; [109] DBI_1.1.2                   withr_2.5.0                </span></span>
<span><span class="co">#&gt; [111] pillar_1.9.0                gsmoothr_0.1.7             </span></span>
<span><span class="co">#&gt; [113] RPushbullet_0.3.4           survival_3.3-1             </span></span>
<span><span class="co">#&gt; [115] RCurl_1.98-1.6              future.apply_1.8.1         </span></span>
<span><span class="co">#&gt; [117] tibble_3.2.1                crayon_1.5.0               </span></span>
<span><span class="co">#&gt; [119] KernSmooth_2.23-20          utf8_1.2.2                 </span></span>
<span><span class="co">#&gt; [121] rmarkdown_2.20              locfit_1.5-9.4             </span></span>
<span><span class="co">#&gt; [123] grid_4.0.2                  data.table_1.14.2          </span></span>
<span><span class="co">#&gt; [125] blob_1.2.2                  SimDesign_2.10.1           </span></span>
<span><span class="co">#&gt; [127] digest_0.6.29               xtable_1.8-4               </span></span>
<span><span class="co">#&gt; [129] tidyr_1.3.0                 textshaping_0.3.6          </span></span>
<span><span class="co">#&gt; [131] stats4_4.0.2                munsell_0.5.0              </span></span>
<span><span class="co">#&gt; [133] bslib_0.3.1                 sessioninfo_1.2.2</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
