<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="mobileRNA">
<title>Explore canditate mobile mRNAs &amp; population-scale mRNA changes in chimeric systems using mobileRNA • mobileRNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Explore canditate mobile mRNAs &amp; population-scale mRNA changes in chimeric systems using mobileRNA">
<meta property="og:description" content="mobileRNA">
<meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/mobileRNA.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/mobileRNA_standard.html">Analyse sRNA populations using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_chimeric.html">Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_mRNA.html">Explore canditate mobile mRNAs &amp; population-scale mRNA changes in chimeric systems using mobileRNA</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="mobileRNA_mRNA_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Explore canditate mobile mRNAs &amp; population-scale mRNA changes in chimeric systems using mobileRNA</h1>
                        <h4 data-toc-skip class="author">Katie Jeynes-Cupper, Marco Catoni</h4>
            
            <h4 data-toc-skip class="date">2023-08-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KJeynesCupper/mobileRNA.git/blob/HEAD/vignettes/mobileRNA_mRNA.Rmd" class="external-link"><code>vignettes/mobileRNA_mRNA.Rmd</code></a></small>
      <div class="d-none name"><code>mobileRNA_mRNA.Rmd</code></div>
    </div>

    
    
<p><br></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Similarly to sRNAs, messenger RNA (mRNA) molecules have also been believed to move systemically throughout plants, and grafting has been a tool to study this. Interspecific grafting could lead to the movement of foreign endogenous mRNAs from one genotype to another. Here, we describe a pipeline to assist with this specific analysis.</p>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>This manual offers a pipeline for the analysis of mRNAseq data taken from chimeric or non-chimeric systems. For example, plant grafting experiments between different cultivars and species (with distinct genome assemblies). This pipeline involves pre-processing and analysis steps for the identification of candidate mobile mRNA molecules and the exploration of population level changes in chimeric systems.</p>
<p><strong>For chimeric systems: </strong> To make the best possible prediction of candidate mobile mRNA, here we have utilised the optimisations of the mapping algorithm in the pre-processing of clean mRNA reads. Here, we supply the mapping algorithms with as much information as possible to best place the mRNA in its most likely location between two genome assemblies. Hence, this includes supplying two genome assemblies at the same time, through the use of a merged genome. A merged genome is a single FASTA file containing multiple genome assemblies. In many databases, the chromosome names have been consistently named using the same patterns. This means that when you merge multiple genome assemblies together it is very challenging to distinguish the individual genomes in the merged file as they have similar chromosome names. Here, we offer a function to merge two genome assemblies or annotations into a single file, and the function is able to maintain the distinguish-ability by adding a distinct pre-fix to chromosome names in each genome. The second piece of extra information we can supply th algorithm is a list of all gene loci identified across the sample replicates in the analysis. This ensure consistency across the analysis and helps prevent false assumptions. Collectively, considering the extra information should optimise the alignment of mRNAs to the bets locations within the merged genome.</p>
<p>The analysis is a simple 5-step workflow to identify both candidate mobile mRNAs and explore the population-scale dynamics in one. Additional features are available and recommend to include in the analysis which includes quality<br>
control analysis (PCA, comparative heatmap) and plotting of the results as a heatmap representing the normalised FPKM and a volcano plot illustrating the logfoldchange against the false discovery rate (FDR). When working with a chimeric system, mapping error can easily be recognised and required functions can remove these when specified (see , and see ).</p>
<p>Finally, during identification of the candidate mobile molecules (see function ) the user can set a threshold parameter to either result in a more stringent or looser outcomes. The threshold parameter filters the data to select a minimum number of replicates in the analysis which contained read counts for a locus. This will enable users to select candidate mobile mRNAs which are consistently represented across the treatment replicates.</p>
<p>Here we introduce the use of statistical analysis to identify mobile mRNA molecules in chimeric systems, and it should be emphasized that this is an optional step. However, statistical analysis for the exploration of population dynamics should be a mandatory step. Here we include a single function which can be instructed to undertake differential analysis using either the edgeR or DESeq2 method (see function ). The function outputs additional columns to your working dataset, including the raw count mean, log2FoldChange, p-value and adjusted p-value.</p>
<p><strong>For non-chimeric systems: </strong> Similar to the pre-processing steps for the chimeric system, it is suggest that users still employ the first “De Novo mRNA identification” step to assist the aligning algorithm. Although, no merged genome will be employed.</p>
<p>Non-chimeric systems involve a single genotype, hence, the pipeline for exploring population-scale mRNA changes is more fitting. This pipeline will allow the user to identify mRNA which are statistically more or less abundant in the treatment compared to the control and whether any unique populations are found within either the treatment or the control samples.</p>
<p>To explore changes in mRNA abundance, we can utilise statistical analysis with either <code>DESeq2</code> or <code>edgeR</code>. For either method, the recommended/standard parameter and steps have been employed. The function outputs additional columns to your working dataset, including the raw count mean, log2FoldChange, p-value and adjusted p-value. The differences in abundance can be explored by extracting statistically significant mRNA-genes from the analysis.</p>
<div class="section level3">
<h3 id="workflow-summary">Workflow Summary<a class="anchor" aria-label="anchor" href="#workflow-summary"></a>
</h3>
<p>The summarised workflow is shown below, where it begins in R-Studio to merge the two genome assemblies into one, then the pre-processing moves into Linux to align each replicate to the merged reference and then back into R-Studio to undertake the analysis to identify potentially mobile RNA species.</p>
<p><img src="../reference/figures/program_flow.png" width="400" height="350"><br></p>
<p>The recommended analysis pipeline in R can be seen below:</p>
<p><img src="../reference/figures/mRNA_pipeline.png" width="665"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The latest version of <code>mobileRNA</code> can be installed via GitHub using the <code>devtools</code> package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"KJeynesCupper/mobileRNA"</span>, ref <span class="op">=</span> <span class="st">"main"</span><span class="op">)</span></span></code></pre></div>
<p>Load package to library:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KJeynesCupper/mobileRNA.git" class="external-link">mobileRNA</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="an-overview-of-the-data-used">An overview of the data used<a class="anchor" aria-label="anchor" href="#an-overview-of-the-data-used"></a>
</h3>
<p>For the following examples, a semi-synthetic messenger RNA-seq data set has been utilised to simulate the movement of mRNA molecules from an Tomato (<em>Solanium lycopersicum</em>) rootstock to a Eggplant (<em>Solanium melongena</em>) scion, a grafting system known to be compatible.</p>
<ul>
<li><p>The three heterograft replicates: known as:<code>heterograft_1</code>, <code>heterograft_2</code> &amp; <code>heterograft_3</code> are individual tomato replicates spiked with 150 tomato sRNA clusters.</p></li>
<li><p>The self-graft replicates: known as <code>selfgraft_1</code>, <code>selfgraft_2</code> &amp; <code>selfgraft_3</code>, are the individual tomato replicates without the spiked tomato sRNA clusters.</p></li>
</ul>
<p>The replicates mirror each other where, for instance, <code>heterograft_1</code> and <code>selfgraft_1</code> are the same replicate, either with or without the spiked clusters. These replicates serve for the aim to analyse mRNA mobility.</p>
<p>The data set, called <code>mRNA_data</code>, stores a matrix containing the pre-processed data from the experiment. As a user, this allows you to see what a full data set might look like and how you might expect it to change across the analysis. Please not that the data has been doctored to reduce the size of the file.</p>
<p>These can be loaded in these R workspace by using the following command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"mRNA_data"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="data-organisation">Data organisation<a class="anchor" aria-label="anchor" href="#data-organisation"></a>
</h3>
<p>There are two key elements required for the pipeline analysis: * sRNA-seq sample replicates (.fastq/.fq) * Two reference genomes (.fasta/.fa)</p>
<p>And one additional element; to improve functional analysis: * Reference genome annotations (.gff/.gff3) <br><br></p>
<div class="section level4">
<h4 id="naming-data-files">Naming data files<a class="anchor" aria-label="anchor" href="#naming-data-files"></a>
</h4>
<p>It is recommended to rename your files to names you wish for them to be represented as within the analysis and shown as labels in plots. Plus, it makes the analysis easier!</p>
<p>For example, instead of names such as:</p>
<ul>
<li>1: <code>sample_1.fq</code>
</li>
<li>2: <code>sample_2.fq</code>
</li>
<li>3: <code>sample_3.fq</code>
</li>
<li>4: <code>sample_4.fq</code>
</li>
<li>5: <code>sample_5.fq</code>
</li>
<li>6: <code>sample_6.fq</code>
</li>
</ul>
<p>For the example data set included in the package, here we have renamed the files based on the condition (treatment or control).For the hetero-grafts, where the is a eggplant scion and an tomato rootstock:</p>
<ul>
<li>1: <code>heterograft_1.fq</code>
</li>
<li>2: <code>heterograft_2.fq</code>
</li>
<li>3: <code>heterograft_3.fq</code>
</li>
</ul>
<p>and for the eggplant self-grafts:</p>
<ul>
<li>4: <code>selfgraft_1.fq</code>
</li>
<li>5: <code>selfgraft_2.fq</code>
</li>
<li>6: <code>selfgraft_3.fq</code> <br>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="the-reference-genomes-and-annotation-files--fasta-fa">The reference genomes and annotation files (.fasta/.fa)<a class="anchor" aria-label="anchor" href="#the-reference-genomes-and-annotation-files--fasta-fa"></a>
</h4>
<p>The pipeline is designed to analyse grafting systems with two distinct genomes, here tomato and eggplant. <br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="pre-processing">Pre-Processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h2>
<p>The pre-processing step involves cleaning raw data and aligning data to the merged genome. Going forward, the pipeline has assumed that the mRNA-seq samples have met quality control standards.</p>
<p>Here, we introduce an alternative mapping method for the analysis of plant heterograft samples. The heterograft system involves two genotypes; here the two genome references are merged into a single reference to which samples are aligned to.</p>
<p><code>mobileRNA</code> offers a function to merge two FASTA reference genomes into one. To distinguish between the reference genomes in a merged file, it is important to make sure the chromosome names between the genomes are different and distinguishable. The function below added a particular character string to the start of each chromosome name in each reference genome. As standard, the string <code>A_</code> is added to the reference genome supplied to <code>genomeA</code> and <code>B_</code> is added to the reference genome supplied to <code>genomeB.</code> These can be customized to the users preference, see manual for more information. <br></p>
<div class="section level3">
<h3 id="pre-mapping">Pre-mapping<a class="anchor" aria-label="anchor" href="#pre-mapping"></a>
</h3>
<p>Here we will merge two genome assemblies or annotations into a merged file. It is of the most importance that both the merged genome and merged annotation contained the same modifications to the chromosome names.</p>
<div class="section level4">
<h4 id="merging-genome-assemblies">Merging Genome Assemblies<a class="anchor" aria-label="anchor" href="#merging-genome-assemblies"></a>
</h4>
<p>Here we merge the two reference genomes into a single merged genome using the [mobileRNA::RNAgenomeMerge()] function. As default, this function changes the chromosome names of each genome to ensure they are distinguishable. To do so, the function requires an input of the initials of each organism’s Latin name. But why do this:</p>
<ul>
<li><p>If the two genomes use the same pattern to name the chromosomes, the user will not be able to differentiate the chromosomes from one another in the merged genome. This could be solves by adding letters to the chromosomes of one of the genomes, for example, <code>SM</code> to represent the Latin name of eggplant.</p></li>
<li><p>If a chromosome naming pattern contains punctuation, the mapping step will not work.</p></li>
</ul>
<p>In the example, the <em>Solanum lycopersicum</em> (version 4) genome contains a full-stop/period within each chromosome name which needs to be removed as well. Here we rename the chromosomes of the tomato genome to <code>SL</code> and the chromosomes related to the eggplant genome (<em>Solanum melongena</em>) to <code>SM</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RNAmergeGenomes.html">RNAmergeGenomes</a></span><span class="op">(</span>genomeA <span class="op">=</span><span class="st">"/Users/user1/projectname/workplace/reference/ref1.fa"</span>,</span>
<span>               genomeB <span class="op">=</span><span class="st">"/Users/user1/projectname/workplace/reference/ref2.fa"</span>,</span>
<span>               abbreviationGenomeA <span class="op">=</span> <span class="st">"SL"</span>,</span>
<span>               abbreviationGenomeB <span class="op">=</span> <span class="st">"SM"</span>,</span>
<span>               out_dir <span class="op">=</span> <span class="st">"/Users/user1/projectname/workplace/reference/merge/</span></span>
<span><span class="st">               ref_merged.fa"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="merging-genome-annotations">Merging Genome Annotations<a class="anchor" aria-label="anchor" href="#merging-genome-annotations"></a>
</h3>
<p>To assist the processing of raw count files, the genomic annotation files (GFF) of each reference genome must be merged in the same way the reference genomes were. This includes following the same naming patterns for the chromosomes in each genome as undertaken earlier with the <code><a href="../reference/RNAmergeGenomes.html">RNAmergeGenomes()</a></code> function. To undertake a similar merging between the two GFF files, use the <code>RNAmergeAnnotations</code> function.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co">## merge the annotation files into a single annotation file</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="kw">RNAmergeAnnotation</span>(<span class="dt">annotationA =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="st">                   annotation/annotation_1.gff3"</span>,</span>
<span id="cb5-4"><a href="#cb5-4"></a>                   <span class="dt">annotationB =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="st">                   annotation/annotation_1.gff3"</span>,</span>
<span id="cb5-6"><a href="#cb5-6"></a>                   <span class="dt">abbreviationAnnoA =</span> <span class="st">"SL"</span>,</span>
<span id="cb5-7"><a href="#cb5-7"></a>                   <span class="dt">abbreviationAnnoB =</span> <span class="st">"SM"</span></span>
<span id="cb5-8"><a href="#cb5-8"></a>                  <span class="dt">out_dir =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="st">                  annotation/merge/anno_merged.gff"</span>)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="auto-detection-of-mrna-loci">Auto-Detection of mRNA loci<a class="anchor" aria-label="anchor" href="#auto-detection-of-mrna-loci"></a>
</h3>
<p>Here we identify and build a list mRNA loci within each sample to assist the mapping step later on to ensure consistency across the analysis. This step is very important to consider in the analysis as it prevents false assumptions. This step aligns a replicate to the given genome assembly and the output contains loci where mRNA reads aligned.</p>
<div class="section level5">
<h5 id="step-1">Step 1<a class="anchor" aria-label="anchor" href="#step-1"></a>
</h5>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="step-2">Step 2<a class="anchor" aria-label="anchor" href="#step-2"></a>
</h5>
<p>Now, we collate all the mRNA loci information from each sample into a text file.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="mapping">Mapping<a class="anchor" aria-label="anchor" href="#mapping"></a>
</h3>
<p>Each sample is mapped to the merged reference genome with the list of mRNA clusters.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="raw-count-generation">Raw count generation<a class="anchor" aria-label="anchor" href="#raw-count-generation"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>Here, the analysis of the pre-processed data occurs, with the aim to identify if any mobile mRNA molecules and population-scale mRNA changes in (non)chimeric systems. <br></p>
<div class="section level3">
<h3 id="import-data">Import Data<a class="anchor" aria-label="anchor" href="#import-data"></a>
</h3>
<p>In the pre-processing steps, the data was cleaned and aligned to the merged reference genome. During the mapping step, a folder for each sample is created which stores all the results in. The analysis steps requires the information which is stored in the <code>Results.txt</code>.</p>
<p>The <code><a href="../reference/RNAimport.html">RNAimport()</a></code> function imports the data from the folder containing all sample folders. The function extracts the required information, stores it in a matrix.</p>
<p>The function requires some information to coordinate the importation and calculations. It requires a directory path to your processed samples; this is the path to the folder containing all the individual sample folders which is stores in argument <code>directory</code>. The second directory path is to the loci file containing information on the loci of the sRNA clusters across the analysis, stored in the argument <code>loci</code>. The last pieces of information the function requires a vector containing the sample names, both treatment and control replicates which is stored in the argument <code>samples</code>. The sample names must match the names of the folders produced in the mapping, and stored in the directory supplied to the argument <code>directory</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Import &amp; organise data.</span></span>
<span><span class="va">results_dir</span> <span class="op">&lt;-</span>  <span class="st">"./analysis/alignment_unique_two/"</span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>,<span class="st">"heterograft_2"</span>, <span class="st">"heterograft_3"</span>,</span>
<span>                  <span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span>, <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">mRNA_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAimport.html">RNAimport</a></span><span class="op">(</span>input <span class="op">=</span> <span class="st">"mRNA"</span>,</span>
<span>                       annotation <span class="op">=</span> <span class="st">"./annotation/merged_anno.gff"</span>,</span>
<span>                       directory <span class="op">=</span> <span class="va">results_dir</span>,</span>
<span>                       samples <span class="op">=</span> <span class="va">sample_names</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mRNA_data</span><span class="op">)</span></span>
<span><span class="co"># example dataframe </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mRNA_data</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="sample-level-quality-control">Sample-level quality control<a class="anchor" aria-label="anchor" href="#sample-level-quality-control"></a>
</h3>
<p>A handy step in the analysis is to assess the overall similarity between sample replicates to understand which samples are the most similar and which are different, and where the most variation is introduced in the data set. As well as understanding whether the data set meets your expectations. It is expected that between the conditions, the sample replicates show enough variation to suggest that the replicates are from different groups.</p>
<p>To investigate the sample similarity/difference, we will undertake sample-level quality control using three different methods: - Principle component analysis (PCA) - Hierarchical clustering Heatmap</p>
<p>This will show us how well samples within each condition cluster together, which may highlight outliers. Plus, to show whether our experimental conditions represent the main source of variation in the data set.</p>
<p>Here we will be employing an unsupervised clustering methods for the PCA and hierarchical clustering Heatmap. This involves an unbiased log2-transformation of the counts which will emphasis the the sample clustering to improve visualization. The DESeq2 package contains a particularly useful function to undertake regularized log transformation (rlog) which controls the variance across the mean, and in this package we have utilized this for the quality control steps.</p>
<p><br></p>
<div class="section level4">
<h4 id="principle-component-analysis-to-assess-sample-distance">Principle component analysis to assess sample distance<a class="anchor" aria-label="anchor" href="#principle-component-analysis-to-assess-sample-distance"></a>
</h4>
<p>Principal Component Analysis (PCA) is a useful technique to illustrate sample distance as it emphasizes the variation through the reduction of dimensions in the data set. Here, we introduce the function <code><a href="../reference/plotSamplePCA.html">plotSamplePCA()</a></code></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSamplePCA.html">plotSamplePCA</a></span><span class="op">(</span><span class="va">mRNA_data</span>, <span class="va">group</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="hierarchical-clustered-heatmap-to-assess-sample-distance">Hierarchical clustered heatmap to assess sample distance<a class="anchor" aria-label="anchor" href="#hierarchical-clustered-heatmap-to-assess-sample-distance"></a>
</h4>
<p>Similarly, to a PCA plot, the <code><a href="../reference/plotSampleDistance.html">plotSampleDistance()</a></code> function undertakes hierarchical clustering with an unbiased log transformation to calculate sample distance and is plotted in the form of a heatmap.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSampleDistance.html">plotSampleDistance</a></span><span class="op">(</span><span class="va">mRNA_data</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis-with-deseq2-or-edger">Differential Expression analysis with DESeq2 or edgeR<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-with-deseq2-or-edger"></a>
</h3>
<p>Differential expression (DE) analysis is undertaken to identify mRNA which are statistically significant to discover quantitative changes in the expression levels between the treatment (hetero-grafting) and the control (self-grafting) groups. This technique can be undertaken with a variety of tools, in <code>mobileRNA</code> users have the option to use the <code>DESeq2</code> or <code>edgeR</code> analytical method.</p>
<p>In particular case, one method may be preferred over the other. For instance, the <code>DESeq2</code> method is not appropriate when the experiment does not contain replicate (ie. one sample replicate per condition). On the other hand, edgeR can be used. Here, we have included the recommend practice for edgeR when the data set does not contain replicates. This option can be employed by setting a custom dispersion value, see argument <code>dispersionValue</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sample conditions in order within dataframe</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">## Differential analysis of whole dataset: DEseq2 method </span></span>
<span><span class="va">mRNA_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mRNA_data</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"DESeq2"</span> <span class="op">)</span></span>
<span></span>
<span><span class="co">## Differential analysis of whole dataset: edgeR method </span></span>
<span><span class="va">mRNA_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mRNA_data</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"edgeR"</span> <span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="identify-the-mobile-molecules">Identify the mobile molecules<a class="anchor" aria-label="anchor" href="#identify-the-mobile-molecules"></a>
</h3>
<p>It is essential to remove the noise from the data to isolate potential mobile molecules through the removal mRNA molecules which map to the tissue-origin genotype, and mapping errors.</p>
<p>To remove molecules mapped to a specific genome set the <code>"task"</code> parameter to “remove”, while to keep molecules mapped to a specific genome set as “keep”. The function can also consider statistical significance filter, whereby mRNA molecules can be selected based on an adjusted p-value threshold, which by default is set to 0.05. Similarly, if you would prefer to extract the mobile mRNA based on the p-value, rather than the adjusted p-values, a numeric threshold can be set for the argument <code>p.value</code> which will override the adjusted p-value settings. As default, the function does not consider the statistical values and properties of the sRNAs. To consider, use <code>statistical=TRUE</code></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># vector of control names</span></span>
<span><span class="va">control_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span>, <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Identify potential tomato mobile molecules, before subsetting classes</span></span>
<span><span class="va">mobile_mRNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mRNA_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         genome.ID <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span>, </span>
<span>                         statistical <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="heatmap-plots-to-represent-mobile-molecules">Heatmap plots to represent mobile molecules<a class="anchor" aria-label="anchor" href="#heatmap-plots-to-represent-mobile-molecules"></a>
</h4>
<p>Here we can use a heatmap created using the FPKM values for each sample. The function normalises the FPKM values and undertakes a log transformation before plotting.</p>
<p><br></p>
</div>
<div class="section level4">
<h4 id="mobile-molecules-from-the-24-nt-srna-dataset">Mobile molecules from the 24-nt sRNA dataset<a class="anchor" aria-label="anchor" href="#mobile-molecules-from-the-24-nt-srna-dataset"></a>
</h4>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p10</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">mobile_mRNA</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="save-output">Save output<a class="anchor" aria-label="anchor" href="#save-output"></a>
</h4>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">mobile_sRNA</span>, <span class="st">"./output.txt"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="explore-population-scale-srna-differences">Explore population-scale sRNA differences<a class="anchor" aria-label="anchor" href="#explore-population-scale-srna-differences"></a>
</h3>
<p>Here, we will look at comparing the native population of mRNAs and how they might have changes in the tissue due to the chimeric system compared to the control.</p>
<p>We have already undertaken the statistical analysis in the previous steps, we will be utilising this dataset going forward.</p>
<div class="section level4">
<h4 id="srna-abundance">sRNA abundance<a class="anchor" aria-label="anchor" href="#srna-abundance"></a>
</h4>
<p>When comparing treatment to control conditions, it might be the case that the same mRNA are found within both but there are differences in the total abundance.</p>
<p>The statistical analysis calculated the log2FC values for each mRNA by comparing the normalised counts between treatment and control. Here, a positive log2FC indicates an increased abundance of transcripts for a given mRNA in the treatment compared to the control, while negative log2FC indicates decreased abundance of transcripts. The statistical significance of the log2FC is determined by the (adjusted) pvalue also calculated in the previous step.</p>
<p>Here we will filter the data to select sRNA clusters which are statistically significant, and then plot the results as a heatmap to compare the conditions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select only significant sRNAs</span></span>
<span><span class="va">significant_mRNAs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsignificant.html">RNAsignificant</a></span><span class="op">(</span><span class="va">mRNA_DESeq2</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#plot</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">significant_mRNAs</span>, cluster <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Lets look at the plot:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span><span class="op">$</span><span class="va">plot</span> </span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="step-6-identify-absolute-mrna-population-difference">Step 6: Identify absolute mRNA population difference<a class="anchor" aria-label="anchor" href="#step-6-identify-absolute-mrna-population-difference"></a>
</h3>
<p>Finally, the <code><a href="../reference/RNApopulation.html">RNApopulation()</a></code> function can be utilised to identify unique mRNA populations found in the treatment or control condition.</p>
<p>First lets look at the mRNAs which are unique to the treatment condition. In the chimeric heterografts, we expect that the foreign sRNAs will also be selected in this pick-up, therefore, we can use the parameter <code>genome.ID</code> to remove mRNA related to the foreign genome.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select sRNA clusters only found in treatment &amp; not in the control samples</span></span>
<span><span class="va">treatment_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span> , <span class="st">"heterograft_3"</span><span class="op">)</span></span>
<span><span class="va">unique_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNApopulation.html">RNApopulation</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mRNA_DESeq2</span>, </span>
<span>                                  conditions <span class="op">=</span> <span class="va">treatment_reps</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># look at number of sRNA cluster only found in treatment </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">unique_treatment</span><span class="op">)</span></span></code></pre></div>
<p>Now, the mRNAs that are unique to the control condition:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select sRNA clusters only found in control &amp; not in the treatment samples</span></span>
<span><span class="va">control_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span> , <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span><span class="va">unique_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNApopulation.html">RNApopulation</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mRNA_DESeq2</span>, </span>
<span>                                   conditions <span class="op">=</span> <span class="va">control_reps</span><span class="op">)</span></span>
<span><span class="co"># look at number of sRNA cluster only found in control  </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">unique_control</span><span class="op">)</span> </span></code></pre></div>
<p><br></p>
<p>Several plotting functions within the package can be used to display these results, specifically and .</p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.0.2 (2020-06-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS  10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] mobileRNA_0.99.0 BiocStyle_2.16.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] colorspace_2.0-3            rprojroot_2.0.2            </span></span>
<span><span class="co">#&gt;   [3] DNAcopy_1.62.0              XVector_0.28.0             </span></span>
<span><span class="co">#&gt;   [5] GenomicRanges_1.40.0        fs_1.5.2                   </span></span>
<span><span class="co">#&gt;   [7] rstudioapi_0.14             listenv_0.8.0              </span></span>
<span><span class="co">#&gt;   [9] audio_0.1-10                affyio_1.58.0              </span></span>
<span><span class="co">#&gt;  [11] ggrepel_0.9.1               bit64_4.0.5                </span></span>
<span><span class="co">#&gt;  [13] AnnotationDbi_1.50.3        fansi_1.0.2                </span></span>
<span><span class="co">#&gt;  [15] codetools_0.2-18            splines_4.0.2              </span></span>
<span><span class="co">#&gt;  [17] Repitools_1.34.0            doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;  [19] cachem_1.0.6                geneplotter_1.66.0         </span></span>
<span><span class="co">#&gt;  [21] knitr_1.42                  jsonlite_1.8.4             </span></span>
<span><span class="co">#&gt;  [23] Rsamtools_2.4.0             annotate_1.66.0            </span></span>
<span><span class="co">#&gt;  [25] cluster_2.1.2               vsn_3.56.0                 </span></span>
<span><span class="co">#&gt;  [27] png_0.1-7                   pheatmap_1.0.12            </span></span>
<span><span class="co">#&gt;  [29] BiocManager_1.30.16         compiler_4.0.2             </span></span>
<span><span class="co">#&gt;  [31] Matrix_1.4-0                fastmap_1.1.0              </span></span>
<span><span class="co">#&gt;  [33] limma_3.44.3                cli_3.6.0                  </span></span>
<span><span class="co">#&gt;  [35] beepr_1.3                   htmltools_0.5.5            </span></span>
<span><span class="co">#&gt;  [37] tools_4.0.2                 gtable_0.3.0               </span></span>
<span><span class="co">#&gt;  [39] glue_1.6.2                  GenomeInfoDbData_1.2.3     </span></span>
<span><span class="co">#&gt;  [41] affy_1.66.0                 dplyr_1.1.2                </span></span>
<span><span class="co">#&gt;  [43] Rcpp_1.0.8.3                Biobase_2.48.0             </span></span>
<span><span class="co">#&gt;  [45] jquerylib_0.1.4             pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt;  [47] vctrs_0.6.2                 Biostrings_2.56.0          </span></span>
<span><span class="co">#&gt;  [49] preprocessCore_1.50.0       progressr_0.10.0           </span></span>
<span><span class="co">#&gt;  [51] rtracklayer_1.48.0          iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [53] xfun_0.37                   stringr_1.5.0              </span></span>
<span><span class="co">#&gt;  [55] globals_0.14.0              lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt;  [57] gtools_3.9.2                XML_3.99-0.9               </span></span>
<span><span class="co">#&gt;  [59] future_1.24.0               edgeR_3.30.3               </span></span>
<span><span class="co">#&gt;  [61] zlibbioc_1.34.0             MASS_7.3-55                </span></span>
<span><span class="co">#&gt;  [63] scales_1.2.1                BSgenome_1.56.0            </span></span>
<span><span class="co">#&gt;  [65] ragg_1.2.2                  parallel_4.0.2             </span></span>
<span><span class="co">#&gt;  [67] SummarizedExperiment_1.18.2 RColorBrewer_1.1-2         </span></span>
<span><span class="co">#&gt;  [69] curl_4.3.2                  yaml_2.3.5                 </span></span>
<span><span class="co">#&gt;  [71] gridExtra_2.3               pbapply_1.7-0              </span></span>
<span><span class="co">#&gt;  [73] memoise_2.0.1               ggplot2_3.4.2              </span></span>
<span><span class="co">#&gt;  [75] sass_0.4.0                  stringi_1.7.6              </span></span>
<span><span class="co">#&gt;  [77] RSQLite_2.2.10              highr_0.9                  </span></span>
<span><span class="co">#&gt;  [79] genefilter_1.70.0           S4Vectors_0.26.1           </span></span>
<span><span class="co">#&gt;  [81] desc_1.4.1                  foreach_1.5.2              </span></span>
<span><span class="co">#&gt;  [83] Ringo_1.52.0                caTools_1.18.2             </span></span>
<span><span class="co">#&gt;  [85] BiocGenerics_0.34.0         BiocParallel_1.22.0        </span></span>
<span><span class="co">#&gt;  [87] truncnorm_1.0-8             GenomeInfoDb_1.24.2        </span></span>
<span><span class="co">#&gt;  [89] rlang_1.1.0                 pkgconfig_2.0.3            </span></span>
<span><span class="co">#&gt;  [91] systemfonts_1.0.4           matrixStats_0.61.0         </span></span>
<span><span class="co">#&gt;  [93] bitops_1.0-7                Rsolnp_1.16                </span></span>
<span><span class="co">#&gt;  [95] evaluate_0.15               lattice_0.20-45            </span></span>
<span><span class="co">#&gt;  [97] purrr_1.0.1                 GenomicAlignments_1.24.0   </span></span>
<span><span class="co">#&gt;  [99] bit_4.0.4                   tidyselect_1.2.0           </span></span>
<span><span class="co">#&gt; [101] parallelly_1.30.0           magrittr_2.0.3             </span></span>
<span><span class="co">#&gt; [103] bookdown_0.32               DESeq2_1.28.1              </span></span>
<span><span class="co">#&gt; [105] R6_2.5.1                    IRanges_2.22.2             </span></span>
<span><span class="co">#&gt; [107] gplots_3.1.1                generics_0.1.2             </span></span>
<span><span class="co">#&gt; [109] DelayedArray_0.14.1         DBI_1.1.2                  </span></span>
<span><span class="co">#&gt; [111] pillar_1.9.0                gsmoothr_0.1.7             </span></span>
<span><span class="co">#&gt; [113] RPushbullet_0.3.4           survival_3.3-1             </span></span>
<span><span class="co">#&gt; [115] RCurl_1.98-1.6              future.apply_1.8.1         </span></span>
<span><span class="co">#&gt; [117] tibble_3.2.1                crayon_1.5.0               </span></span>
<span><span class="co">#&gt; [119] KernSmooth_2.23-20          utf8_1.2.2                 </span></span>
<span><span class="co">#&gt; [121] rmarkdown_2.20              viridis_0.6.2              </span></span>
<span><span class="co">#&gt; [123] locfit_1.5-9.4              grid_4.0.2                 </span></span>
<span><span class="co">#&gt; [125] data.table_1.14.2           blob_1.2.2                 </span></span>
<span><span class="co">#&gt; [127] SimDesign_2.10.1            digest_0.6.29              </span></span>
<span><span class="co">#&gt; [129] xtable_1.8-4                tidyr_1.3.0                </span></span>
<span><span class="co">#&gt; [131] textshaping_0.3.6           stats4_4.0.2               </span></span>
<span><span class="co">#&gt; [133] munsell_0.5.0               viridisLite_0.4.0          </span></span>
<span><span class="co">#&gt; [135] bslib_0.3.1                 sessioninfo_1.2.2</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
