<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="mobileRNA">
<title>Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA • mobileRNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA">
<meta property="og:description" content="mobileRNA">
<meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/mobileRNA.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/mobileRNA_standard.html">Analyse sRNA populations using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_chimeric.html">Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_mRNA.html">Explore canditate mobile mRNAs &amp; population-scale mRNA changes in chimeric systems using mobileRNA</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="mobileRNA_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA</h1>
                        <h4 data-toc-skip class="author">Katie Jeynes-Cupper, Marco Catoni</h4>
            
            <h4 data-toc-skip class="date">2023-08-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/KJeynesCupper/mobileRNA.git/blob/HEAD/vignettes/mobileRNA.Rmd" class="external-link"><code>vignettes/mobileRNA.Rmd</code></a></small>
      <div class="d-none name"><code>mobileRNA.Rmd</code></div>
    </div>

    
    
<p><br></p>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In plants, systemic signalling is an elaborated molecular system which coordinates plant development, integrating and transmitting the information perceived from environment to distant organs.An important role in long-distance signalling is played by small RNA molecules (sRNAs). These endogenous sRNAs can be produced in a tissue and then transported systemically across the vascular system into recipient organs, where they can induce a molecular response and coordinate physiological changes. Plant grafting can be utilse to create chimeric plant systems composed of two genotype. For example, grafting can join tomato and eggplant or different species of grapevine. Grafting of horticultural crops can lead to favorable traits, such as improved plant vigour or architecture, and resistance to stress. sRNAs have been known to move across a graft junction, hence, understanding the complex regulatory networks controlled by sRNAs traveling from different genomes could assist the development of novel methods for transferring traits into crops, as alternatives to traditional breeding or transgenic approaches.</p>
<p>Currently, for each study it has been necessary to adapt general RNA sequencing pipelines to the specific study of mobile molecules. Several genomic studies used grafting to explore and characterise mobile sRNAs in chimeric systems, however, it is clear that there is not a standardised approach.</p>
<p>Here we introduce the R Package , to aid the identification of mobile RNA molecules and explore population-scale changes in chimeric systems in comparison to its non-chimeric control. The tool offers a workflow for pre-processing and analysis. Ultimately, the pipeline offers an optimised mapping/alignment of samples to genome assemblies to improve sRNA origin, and offers analysis steps to remove mapping errors, visualization of the results, and supports output for functional analysis. <br></p>
</div>
<div class="section level2">
<h2 id="methods">Methods<a class="anchor" aria-label="anchor" href="#methods"></a>
</h2>
<p>This manual offers a pipeline for the analysis of sRNAseq data taken from chimeric systems. For example, plant grafting experiments between different cultivars and species (with distinct genome assemblies). This pipeline involves pre-processing and analysis steps for the identification of candidate mobile sRNA molecules and the exploration of population level changes in chimeric systems.</p>
<p>There are two clear area which must be optimised to make better predicts of the<br>
sRNA mobility and population dynamics in chimeric systems: * Optimisation of the alignment steps * Optimisation of analysis: sRNA classification and mapping error removal.</p>
<p>The first involves optimising the use of mapping algorithm to best place sRNA to each genome. Here, we supply the mapping algorithms with as much information as possible to make its predictions. This includes supplying both genome assemblies at the same time, through the use of a merged genome. A merged genome is a single FASTA file containing multiple genome assemblies. In many databases, the chromosome names have been consistently named in the same patterns. This means that when you merge multiple genome assemblies together it is very challenging to distinguish the individual genomes in the merged file. here, we offer a function to merge two genome assemblies into a single file and is able to maintain the distinguishability by adding a distinct pre-fix to chromosome names in each genome. The other bit of extra information to supply to the aligner, is a list of loci where sRNA clusters have been identified in each sample in the analysis. This ensure consistency across the analysis and helps prevent false assumptions. The first thing to understand is that sRNA molecules are expected to have a very low abundance. It might be the case that in one replicate the abundance of reads at a given sRNA locus was not significant enough to record as a sRNA cluster, while in another replicate it was recorded. Therefore, it would be wrong to assume there are zero reads at the loci. Collectively, considering the extra information should optimise the alignment of sRNAs to the bets locations within the merged genome.</p>
<p>The second area to optimise is the analysis in R. The analysis workflow can be completed in a simple 3-step process to identify potentially mobile sRNAs and similarly to explore the population-scale dynamics this can be achieved in a simple 4-step process. Across the analysis, additional features are available to advance the analysis. This includes plotting principal component analysis results, heatmaps and distribution plots, as well as including statistical analysis through the deployment of differential analysis method (edgeR or DESeq2), and steps to assist functional analysis. At many of the core steps, there is the ability to remove mapping errors by comparing controls to the treatment. This is a step not obviously described by other workflows but adds huge value to the analysis. Similarly, an important core step to understand is consensus dicer classification (see function ). For a given sRNA cluster, each replicate at the mapping and clustering pre-processing step will have predicted the class (default: 20-24nt). The consensus dicer classification step uses this information to define the classification of the cluster. There is a range of parameters to control how these are drawn, making it either stringent or loose, which will alter the downstream analysis. Finally, the step to identify the candidate mobile molecules (see function ) also has a parameter which can either result in stringent or looser outcomes. This parameter utilses the number of replicates which contributed to the consensus dicer classification as a threshold value. For instance, it would be more stringent to select sRNA clusters which were unambiguously classified across condition replicates.</p>
<p>Here we introduce the use of statistical analysis to identify mobile sRNA molecules in chimeric systems, and it should be emphasized that this is an optional step. However, statistical analysis for the exploration of population dynamics should be a mandatory step. Here we include a single function which can be instructed to undertake differential analysis using either the edgeR or DESeq2 method (see function ). The function outputs additional columns to your working dataset, including the raw count mean, log2FoldChange, p-value and adjusted p-value.</p>
<div class="section level3">
<h3 id="workflow-summary">Workflow Summary<a class="anchor" aria-label="anchor" href="#workflow-summary"></a>
</h3>
<p>The summarised workflow is shown below, where it begins in R-Studio to merge the two genome assemblies into one, then the pre-processing moves into Linux to align each replicate to the merged reference and then back into R-Studio to undertake the analysis to identify potentially mobile RNA species.</p>
<p><img src="../reference/figures/program_flow.png" width="400" height="350"><br></p>
<p>The analysis pipeline, splits into two tracks when trying to identify candidate mobile sRNAs and population dynamics. This can be seen in the diagram below:</p>
<p><img src="../reference/figures/basic_vs_advanced_mobile.png" width="1211"><br></p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The latest version of <code>mobileRNA</code> can be installed via GitHub using the <code>devtools</code> package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"KJeynesCupper/mobileRNA"</span>, ref <span class="op">=</span> <span class="st">"main"</span><span class="op">)</span></span></code></pre></div>
<p>Load package to library:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/KJeynesCupper/mobileRNA.git" class="external-link">mobileRNA</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="an-overview-of-the-data-used">An overview of the data used<a class="anchor" aria-label="anchor" href="#an-overview-of-the-data-used"></a>
</h3>
<p>For the following examples, a semi-synthetic small RNA-seq data set has been utilised to simulate the movement of sRNA molecules from an Tomato (<em>Solanium lycopersicum</em>) rootstock to a Eggplant (<em>Solanium melongena</em>) scion, a grafting system known to be compatible.</p>
<ul>
<li><p>The three heterograft replicates: known as:<code>heterograft_1</code>, <code>heterograft_2</code> &amp; <code>heterograft_3</code> are individual tomato replicates spiked with 150 tomato sRNA clusters.</p></li>
<li><p>The self-graft replicates: known as <code>selfgraft_1</code>, <code>selfgraft_2</code> &amp; <code>selfgraft_3</code>, are the individual tomato replicates without the spiked tomato sRNA clusters.</p></li>
</ul>
<p>The replicates mirror each other where, for instance, <code>heterograft_1</code> and <code>selfgraft_1</code> are the same replicate, either with or without the spiked clusters. These replicates serve for the aim to analyse sRNA mobility.</p>
<p>The data set, called <code>sRNA_data</code>, stores a matrix containing the pre-processed data from the experiment. As a user, this allows you to see what a full data set might look like and how you might expect it to change across the analysis. Please not that the data has been doctored to reduce the size of the file.</p>
<p>These can be loaded in these R workspace by using the following command:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sRNA_data"</span><span class="op">)</span></span></code></pre></div>
<p>In addition, there are two further datasets called “sRNA_data_consensus” and “sRNA_data_mobile” which are utilised in the function documentations. Both build on the “sRNA_data” dataset where “sRNA_data_consensus” contains the RNA consensus results formed by the function [RNAdicercall()]. While, the “sRNA_data_mobile” contains only the likely mobile sRNAs selected by the function [RNAmobile()] under the default parameters. <br></p>
</div>
<div class="section level3">
<h3 id="data-organisation">Data organisation<a class="anchor" aria-label="anchor" href="#data-organisation"></a>
</h3>
<p>There are two key elements required for the pipeline analysis: * sRNA-seq sample replicates (.fastq/.fq) * Two reference genomes (.fasta/.fa)</p>
<p>And one additional element; to improve functional analysis: * Reference genome annotations (.gff/.gff3) <br><br></p>
<div class="section level4">
<h4 id="naming-data-files">Naming data files<a class="anchor" aria-label="anchor" href="#naming-data-files"></a>
</h4>
<p>It is recommended to rename your files to names you wish for them to be represented as within the analysis and shown as labels in plots. Plus, it makes the analysis easier!</p>
<p>For example, instead of names such as:</p>
<ul>
<li>1: <code>sample_1.fq</code>
</li>
<li>2: <code>sample_2.fq</code>
</li>
<li>3: <code>sample_3.fq</code>
</li>
<li>4: <code>sample_4.fq</code>
</li>
<li>5: <code>sample_5.fq</code>
</li>
<li>6: <code>sample_6.fq</code>
</li>
</ul>
<p>For the example data set included in the package, here we have renamed the files based on the condition (treatment or control).For the hetero-grafts, where the is a eggplant scion and an tomato rootstock:</p>
<ul>
<li>1: <code>heterograft_1.fq</code>
</li>
<li>2: <code>heterograft_2.fq</code>
</li>
<li>3: <code>heterograft_3.fq</code>
</li>
</ul>
<p>and for the eggplant self-grafts:</p>
<ul>
<li>4: <code>selfgraft_1.fq</code>
</li>
<li>5: <code>selfgraft_2.fq</code>
</li>
<li>6: <code>selfgraft_3.fq</code> <br>
</li>
</ul>
</div>
<div class="section level4">
<h4 id="the-reference-genomes-and-annotation-files--fasta-fa">The reference genomes and annotation files (.fasta/.fa)<a class="anchor" aria-label="anchor" href="#the-reference-genomes-and-annotation-files--fasta-fa"></a>
</h4>
<p>The pipeline is designed to analyse grafting systems with two distinct genomes, here tomato and eggplant. <br></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="pre-processing">Pre-Processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h2>
<p>The pre-processing step involves cleaning raw data and aligning data to the merged genome. Going forward, the pipeline has assumed that the sRNA-seq samples have met quality control standards.</p>
<p>We recommend installing and using the <code>ShortStack</code> (<a href="https://github.com/MikeAxtell/ShortStack" class="external-link uri">https://github.com/MikeAxtell/ShortStack</a>) program to aligns samples to the merged genome and undertake cluster analysis. This program is specifically formulated for sRNAseq analysis, utilising Bowtie (Version 1) to map samples and a specifically formulated algorithm to cluster sRNAs.</p>
<p>Here, we introduce an alternative mapping method for the analysis of plant heterograft samples. The heterograft system involves two genotypes; here the two genome references are merged into a single reference to which samples are aligned to.</p>
<p><code>mobileRNA</code> offers a function to merge two FASTA reference genomes into one. To distinguish between the reference genomes in a merged file, it is important to make sure the chromosome names between the genomes are different and distinguishable. The function below added a particular character string to the start of each chromosome name in each reference genome. As standard, the string <code>A_</code> is added to the reference genome supplied to <code>genomeA</code> and <code>B_</code> is added to the reference genome supplied to <code>genomeB.</code> These can be customized to the users preference, see manual for more information. <br></p>
<div class="section level3">
<h3 id="pre-mapping">Pre-mapping<a class="anchor" aria-label="anchor" href="#pre-mapping"></a>
</h3>
<div class="section level4">
<h4 id="merging-genome-assemblies">Merging Genome Assemblies<a class="anchor" aria-label="anchor" href="#merging-genome-assemblies"></a>
</h4>
<p>Here we merge the two reference genomes into a single merged genome using the [mobileRNA::RNAgenomeMerge()] function. As default, this function changes the chromosome names of each genome to ensure they are distinguishable. To do so, the function requires an input of the initials of each organism’s Latin name. But why do this:</p>
<ul>
<li><p>If the two genomes use the same pattern to name the chromosomes, the user will not be able to differentiate the chromosomes from one another in the merged genome. This could be solves by adding letters to the chromosomes of one of the genomes, for example, <code>SM</code> to represent the Latin name of eggplant.</p></li>
<li><p>If a chromosome naming pattern contains punctuation, the mapping step will not work.</p></li>
</ul>
<p>In the example, the <em>Solanum lycopersicum</em> (version 4) genome contains a full-stop/period within each chromosome name which needs to be removed as well. Here we rename the chromosomes of the tomato genome to <code>SL</code> and the chromosomes related to the eggplant genome (<em>Solanum melongena</em>) to <code>SM</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/RNAmergeGenomes.html">RNAmergeGenomes</a></span><span class="op">(</span>genomeA <span class="op">=</span><span class="st">"/Users/user1/projectname/workplace/reference/ref1.fa"</span>,</span>
<span>               genomeB <span class="op">=</span><span class="st">"/Users/user1/projectname/workplace/reference/ref2.fa"</span>,</span>
<span>               abbreviationGenomeA <span class="op">=</span> <span class="st">"SL"</span>,</span>
<span>               abbreviationGenomeB <span class="op">=</span> <span class="st">"SM"</span>,</span>
<span>               out_dir <span class="op">=</span> <span class="st">"/Users/user1/projectname/workplace/reference/merge/</span></span>
<span><span class="st">               ref_merged.fa"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="merging-genome-annotations">Merging Genome Annotations<a class="anchor" aria-label="anchor" href="#merging-genome-annotations"></a>
</h3>
<p>To identify associates between sRNA clusters and genomic features, the genomic annotation files (GFF) of each reference genome must be merged in the same way the reference genomes were. This includes following the same naming patterns for the chromosomes in each genome as undertaken earlier with the <code><a href="../reference/RNAmergeGenomes.html">RNAmergeGenomes()</a></code> function. To undertake a similar merging between the two GFF files, use the <code>RNAmergeAnnotations</code> function.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co">## merge the annotation files into a single annotation file</span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="kw">RNAmergeAnnotation</span>(<span class="dt">annotationA =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="st">                   annotation/annotation_1.gff3"</span>,</span>
<span id="cb5-4"><a href="#cb5-4"></a>                   <span class="dt">annotationB =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="st">                   annotation/annotation_1.gff3"</span>,</span>
<span id="cb5-6"><a href="#cb5-6"></a>                   <span class="dt">abbreviationAnnoA =</span> <span class="st">"SL"</span>,</span>
<span id="cb5-7"><a href="#cb5-7"></a>                   <span class="dt">abbreviationAnnoB =</span> <span class="st">"SM"</span></span>
<span id="cb5-8"><a href="#cb5-8"></a>                  <span class="dt">out_dir =</span> <span class="st">"/Users/user1/projectname/workplace/</span></span>
<span id="cb5-9"><a href="#cb5-9"></a><span class="st">                  annotation/merge/anno_merged.gff"</span>)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="auto-detection-of-srna-cluster">Auto-Detection of sRNA Cluster<a class="anchor" aria-label="anchor" href="#auto-detection-of-srna-cluster"></a>
</h3>
<p>Here we identify and build a list sRNA clusters within each sample to assist the mapping step later on to ensure consistency across the analysis. This step is very important to consider in the analysis as it prevents false assumptions. This step aligns a replicate to the given genome assembly and the outputs contains loci where sRNA clusters were found. Across your replicates, there will be difference in the cluster identified. In some instance, it might be the case that the abundance of reads at a given sRNA locus was not significant enough to record as a sRNA cluster, while in another replicate it was recorded. Therefore, it would be wrong to assume there are zero reads at the loci. Here, we identify all the potential sRNA cluster loci and use this information to improve the mapping.</p>
<p>We recommend installing the <code>ShortStack</code> (<a href="https://github.com/MikeAxtell/ShortStack" class="external-link uri">https://github.com/MikeAxtell/ShortStack</a>) program to detect sRNA clusters and align sRNAseq samples. <br></p>
<div class="section level5">
<h5 id="step-1---cluster-analysis-with-shortstack">Step 1 - Cluster analysis with ShortStack<a class="anchor" aria-label="anchor" href="#step-1---cluster-analysis-with-shortstack"></a>
</h5>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb6-2"><a href="#cb6-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb6-3"><a href="#cb6-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb6-4"><a href="#cb6-4"></a>--bowtie_cores 6 \</span>
<span id="cb6-5"><a href="#cb6-5"></a>--mmap n \</span>
<span id="cb6-6"><a href="#cb6-6"></a>--mismatches 0 \</span>
<span id="cb6-7"><a href="#cb6-7"></a>--nohp \</span>
<span id="cb6-8"><a href="#cb6-8"></a>--outdir <span class="op">&lt;</span>./output/directory<span class="op">&gt;</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="step-2---build-srna-cluster-list">Step 2 - Build sRNA cluster list<a class="anchor" aria-label="anchor" href="#step-2---build-srna-cluster-list"></a>
</h5>
<p>Now, we collate all the sRNA loci information from each sample into a text file.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># location of step 1 output</span></span>
<span id="cb7-2"><a href="#cb7-2"></a>folder &lt;-<span class="st"> </span><span class="er">&lt;</span>.<span class="op">/</span>output<span class="op">/</span>directory<span class="op">/</span>from<span class="op">/</span>step<span class="op">/</span><span class="dv">1</span><span class="op">/</span><span class="er">&gt;</span></span>
<span id="cb7-3"><a href="#cb7-3"></a></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co"># name and location to save output file to (must be .txt)</span></span>
<span id="cb7-5"><a href="#cb7-5"></a>save_folder &lt;-<span class="st"> </span><span class="er">&lt;</span>.<span class="op">/</span>output<span class="op">/</span>directory<span class="op">/</span>ClustersInfo.txt<span class="op">&gt;</span></span>
<span id="cb7-6"><a href="#cb7-6"></a></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co"># names of samples (ie. folder names)</span></span>
<span id="cb7-8"><a href="#cb7-8"></a>sample_names &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"&lt;treatment_1&gt;"</span>, <span class="st">"&lt;treatment_2&gt;"</span>, <span class="st">"&lt;control_1&gt;"</span>,<span class="st">"&lt;control_2&gt;"</span>)</span>
<span id="cb7-9"><a href="#cb7-9"></a></span>
<span id="cb7-10"><a href="#cb7-10"></a></span>
<span id="cb7-11"><a href="#cb7-11"></a>loci_info &lt;-<span class="st"> </span><span class="kw">RNAloci</span>(<span class="dt">files =</span> folder, </span>
<span id="cb7-12"><a href="#cb7-12"></a>             <span class="dt">out =</span> save_folder,</span>
<span id="cb7-13"><a href="#cb7-13"></a>             <span class="dt">samples =</span> sample_names)</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="mapping">Mapping<a class="anchor" aria-label="anchor" href="#mapping"></a>
</h3>
<p>Each sample is mapped to the merged reference genome with the list of sRNA clusters.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1"></a><span class="ex">ShortStack</span> \</span>
<span id="cb8-2"><a href="#cb8-2"></a>--readfile <span class="op">&lt;</span>control_1.fastq<span class="op">&gt;</span> \</span>
<span id="cb8-3"><a href="#cb8-3"></a>--genomefile <span class="op">&lt;</span>merged_reference.fa<span class="op">&gt;</span> \</span>
<span id="cb8-4"><a href="#cb8-4"></a>--locifile <span class="op">&lt;</span>./output/directory/ClustersInfo.gff3<span class="op">&gt;</span> \</span>
<span id="cb8-5"><a href="#cb8-5"></a>--bowtie_cores 6 \</span>
<span id="cb8-6"><a href="#cb8-6"></a>--mmap n \</span>
<span id="cb8-7"><a href="#cb8-7"></a>--mismatches 0 \</span>
<span id="cb8-8"><a href="#cb8-8"></a>--nohp \</span>
<span id="cb8-9"><a href="#cb8-9"></a>--mincov 5 \</span>
<span id="cb8-10"><a href="#cb8-10"></a>--outdir <span class="op">&lt;</span>./output/directory/step2/<span class="op">&gt;</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>Here, the analysis of the pre-processed data occurs, with the aim to identify if any mobile sRNA molecules are traveling from one genotype to the other in your grafting system. <br></p>
<div class="section level3">
<h3 id="import-data">Import Data<a class="anchor" aria-label="anchor" href="#import-data"></a>
</h3>
<p>In the pre-processing steps, the data was cleaned and aligned to the merged reference genome. During the mapping step, a folder for each sample is created which stores all the results in. The analysis steps requires the information which is stored in the <code>Results.txt</code>.</p>
<p>The <code><a href="../reference/RNAimport.html">RNAimport()</a></code> function imports the data from the folder containing all sample folders. The function extracts the required information, stores it in a matrix.</p>
<p>The function requires some information to coordinate the importation and calculations. It requires a directory path to your processed samples; this is the path to the folder containing all the individual sample folders which is stores in argument <code>directory</code>. The second directory path is to the loci file containing information on the loci of the sRNA clusters across the analysis, stored in the argument <code>loci</code>. The last pieces of information the function requires a vector containing the sample names, both treatment and control replicates which is stored in the argument <code>samples</code>. The sample names must match the names of the folders produced in the mapping, and stored in the directory supplied to the argument <code>directory</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Import &amp; organise data.</span></span>
<span><span class="va">results_dir</span> <span class="op">&lt;-</span>  <span class="st">"./analysis/alignment_unique_two/"</span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>,<span class="st">"heterograft_2"</span>, <span class="st">"heterograft_3"</span>,</span>
<span>                  <span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span>, <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">sRNA_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAimport.html">RNAimport</a></span><span class="op">(</span>input <span class="op">=</span> <span class="st">"sRNA"</span>,</span>
<span>                       directory <span class="op">=</span> <span class="va">results_dir</span>,</span>
<span>                           samples <span class="op">=</span> <span class="va">sample_names</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># example dataframe </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sRNA_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   Locus chr     start       end   Cluster</span></span>
<span><span class="co">#&gt; 1 1:100000956-100001076   1 100000956 100001076 cluster_1</span></span>
<span><span class="co">#&gt; 2 1:100001486-100001554   1 100001486 100001554 cluster_2</span></span>
<span><span class="co">#&gt; 3 1:100001646-100001731   1 100001646 100001731 cluster_3</span></span>
<span><span class="co">#&gt; 4   1:10001034-10001073   1  10001034  10001073 cluster_4</span></span>
<span><span class="co">#&gt; 5 1:100011795-100011930   1 100011795 100011930 cluster_5</span></span>
<span><span class="co">#&gt; 6   1:10002000-10002065   1  10002000  10002065 cluster_6</span></span>
<span><span class="co">#&gt;   DicerCall_heterograft_1 Count_heterograft_1 RPM_heterograft_1</span></span>
<span><span class="co">#&gt; 1                      24                  19             1.908</span></span>
<span><span class="co">#&gt; 2                      24                   3             0.301</span></span>
<span><span class="co">#&gt; 3                       N                   3             0.301</span></span>
<span><span class="co">#&gt; 4                      24                   1             0.100</span></span>
<span><span class="co">#&gt; 5                      24                  11             1.105</span></span>
<span><span class="co">#&gt; 6                      24                   9             0.904</span></span>
<span><span class="co">#&gt;      MajorRNA_heterograft_1 DicerCall_heterograft_2 Count_heterograft_2</span></span>
<span><span class="co">#&gt; 1  AGUGACGGUCUCAUGUAACAUAGA                      24                  14</span></span>
<span><span class="co">#&gt; 2  AAAGAGUAGACAUAUAUUAAAUGG                      20                   5</span></span>
<span><span class="co">#&gt; 3 CUGUCUCAUGUGACACUCUUUCACU                      23                   6</span></span>
<span><span class="co">#&gt; 4  AAAUUGUCAUUUAGUCAGAGGAAU                       N                   0</span></span>
<span><span class="co">#&gt; 5  AUAGACAAAUCGCACUCGCUCUCG                      24                   2</span></span>
<span><span class="co">#&gt; 6  UUAGAUUUCUGGAGUAACAACUUU                       N                   2</span></span>
<span><span class="co">#&gt;   RPM_heterograft_2   MajorRNA_heterograft_2 DicerCall_heterograft_3</span></span>
<span><span class="co">#&gt; 1             1.803 AGUGACGGUCUCAUGUAACAUAGA                      24</span></span>
<span><span class="co">#&gt; 2             0.644     ACAUAUUUUUUUGGACGAAC                      24</span></span>
<span><span class="co">#&gt; 3             0.773  UCUGUCUCAUGUGACACUCUUUC                      24</span></span>
<span><span class="co">#&gt; 4             0.000                        N                      24</span></span>
<span><span class="co">#&gt; 5             0.258 AAGUGUCUACCCAUUGUGAUUCGC                      24</span></span>
<span><span class="co">#&gt; 6             0.258      UUAUUGACAAACUAUAACU                      22</span></span>
<span><span class="co">#&gt;   Count_heterograft_3 RPM_heterograft_3   MajorRNA_heterograft_3</span></span>
<span><span class="co">#&gt; 1                  30             2.953 AGUGACGGUCUCAUGUAACAUAGA</span></span>
<span><span class="co">#&gt; 2                   1             0.098 UAUGUCUACUCUUUAUUACUCUCC</span></span>
<span><span class="co">#&gt; 3                  11             1.083 AUCUUGUCAAGUGAAAGAGUGUCA</span></span>
<span><span class="co">#&gt; 4                  11             1.083 AUUCCUCUGACUAAAUGACAAUUU</span></span>
<span><span class="co">#&gt; 5                   5             0.492 AAUCACAAUGGGUAGACACUUAGA</span></span>
<span><span class="co">#&gt; 6                   3             0.295 UUAGAUUUCUGGAGUAACAACUUU</span></span>
<span><span class="co">#&gt;   DicerCall_selfgraft_1 Count_selfgraft_1 RPM_selfgraft_1</span></span>
<span><span class="co">#&gt; 1                    24                19           1.909</span></span>
<span><span class="co">#&gt; 2                    24                 3           0.301</span></span>
<span><span class="co">#&gt; 3                     N                 3           0.301</span></span>
<span><span class="co">#&gt; 4                    24                 1           0.100</span></span>
<span><span class="co">#&gt; 5                    24                11           1.105</span></span>
<span><span class="co">#&gt; 6                    24                 9           0.904</span></span>
<span><span class="co">#&gt;        MajorRNA_selfgraft_1 DicerCall_selfgraft_2 Count_selfgraft_2</span></span>
<span><span class="co">#&gt; 1  AGUGACGGUCUCAUGUAACAUAGA                    24                14</span></span>
<span><span class="co">#&gt; 2  AAAGAGUAGACAUAUAUUAAAUGG                    20                 5</span></span>
<span><span class="co">#&gt; 3 CUGUCUCAUGUGACACUCUUUCACU                    23                 6</span></span>
<span><span class="co">#&gt; 4  AAAUUGUCAUUUAGUCAGAGGAAU                     N                 0</span></span>
<span><span class="co">#&gt; 5  AAUGGGUAGACACUUAGAUACAUG                    24                 2</span></span>
<span><span class="co">#&gt; 6  UUAGAUUUCUGGAGUAACAACUUU                     N                 2</span></span>
<span><span class="co">#&gt;   RPM_selfgraft_2     MajorRNA_selfgraft_2 DicerCall_selfgraft_3</span></span>
<span><span class="co">#&gt; 1           1.804 AGUGACGGUCUCAUGUAACAUAGA                    24</span></span>
<span><span class="co">#&gt; 2           0.644     ACAUAUUUUUUUGGACGAAC                    24</span></span>
<span><span class="co">#&gt; 3           0.773  UCUGUCUCAUGUGACACUCUUUC                    24</span></span>
<span><span class="co">#&gt; 4           0.000                        N                    24</span></span>
<span><span class="co">#&gt; 5           0.258 AUAGACAAAUCGCACUCGCUCUCG                    24</span></span>
<span><span class="co">#&gt; 6           0.258  AGAUUUCUGGAGUAACAACUUUG                    22</span></span>
<span><span class="co">#&gt;   Count_selfgraft_3 RPM_selfgraft_3     MajorRNA_selfgraft_3</span></span>
<span><span class="co">#&gt; 1                30           2.954 AGUGACGGUCUCAUGUAACAUAGA</span></span>
<span><span class="co">#&gt; 2                 1           0.098 UAUGUCUACUCUUUAUUACUCUCC</span></span>
<span><span class="co">#&gt; 3                11           1.083 AUCUUGUCAAGUGAAAGAGUGUCA</span></span>
<span><span class="co">#&gt; 4                11           1.083 AUUCCUCUGACUAAAUGACAAUUU</span></span>
<span><span class="co">#&gt; 5                 5           0.492 AAUCACAAUGGGUAGACACUUAGA</span></span>
<span><span class="co">#&gt; 6                 3           0.295   UUGACAAACUAUAACUUAAAUU</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="sample-level-quality-control">Sample-level quality control<a class="anchor" aria-label="anchor" href="#sample-level-quality-control"></a>
</h3>
<p>A handy step in the analysis is to assess the overall similarity between sample replicates to understand which samples are the most similar and which are different, and where the most variation is introduced in the data set. As well as understanding whether the data set meets your expectations. It is expected that between the conditions, the sample replicates show enough variation to suggest that the replicates are from different groups.</p>
<p>To investigate the sample similarity/difference, we will undertake sample-level quality control using three different methods:</p>
<ul>
<li>Distribution of RNA classes within each sample</li>
<li>Principle component analysis (PCA)</li>
<li>hierarchical clustering Heatmap</li>
</ul>
<p>This will show us how well samples within each condition cluster together, which may highlight outliers. Plus, to show whether our experimental conditions represent the main source of variation in the data set.</p>
<p>Here we will be employing an unsupervised clustering methods for the PCA and hierarchical clustering Heatmap. This involves an unbiased log2-transformation of the counts which will emphasis the the sample clustering to improve visualization. The DESeq2 package contains a particularly useful function to undertake regularized log transformation (rlog) which controls the variance across the mean, and in this package we have utilized this for the quality control steps.</p>
<p><br></p>
<div class="section level4">
<h4 id="plot-the-distribution-of-rna-classes-within-each-sample">Plot the distribution of RNA classes within each sample<a class="anchor" aria-label="anchor" href="#plot-the-distribution-of-rna-classes-within-each-sample"></a>
</h4>
<p>Explore the number of each RNA class identified within each sample using the <code>RNAdistribution</code> function which produces a table and plots the results. The results can be plotted as either a bar chart (<code>style = "bar"</code>) or a line graph (<code>style = "line"</code>) for each sample and can be shown in a bar chart facet (<code>facet = TRUE</code>) or in as a single line graph, where each line represents a sample (<code>together=TRUE</code>). Alternatively, the results can be plotted individually for each sample in either format.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># plot each replicate as a line, separately, and facet </span></span>
<span><span class="va">sample_distribution_line</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span><span class="va">sRNA_data</span>,</span>
<span>                                            style <span class="op">=</span> <span class="st">"line"</span>,</span>
<span>                                            together <span class="op">=</span> <span class="cn">FALSE</span>,facet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>                                            </span>
<span> <span class="co"># plot each replicate as a bar, separately, and facet </span></span>
<span><span class="va">sample_distribution_bar</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span><span class="va">sRNA_data</span>,</span>
<span>                                           style <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>                                           facet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot each replicate on a single plot</span></span>
<span><span class="va">sample_distribution</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span><span class="va">sRNA_data</span>,</span>
<span>                                           style <span class="op">=</span> <span class="st">"line"</span>,</span>
<span>                                           together <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Lets view these plots:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># View plot (only)</span></span>
<span><span class="va">sample_distribution_line</span><span class="op">$</span><span class="va">plot</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># View plot (only)</span></span>
<span><span class="va">sample_distribution_bar</span><span class="op">$</span><span class="va">plot</span></span>
<span></span>
<span><span class="co"># View plot</span></span>
<span><span class="va">sample_distribution</span><span class="op">$</span><span class="va">plot</span></span>
<span><span class="co"># View data </span></span>
<span><span class="va">sample_distribution</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="co">#&gt;    Class heterograft_1 heterograft_2 heterograft_3 selfgraft_1 selfgraft_2</span></span>
<span><span class="co">#&gt; 1:    20             9            14             7           8          13</span></span>
<span><span class="co">#&gt; 2:    21            27            28            24          25          26</span></span>
<span><span class="co">#&gt; 3:    22            19            27            26          19          27</span></span>
<span><span class="co">#&gt; 4:    23            25            38            42          18          31</span></span>
<span><span class="co">#&gt; 5:    24           597           491           575         459         353</span></span>
<span><span class="co">#&gt;    selfgraft_3</span></span>
<span><span class="co">#&gt; 1:           6</span></span>
<span><span class="co">#&gt; 2:          22</span></span>
<span><span class="co">#&gt; 3:          26</span></span>
<span><span class="co">#&gt; 4:          35</span></span>
<span><span class="co">#&gt; 5:         437</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-11-1.png" alt="An example line graph (A) and bar graph (B), to show the distribution of
RNA classes within each sample. Figure C shows all replicates plotted on a 
single plot and the accompanying data." width="33%"><img src="mobileRNA_files/figure-html/unnamed-chunk-11-2.png" alt="An example line graph (A) and bar graph (B), to show the distribution of
RNA classes within each sample. Figure C shows all replicates plotted on a 
single plot and the accompanying data." width="33%"><img src="mobileRNA_files/figure-html/unnamed-chunk-11-3.png" alt="An example line graph (A) and bar graph (B), to show the distribution of
RNA classes within each sample. Figure C shows all replicates plotted on a 
single plot and the accompanying data." width="33%"><p class="caption">
An example line graph (A) and bar graph (B), to show the distribution of RNA classes within each sample. Figure C shows all replicates plotted on a single plot and the accompanying data.
</p>
</div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="principle-component-analysis-to-assess-sample-distance">Principle component analysis to assess sample distance<a class="anchor" aria-label="anchor" href="#principle-component-analysis-to-assess-sample-distance"></a>
</h4>
<p>Principal Component Analysis (PCA) is a useful technique to illustrate sample distance as it emphasizes the variation through the reduction of dimensions in the data set. Here, we introduce the function <code><a href="../reference/plotSamplePCA.html">plotSamplePCA()</a></code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotSamplePCA.html">plotSamplePCA</a></span><span class="op">(</span><span class="va">sRNA_data</span>, <span class="va">group</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking and organising data </span></span>
<span><span class="co">#&gt; Transforming the count data to the log2 scale </span></span>
<span><span class="co">#&gt; Organising principal component analysis</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-13-1.png" alt="An example of a PCA, illustracting the sRNA data set sample similarity" width="700"><p class="caption">
An example of a PCA, illustracting the sRNA data set sample similarity
</p>
</div>
<p>This PCA is somewhat not what we would expect to see in a real dataset. This is because PC1 is showing the variation between the non-spiked and spiked version of the replicate while PC2 shows the natural variation between the replicates. In a real dataset, we would expect there to be little difference between the treatment and control. There should be much more natural variation between the replicates, but it is unlikely that there will be a huge distinctive variation between the samples across the conditions. This is because it is likely that only a very small number of endogenous mobile sRNAs are present in the heterograft system compared to the self-graft controls while the population of sRNAs present in the destination tissue is expected to be largely similar. Although, there is likely to be some difference in this too due to the endogenous changes. <br></p>
</div>
<div class="section level4">
<h4 id="hierarchical-clustered-heatmap-to-assess-sample-distance">Hierarchical clustered heatmap to assess sample distance<a class="anchor" aria-label="anchor" href="#hierarchical-clustered-heatmap-to-assess-sample-distance"></a>
</h4>
<p>Similarly, to a PCA plot, the <code><a href="../reference/plotSampleDistance.html">plotSampleDistance()</a></code> function undertakes hierarchical clustering with an unbiased log transformation to calculate sample distance and is plotted in the form of a heatmap.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotSampleDistance.html">plotSampleDistance</a></span><span class="op">(</span><span class="va">sRNA_data</span><span class="op">)</span></span>
<span><span class="co">#&gt; Checking data </span></span>
<span><span class="co">#&gt; Transforming the count data to the log2 scale </span></span>
<span><span class="co">#&gt; Calculating distance matrix </span></span>
<span><span class="co">#&gt; Creating sample distance plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-15-1.png" alt="An example of a heatmap,illustracting the sRNA data set sample similarity" width="700"><p class="caption">
An example of a heatmap,illustracting the sRNA data set sample similarity
</p>
</div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="define-the-consensus-rna-class">Define the consensus RNA class<a class="anchor" aria-label="anchor" href="#define-the-consensus-rna-class"></a>
</h3>
<p>Have a look at the <code>sRNA_data</code> object, you will see that for each sample the class of sRNA has been determined (see columns with names starting with DicerCall). The result will either state a number from 20-24 which represents the nucleotide length of the sRNA, or, state NA/N which represents that the sRNA was unclassified in the sample. There is usually some discrepancies between the classification of a sRNA, therefore, a consensus class can be assigned based off the information from the treatment samples. The <code><a href="../reference/RNAdicercall.html">RNAdicercall()</a></code> function is used to compute this by counting the number of occurrences of classifications for a sRNA. The consensus is determined based on the class with the most counts.</p>
<p>When working with samples from a chimeric system, the function contains a specialised parameter which should be utilized. This is <code>chimeric=TRUE</code>, along with the <code>id</code> and <code>controls</code> parameters. This helps optimise classification by removal of mapping errors.</p>
<p>Moreover, the user would prefer to form the dicer consensus based on the results of specific samples this can be specified to the <code>conditions</code> parameter. The function also works to remove noise from your data, which as default is not undertaken. If the user wishes to remove the noise, set <code>tidy=TRUE</code> to promote the removal of unclassified clusters.</p>
<p>For the downstream analysis, it can be useful to define the distinct 24-nt and 21/22-nt populations of sRNA molecules at the beginning of the analysis. Hence, the data should be subset. To subset the data, use the <code><a href="../reference/RNAsubset.html">RNAsubset()</a></code> function to choose which sRNA populations to include: 20-nt, 21-nt, 22-nt, 23-nt, or 24-nt. It is recommended to subset the data into two dataframes: one which includes only 24-nt sRNA and the other to include both 21-nt and 22-nt sRNAs. To instruct the <code><a href="../reference/RNAsubset.html">RNAsubset()</a></code> function to select specific groups, use the <code>type</code> argument.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># define consensus, store as a data summary file.</span></span>
<span><span class="va">sRNA_data_consensus2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdicercall.html">RNAdicercall</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data</span>, chimeric <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                     genome.ID <span class="op">=</span> <span class="st">"SL"</span>, </span>
<span>                                     controls <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, </span>
<span>                                                  <span class="st">"selfgraft_2"</span>, </span>
<span>                                                  <span class="st">"selfgraft_3"</span><span class="op">)</span>,</span>
<span>                                     tidy<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating consensus dicercall based on information from all replicates </span></span>
<span><span class="co">#&gt; The consensus dicercall will be excluded in the case of a tie </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Removing sRNA clusters with no consensus dicercall...</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sRNA_data_consensus</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   Locus chr     start       end   Cluster</span></span>
<span><span class="co">#&gt; 1 1:100000956-100001076   1 100000956 100001076 cluster_1</span></span>
<span><span class="co">#&gt; 2 1:100001486-100001554   1 100001486 100001554 cluster_2</span></span>
<span><span class="co">#&gt; 4   1:10001034-10001073   1  10001034  10001073 cluster_4</span></span>
<span><span class="co">#&gt; 5 1:100011795-100011930   1 100011795 100011930 cluster_5</span></span>
<span><span class="co">#&gt; 7   1:10002223-10002449   1  10002223  10002449 cluster_7</span></span>
<span><span class="co">#&gt; 8   1:10002531-10002593   1  10002531  10002593 cluster_8</span></span>
<span><span class="co">#&gt;   DicerCall_heterograft_1 Count_heterograft_1 RPM_heterograft_1</span></span>
<span><span class="co">#&gt; 1                      24                  19             1.908</span></span>
<span><span class="co">#&gt; 2                      24                   3             0.301</span></span>
<span><span class="co">#&gt; 4                      24                   1             0.100</span></span>
<span><span class="co">#&gt; 5                      24                  11             1.105</span></span>
<span><span class="co">#&gt; 7                      24                  20             2.009</span></span>
<span><span class="co">#&gt; 8                      24                  12             1.205</span></span>
<span><span class="co">#&gt;     MajorRNA_heterograft_1 DicerCall_heterograft_2 Count_heterograft_2</span></span>
<span><span class="co">#&gt; 1 AGUGACGGUCUCAUGUAACAUAGA                      24                  14</span></span>
<span><span class="co">#&gt; 2 AAAGAGUAGACAUAUAUUAAAUGG                      20                   5</span></span>
<span><span class="co">#&gt; 4 AAAUUGUCAUUUAGUCAGAGGAAU                       N                   0</span></span>
<span><span class="co">#&gt; 5 AUAGACAAAUCGCACUCGCUCUCG                      24                   2</span></span>
<span><span class="co">#&gt; 7 AACGUGAAUUCUGAUGAAACCAAU                      24                  12</span></span>
<span><span class="co">#&gt; 8 AGGGUAUAUUGGUAUUGGACCAUC                      24                   9</span></span>
<span><span class="co">#&gt;   RPM_heterograft_2   MajorRNA_heterograft_2 DicerCall_heterograft_3</span></span>
<span><span class="co">#&gt; 1             1.803 AGUGACGGUCUCAUGUAACAUAGA                      24</span></span>
<span><span class="co">#&gt; 2             0.644     ACAUAUUUUUUUGGACGAAC                      24</span></span>
<span><span class="co">#&gt; 4             0.000                        N                      24</span></span>
<span><span class="co">#&gt; 5             0.258 AAGUGUCUACCCAUUGUGAUUCGC                      24</span></span>
<span><span class="co">#&gt; 7             1.545 UUAUAACGUGAAUUCUGAUGAAAC                      24</span></span>
<span><span class="co">#&gt; 8             1.159 AUAUUGGUAUUGGACCAUCAGAUU                      24</span></span>
<span><span class="co">#&gt;   Count_heterograft_3 RPM_heterograft_3   MajorRNA_heterograft_3</span></span>
<span><span class="co">#&gt; 1                  30             2.953 AGUGACGGUCUCAUGUAACAUAGA</span></span>
<span><span class="co">#&gt; 2                   1             0.098 UAUGUCUACUCUUUAUUACUCUCC</span></span>
<span><span class="co">#&gt; 4                  11             1.083 AUUCCUCUGACUAAAUGACAAUUU</span></span>
<span><span class="co">#&gt; 5                   5             0.492 AAUCACAAUGGGUAGACACUUAGA</span></span>
<span><span class="co">#&gt; 7                  29             2.855  AUAACGUGAAUUCUGAUGAAACC</span></span>
<span><span class="co">#&gt; 8                  16             1.575 GGUAUUGGACCAUCAGAUUGACGA</span></span>
<span><span class="co">#&gt;   DicerCall_selfgraft_1 Count_selfgraft_1 RPM_selfgraft_1</span></span>
<span><span class="co">#&gt; 1                    24                19           1.909</span></span>
<span><span class="co">#&gt; 2                    24                 3           0.301</span></span>
<span><span class="co">#&gt; 4                    24                 1           0.100</span></span>
<span><span class="co">#&gt; 5                    24                11           1.105</span></span>
<span><span class="co">#&gt; 7                    24                20           2.010</span></span>
<span><span class="co">#&gt; 8                    24                12           1.206</span></span>
<span><span class="co">#&gt;       MajorRNA_selfgraft_1 DicerCall_selfgraft_2 Count_selfgraft_2</span></span>
<span><span class="co">#&gt; 1 AGUGACGGUCUCAUGUAACAUAGA                    24                14</span></span>
<span><span class="co">#&gt; 2 AAAGAGUAGACAUAUAUUAAAUGG                    20                 5</span></span>
<span><span class="co">#&gt; 4 AAAUUGUCAUUUAGUCAGAGGAAU                     N                 0</span></span>
<span><span class="co">#&gt; 5 AAUGGGUAGACACUUAGAUACAUG                    24                 2</span></span>
<span><span class="co">#&gt; 7 AACGUGAAUUCUGAUGAAACCAAU                    24                12</span></span>
<span><span class="co">#&gt; 8 AGGGUAUAUUGGUAUUGGACCAUC                    24                 9</span></span>
<span><span class="co">#&gt;   RPM_selfgraft_2     MajorRNA_selfgraft_2 DicerCall_selfgraft_3</span></span>
<span><span class="co">#&gt; 1           1.804 AGUGACGGUCUCAUGUAACAUAGA                    24</span></span>
<span><span class="co">#&gt; 2           0.644     ACAUAUUUUUUUGGACGAAC                    24</span></span>
<span><span class="co">#&gt; 4           0.000                        N                    24</span></span>
<span><span class="co">#&gt; 5           0.258 AUAGACAAAUCGCACUCGCUCUCG                    24</span></span>
<span><span class="co">#&gt; 7           1.546 UUAUAACGUGAAUUCUGAUGAAAC                    24</span></span>
<span><span class="co">#&gt; 8           1.160 AUAUUGGUAUUGGACCAUCAGAUU                    24</span></span>
<span><span class="co">#&gt;   Count_selfgraft_3 RPM_selfgraft_3     MajorRNA_selfgraft_3 DicerCounts</span></span>
<span><span class="co">#&gt; 1                30           2.954 AGUGACGGUCUCAUGUAACAUAGA           3</span></span>
<span><span class="co">#&gt; 2                 1           0.098 UAUGUCUACUCUUUAUUACUCUCC           3</span></span>
<span><span class="co">#&gt; 4                11           1.083 AUUCCUCUGACUAAAUGACAAUUU           2</span></span>
<span><span class="co">#&gt; 5                 5           0.492 AAUCACAAUGGGUAGACACUUAGA           3</span></span>
<span><span class="co">#&gt; 7                29           2.856  AUAACGUGAAUUCUGAUGAAACC           3</span></span>
<span><span class="co">#&gt; 8                16           1.576 GGUAUUGGACCAUCAGAUUGACGA           3</span></span>
<span><span class="co">#&gt;   DicerConsensus</span></span>
<span><span class="co">#&gt; 1             24</span></span>
<span><span class="co">#&gt; 2             24</span></span>
<span><span class="co">#&gt; 4             24</span></span>
<span><span class="co">#&gt; 5             24</span></span>
<span><span class="co">#&gt; 7             24</span></span>
<span><span class="co">#&gt; 8             24</span></span>
<span></span>
<span><span class="co"># Subset data for analysis: 24-nt sRNAs</span></span>
<span><span class="va">sRNA_24</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsubset.html">RNAsubset</a></span><span class="op">(</span><span class="va">sRNA_data_consensus</span>, type <span class="op">=</span> <span class="fl">24</span><span class="op">)</span></span>
<span><span class="co"># Subset data for analysis: 24 21/22-nt sRNAs</span></span>
<span><span class="va">sRNA_2122</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsubset.html">RNAsubset</a></span><span class="op">(</span><span class="va">sRNA_data_consensus</span>, type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">21</span>, <span class="fl">22</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="plot-consensus-classification-of-srna">Plot consensus classification of sRNA<a class="anchor" aria-label="anchor" href="#plot-consensus-classification-of-srna"></a>
</h4>
<p>Use <code>RNAdistriution()</code> to visualize the distribution of the consensus sRNA classes within the data. This step can be used as a quality control step to illustrate the distribution of each sRNA class within the data. The function can plot the absolute or relative frequency, use argument <code>relative</code>. It can be beneficial to view the absolute frequency as well as the relative to observe whether the data meets your expectations.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAdistribution.html">RNAdistribution</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>,</span>
<span>                                  consensus <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                  style <span class="op">=</span> <span class="st">"consensus"</span><span class="op">)</span></span>
<span><span class="co"># view </span></span>
<span><span class="va">consensus_plot</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="mobileRNA_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus_plot</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="co">#&gt;   DicerConsensus   n</span></span>
<span><span class="co">#&gt; 1             20   5</span></span>
<span><span class="co">#&gt; 2             21  20</span></span>
<span><span class="co">#&gt; 3             22  14</span></span>
<span><span class="co">#&gt; 4             23  19</span></span>
<span><span class="co">#&gt; 5             24 601</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis-with-deseq2-or-edger">Differential Expression analysis with DESeq2 or edgeR<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-with-deseq2-or-edger"></a>
</h3>
<p>Differential expression (DE) analysis is undertaken to identify sRNA which are statistically significant to discover quantitative changes in the expression levels between the treatment (hetero-grafting) and the control (self-grafting) groups. This technique can be undertaken with a variety of tools, in <code>mobileRNA</code> users have the option to use the <code>DESeq2</code> or <code>edgeR</code> analytical method.</p>
<p>Although both methods will produce different results, it is typically they case that that at the step where we isolate the mobile molecules, the same results are produce. What is meant by this, is that the same mobile molecules are typically identified at the end of the analysis regardless of the analytical method. However, in particular case, one method will be preferred over the other. For instance, the <code>DESeq2</code> method is not appropriate when the experiment does not contain replicate (ie. one sample replicate per condition). On the other hand, edgeR can be used. Here, we have included the recommend practice for edgeR when the data set does not contain replicates. This option can be employed by setting a custom dispersion value, see argument <code>dispersionValue</code>.</p>
<p>Please note that this step is optional and not a required part of the analysis.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># sample conditions in order within dataframe</span></span>
<span><span class="va">groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>, <span class="st">"Heterograft"</span>,</span>
<span>            <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span>, <span class="st">"Selfgraft"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Differential analysis of whole dataset: DEseq2 method </span></span>
<span><span class="va">sRNA_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"DESeq2"</span> <span class="op">)</span></span>
<span></span>
<span><span class="co">## Differential analysis: DEseq2 method</span></span>
<span><span class="co"># 24-nt sRNA data-set</span></span>
<span><span class="va">sRNA_24_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_24</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"DESeq2"</span> <span class="op">)</span></span>
<span><span class="co"># 2122-nt sRNA data-set</span></span>
<span><span class="va">sRNA_2122_DESeq2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_2122</span>,</span>
<span>                                group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                                method <span class="op">=</span> <span class="st">"DESeq2"</span> <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Differential analysis of whole dataset: edgeR method </span></span>
<span><span class="va">sRNA_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>,</span>
<span>                              group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                              method <span class="op">=</span> <span class="st">"edgeR"</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Design matrix not provided. Switch to the classic mode.</span></span>
<span></span>
<span></span>
<span><span class="co">## Differential analysis: edgeR method</span></span>
<span><span class="va">sRNA_24_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_24</span>,</span>
<span>                             group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                             method <span class="op">=</span> <span class="st">"edgeR"</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Design matrix not provided. Switch to the classic mode.</span></span>
<span></span>
<span><span class="va">sRNA_2122_edgeR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAanalysis.html">RNAanalysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_2122</span> ,</span>
<span>                               group <span class="op">=</span> <span class="va">groups</span>,</span>
<span>                               method <span class="op">=</span> <span class="st">"edgeR"</span> <span class="op">)</span></span>
<span><span class="co">#&gt; Design matrix not provided. Switch to the classic mode.</span></span>
<span></span>
<span><span class="co"># view example output:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sRNA_edgeR</span><span class="op">)</span></span>
<span><span class="co">#&gt;                   Locus chr     start       end   Cluster</span></span>
<span><span class="co">#&gt; 1 1:100000956-100001076   1 100000956 100001076 cluster_1</span></span>
<span><span class="co">#&gt; 2 1:100001486-100001554   1 100001486 100001554 cluster_2</span></span>
<span><span class="co">#&gt; 4   1:10001034-10001073   1  10001034  10001073 cluster_4</span></span>
<span><span class="co">#&gt; 5 1:100011795-100011930   1 100011795 100011930 cluster_5</span></span>
<span><span class="co">#&gt; 7   1:10002223-10002449   1  10002223  10002449 cluster_7</span></span>
<span><span class="co">#&gt; 8   1:10002531-10002593   1  10002531  10002593 cluster_8</span></span>
<span><span class="co">#&gt;   DicerCall_heterograft_1 Count_heterograft_1 RPM_heterograft_1</span></span>
<span><span class="co">#&gt; 1                      24                  19             1.908</span></span>
<span><span class="co">#&gt; 2                      24                   3             0.301</span></span>
<span><span class="co">#&gt; 4                      24                   1             0.100</span></span>
<span><span class="co">#&gt; 5                      24                  11             1.105</span></span>
<span><span class="co">#&gt; 7                      24                  20             2.009</span></span>
<span><span class="co">#&gt; 8                      24                  12             1.205</span></span>
<span><span class="co">#&gt;     MajorRNA_heterograft_1 DicerCall_heterograft_2 Count_heterograft_2</span></span>
<span><span class="co">#&gt; 1 AGUGACGGUCUCAUGUAACAUAGA                      24                  14</span></span>
<span><span class="co">#&gt; 2 AAAGAGUAGACAUAUAUUAAAUGG                      20                   5</span></span>
<span><span class="co">#&gt; 4 AAAUUGUCAUUUAGUCAGAGGAAU                       N                   0</span></span>
<span><span class="co">#&gt; 5 AUAGACAAAUCGCACUCGCUCUCG                      24                   2</span></span>
<span><span class="co">#&gt; 7 AACGUGAAUUCUGAUGAAACCAAU                      24                  12</span></span>
<span><span class="co">#&gt; 8 AGGGUAUAUUGGUAUUGGACCAUC                      24                   9</span></span>
<span><span class="co">#&gt;   RPM_heterograft_2   MajorRNA_heterograft_2 DicerCall_heterograft_3</span></span>
<span><span class="co">#&gt; 1             1.803 AGUGACGGUCUCAUGUAACAUAGA                      24</span></span>
<span><span class="co">#&gt; 2             0.644     ACAUAUUUUUUUGGACGAAC                      24</span></span>
<span><span class="co">#&gt; 4             0.000                        N                      24</span></span>
<span><span class="co">#&gt; 5             0.258 AAGUGUCUACCCAUUGUGAUUCGC                      24</span></span>
<span><span class="co">#&gt; 7             1.545 UUAUAACGUGAAUUCUGAUGAAAC                      24</span></span>
<span><span class="co">#&gt; 8             1.159 AUAUUGGUAUUGGACCAUCAGAUU                      24</span></span>
<span><span class="co">#&gt;   Count_heterograft_3 RPM_heterograft_3   MajorRNA_heterograft_3</span></span>
<span><span class="co">#&gt; 1                  30             2.953 AGUGACGGUCUCAUGUAACAUAGA</span></span>
<span><span class="co">#&gt; 2                   1             0.098 UAUGUCUACUCUUUAUUACUCUCC</span></span>
<span><span class="co">#&gt; 4                  11             1.083 AUUCCUCUGACUAAAUGACAAUUU</span></span>
<span><span class="co">#&gt; 5                   5             0.492 AAUCACAAUGGGUAGACACUUAGA</span></span>
<span><span class="co">#&gt; 7                  29             2.855  AUAACGUGAAUUCUGAUGAAACC</span></span>
<span><span class="co">#&gt; 8                  16             1.575 GGUAUUGGACCAUCAGAUUGACGA</span></span>
<span><span class="co">#&gt;   DicerCall_selfgraft_1 Count_selfgraft_1 RPM_selfgraft_1</span></span>
<span><span class="co">#&gt; 1                    24                19           1.909</span></span>
<span><span class="co">#&gt; 2                    24                 3           0.301</span></span>
<span><span class="co">#&gt; 4                    24                 1           0.100</span></span>
<span><span class="co">#&gt; 5                    24                11           1.105</span></span>
<span><span class="co">#&gt; 7                    24                20           2.010</span></span>
<span><span class="co">#&gt; 8                    24                12           1.206</span></span>
<span><span class="co">#&gt;       MajorRNA_selfgraft_1 DicerCall_selfgraft_2 Count_selfgraft_2</span></span>
<span><span class="co">#&gt; 1 AGUGACGGUCUCAUGUAACAUAGA                    24                14</span></span>
<span><span class="co">#&gt; 2 AAAGAGUAGACAUAUAUUAAAUGG                    20                 5</span></span>
<span><span class="co">#&gt; 4 AAAUUGUCAUUUAGUCAGAGGAAU                     N                 0</span></span>
<span><span class="co">#&gt; 5 AAUGGGUAGACACUUAGAUACAUG                    24                 2</span></span>
<span><span class="co">#&gt; 7 AACGUGAAUUCUGAUGAAACCAAU                    24                12</span></span>
<span><span class="co">#&gt; 8 AGGGUAUAUUGGUAUUGGACCAUC                    24                 9</span></span>
<span><span class="co">#&gt;   RPM_selfgraft_2     MajorRNA_selfgraft_2 DicerCall_selfgraft_3</span></span>
<span><span class="co">#&gt; 1           1.804 AGUGACGGUCUCAUGUAACAUAGA                    24</span></span>
<span><span class="co">#&gt; 2           0.644     ACAUAUUUUUUUGGACGAAC                    24</span></span>
<span><span class="co">#&gt; 4           0.000                        N                    24</span></span>
<span><span class="co">#&gt; 5           0.258 AUAGACAAAUCGCACUCGCUCUCG                    24</span></span>
<span><span class="co">#&gt; 7           1.546 UUAUAACGUGAAUUCUGAUGAAAC                    24</span></span>
<span><span class="co">#&gt; 8           1.160 AUAUUGGUAUUGGACCAUCAGAUU                    24</span></span>
<span><span class="co">#&gt;   Count_selfgraft_3 RPM_selfgraft_3     MajorRNA_selfgraft_3 DicerCounts</span></span>
<span><span class="co">#&gt; 1                30           2.954 AGUGACGGUCUCAUGUAACAUAGA           3</span></span>
<span><span class="co">#&gt; 2                 1           0.098 UAUGUCUACUCUUUAUUACUCUCC           3</span></span>
<span><span class="co">#&gt; 4                11           1.083 AUUCCUCUGACUAAAUGACAAUUU           2</span></span>
<span><span class="co">#&gt; 5                 5           0.492 AAUCACAAUGGGUAGACACUUAGA           3</span></span>
<span><span class="co">#&gt; 7                29           2.856  AUAACGUGAAUUCUGAUGAAACC           3</span></span>
<span><span class="co">#&gt; 8                16           1.576 GGUAUUGGACCAUCAGAUUGACGA           3</span></span>
<span><span class="co">#&gt;   DicerConsensus CountMean log2FoldChange    pvalue padjusted    logCPM</span></span>
<span><span class="co">#&gt; 1             24  21.00000     0.09399726 0.7992536         1 10.819068</span></span>
<span><span class="co">#&gt; 2             24   3.00000     0.20765865 0.8863680         1  8.661401</span></span>
<span><span class="co">#&gt; 4             24   4.00000     0.01104373 1.0000000         1  8.831677</span></span>
<span><span class="co">#&gt; 5             24   6.00000     0.07211172 1.0000000         1  9.271215</span></span>
<span><span class="co">#&gt; 7             24  20.33333     0.08714351 0.8448235         1 10.763685</span></span>
<span><span class="co">#&gt; 8             24  12.33333     0.09474876 0.8602119         1 10.144466</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level3">
<h3 id="identify-the-mobile-molecules">Identify the mobile molecules<a class="anchor" aria-label="anchor" href="#identify-the-mobile-molecules"></a>
</h3>
<p>It is essential to remove the noise from the data to isolate potential mobile molecules which include falsely mapped clusters and, in this case, the clusters which map the the scion genotype when the aim is to identify molecules traveling from the rootstock. In the data, the aim is to explore whether there are tomato sRNA traveling up through the graft junction and exiting into the eggplant leaf tissues. We can identify potential mobile molecule by using the <code><a href="../reference/RNAmobile.html">RNAmobile()</a></code> function. It selects clusters mapping to the rootstock genome in the hetero-graft. Therefore, we will remove clusters mapped to chromosome in the genotype used in the self-graft controls.</p>
<p>In the example, the tomato genome has chromosome labeled with “SL40” whilst the the eggplant genome has no labels. To remove clusters associated with the eggplant genome and keep only clusters associated with the tomato genome, set the <code>"task"</code> argument as <code>"keep"</code>.</p>
<p>The <code><a href="../reference/RNAmobile.html">RNAmobile()</a></code> function can take into account the statistically significance. If set to consider this, the function will filter sRNA based on an adjusted p-value threshold = 0.05. The threshold can be changed to make it less stringent by using the argument <code>padj</code>.Similarly, if you would prefer to extract the mobile RNA based on the p-value,rather than the adjusted p-values, a numeric threshold can be set for the argument <code>p.value</code>. As default, the function does not consider the statistical values and properties of the sRNAs. To consider, use <code>statistical=TRUE</code></p>
<p>Below we have used the datasets which contains the additional columns with the statistical values. As stated above, we do not need to use the statistical values to identify the mobile molecules but it can be explored and utilised to plot a heatmap based on the logFC.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># vector of control names</span></span>
<span><span class="va">control_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span>, <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Identify potential tomato mobile molecules, before subsetting classes</span></span>
<span><span class="va">mobile_sRNA</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         genome.ID <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span>, </span>
<span>                         statistical <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or identify within specific classes </span></span>
<span><span class="co"># within 24-nt sRNA datset </span></span>
<span><span class="va">mobile_sRNA_24</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_24_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         genome.ID <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span><span class="op">)</span></span>
<span><span class="co"># within 21/22-nt sRNA datset</span></span>
<span><span class="va">mobile_sRNA_2122</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_2122_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         genome.ID <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># If you wish to consider the statistical values from DESeq2 analysis </span></span>
<span></span>
<span><span class="va">mobile_sRNA_sig24</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_24_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         genome.ID <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span>, </span>
<span>                         statistical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mobile_sRNA_sig2122</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmobile.html">RNAmobile</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_2122_DESeq2</span>, </span>
<span>                         controls <span class="op">=</span> <span class="va">control_names</span>,</span>
<span>                         genome.ID <span class="op">=</span> <span class="st">"SL40"</span>,</span>
<span>                         task <span class="op">=</span> <span class="st">"keep"</span>, </span>
<span>                         statistical <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level4">
<h4 id="heatmap-plots-to-represent-mobile-molecules">Heatmap plots to represent mobile molecules<a class="anchor" aria-label="anchor" href="#heatmap-plots-to-represent-mobile-molecules"></a>
</h4>
<p>Here we can use a hierarchical clustering heatmap to analysis the potential statistically significant mobile sRNAs we have identified. It is expected that the log2 fold changes should be opposite between your treatment and controls where the treatment samples shows up-regulation and the control samples show down-regulation.</p>
<p><br></p>
</div>
<div class="section level4">
<h4 id="mobile-molecules-from-the-24-nt-srna-dataset">Mobile molecules from the 24-nt sRNA dataset<a class="anchor" aria-label="anchor" href="#mobile-molecules-from-the-24-nt-srna-dataset"></a>
</h4>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p10</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">mobile_sRNA_sig24</span><span class="op">)</span></span>
<span><span class="va">p11</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">mobile_sRNA_sig2122</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="mobileRNA_files/figure-html/unnamed-chunk-20-1.png" alt="An example heatmap of the potential mobile RNA molecules Identifed by the `DESeq2` method Class of 24-nt sRNAs (left) and Class of 21/22-nt sRNAs (right). Where the rows represents the sRNA cluster number and the columns represents replicates." width="50%"><img src="mobileRNA_files/figure-html/unnamed-chunk-20-2.png" alt="An example heatmap of the potential mobile RNA molecules Identifed by the `DESeq2` method Class of 24-nt sRNAs (left) and Class of 21/22-nt sRNAs (right). Where the rows represents the sRNA cluster number and the columns represents replicates." width="50%"><p class="caption">
An example heatmap of the potential mobile RNA molecules Identifed by the <code>DESeq2</code> method Class of 24-nt sRNAs (left) and Class of 21/22-nt sRNAs (right). Where the rows represents the sRNA cluster number and the columns represents replicates.
</p>
</div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="save-output">Save output<a class="anchor" aria-label="anchor" href="#save-output"></a>
</h4>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">mobile_sRNA</span>, <span class="st">"./output.txt"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="functional-analysis">Functional Analysis<a class="anchor" aria-label="anchor" href="#functional-analysis"></a>
</h3>
<p>By now you will have retrieved a list of potentially mobile sRNA molecules and you will want to identify whether they play a role in the biological system.</p>
<p>Small interfering RNAs (siRNAs) have a length of 24-nucleotides, and are known to play a role in the RNA-directed DNA methylation pathway via targeting complementary sequencing in the DNA. While, miRNAs have a length of 21/22-nucleotides, and are known to instigate post-transcriptional gene silencing via mRNA degradation or repression of translation.</p>
<p>To predict the function of a mobile sRNA in destination tissues, we can overlap the information with the genome annotation of the origin tissue to determine the local functionality. As well as, retrieving the nucleotide sequence of the most abundant sRNA within a cluster and locating complementary sequences in both the genome of origin and genome of destination using BLAST.</p>
<p>If you have different datatypes, for example RNAseq and BSseq, and want to overlap your data it might be interesting to see whether there are complementary DNA sequences at the promoter regions of your deferentially expressed genes of interest to the identified potential mobile sRNAs. Moreover, in theory, if these regions also show differential methylation it could be a potential indication of a molecular pathway.</p>
<p>IMPORTANT: If chromosome names were altered in either of genome assemblies used in the mapping steps, the chromosome names in the genome annotation files must match the same pattern. If you used the [mobileRNA::RNAmergeGenomes()] function, the chromosome names will by default be altered. The [mobileRNA::RNAmergeAnnotations()] function can be utilsed to alter both genome annotation files in the same way. This functions alters the names, saves each altered genome annotation and then merges them into one.</p>
<p><br></p>
<div class="section level4">
<h4 id="add-genomic-attributes-to-srna-clusters">Add genomic attributes to sRNA clusters<a class="anchor" aria-label="anchor" href="#add-genomic-attributes-to-srna-clusters"></a>
</h4>
<p>Each sRNA cluster contains coordinates, these can be matches with coordinates in an annotation file. If there is a match, the function returns the input data with additional fields, where sRNA cluster will receive information on:</p>
<ul>
<li>
<code>source</code> : name of the program that generated this feature, or the data source (database or project name)</li>
<li>
<code>feature</code> : feature type name, e.g. Gene, Variation, Similarity</li>
<li>
<code>score</code> : A floating point value</li>
<li>
<code>strand</code> : defined as + (forward) or - (reverse)</li>
<li>
<code>frame</code> : One of ‘0’, ‘1’ or ‘2’. ‘0’ indicates that the first base of the feature is the first base of a codon, ‘1’ that the second base is the first base of a codon, and so on</li>
<li>
<code>attribute</code> : A semicolon-separated list of tag-value pairs, providing additional information about each feature.</li>
</ul>
<p>Using the genomic locations of the mobile sRNA clusters, we can identify overlaps in the genome annotation of the origin tissue to predict the implication of the mobile sRNA.</p>
<p>This function can be utilsed at the beginning of the analysis steps where all sRNA clusters from both genotypes can be annotated with the genomic attributes associated to their location. This can be completed in a single step using a merge annotation file using the function [mobileRNA::RNAmergeAnnotations()].</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mobile_df_attributes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAattributes.html">RNAattributes</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mobile_sRNA</span>,</span>
<span>                            annotation <span class="op">=</span> <span class="st">"./annotation/origin_annotation.gff3"</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
<div class="section level4">
<h4 id="summarise-overlaps-with-genomic-features">Summarise overlaps with genomic features<a class="anchor" aria-label="anchor" href="#summarise-overlaps-with-genomic-features"></a>
</h4>
<p>We can also locate specific genomic features which overlap with the the sRNA cluster locus including promoter regions, exon, introns, untranslated regions and repeat regions.The results can either be displayed in the matrix as an absolute value or as a percentage of the total:</p>
<p>NOTE: This function look for the exact matches in the locus (chromosome, start and end regions) between the sRNA cluster and the annotation file.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mobile_df_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAfeatures.html">RNAfeatures</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">mobile_sRNA</span>,</span>
<span>                            annotation <span class="op">=</span> <span class="st">"./annotation/origin_annotation.gff3"</span>, </span>
<span>                      repeats <span class="op">=</span> <span class="st">"./annotation/origin_annotation_repeats.gff3"</span><span class="op">)</span></span></code></pre></div>
<p>The distribution of sRNA clusters across genomic features is strongly related to the class of sRNA based on their nucleotide length. It is expected that 24-nt sRNA are associated with repeat regions and promoter regions, while 21/22-nt sRNA are likely to be associated with coding sequence.</p>
<p>The resulting dataframe can be passed into a plot of your choice. Similarly, this features can be utilsed across different data sets. For example, across the total data set or specific mobile sRNA subsets. <br></p>
</div>
<div class="section level4">
<h4 id="retrieve-srna-sequence-complementary-sequences">Retrieve sRNA sequence &amp; complementary sequences<a class="anchor" aria-label="anchor" href="#retrieve-srna-sequence-complementary-sequences"></a>
</h4>
<p>Lastly, lets extract the sRNA cluster sequences.These can utilised with <code>BLAST</code>, or plant sRNA target prediction tools such as <code>TargetFinder</code> (<a href="https://github.com/carringtonlab/TargetFinder" class="external-link uri">https://github.com/carringtonlab/TargetFinder</a>) or overlapped with <code>mRNAseq</code> data to identify complementary sequences to further elucidate the potential function of sRNAs.</p>
<p>The <code>RNAsequences</code> function identifies whether the most abundant sRNA is consistent across the replicates, and if so, it extracts the sRNA nucleotide sequence and calculates the RNA and DNA complementary sequences, as well as stating the length of the sequence.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">mobile_sequences</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsequences.html">RNAsequences</a></span><span class="op">(</span><span class="va">mobile_sRNA</span><span class="op">)</span></span>
<span><span class="co">#&gt; The minimum number of matches required to form a consensus sRNA sequence is...  1 </span></span>
<span><span class="co">#&gt; The consensus sRNA sequences will be choose at random in the case of a tie</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mobile_sequences</span><span class="op">)</span></span>
<span><span class="co">#&gt;             Cluster Match                 Sequence Width</span></span>
<span><span class="co">#&gt; 1610 cluster_133430    No                     &lt;NA&gt;  &lt;NA&gt;</span></span>
<span><span class="co">#&gt; 2010 cluster_133435   Yes AUGUAACUGAUUGAGUAUAUAUUU    24</span></span>
<span><span class="co">#&gt; 2410 cluster_133450   Yes  CCCGGACUGUCGGACACUACACA    23</span></span>
<span><span class="co">#&gt; 2510 cluster_133451   Yes  ACGGCCUAUUGCACACCUGAAUU    23</span></span>
<span><span class="co">#&gt; 2610 cluster_133452   Yes AGGAUUACUUACCACUCCACCAUA    24</span></span>
<span><span class="co">#&gt; 2810 cluster_133455   Yes ACACUUGUAGGUCUGAGUGAAACU    24</span></span>
<span><span class="co">#&gt;             Complementary_RNA        Complementary_DNA</span></span>
<span><span class="co">#&gt; 1610                       NA                       NA</span></span>
<span><span class="co">#&gt; 2010 UACAUUGACUAACUCAUAUAUAAA TACATTGACTAACTCATATATAAA</span></span>
<span><span class="co">#&gt; 2410  GGGCCUGACAGCCUGUGAUGUGU  GGGCCTGACAGCCTGTGATGTGT</span></span>
<span><span class="co">#&gt; 2510  UGCCGGAUAACGUGUGGACUUAA  TGCCGGATAACGTGTGGACTTAA</span></span>
<span><span class="co">#&gt; 2610 UCCUAAUGAAUGGUGAGGUGGUAU TCCTAATGAATGGTGAGGTGGTAT</span></span>
<span><span class="co">#&gt; 2810 UGUGAACAUCCAGACUCACUUUGA TGTGAACATCCAGACTCACTTTGA</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level3">
<h3 id="explore-population-scale-srna-difference">Explore population-scale sRNA difference<a class="anchor" aria-label="anchor" href="#explore-population-scale-srna-difference"></a>
</h3>
<p>Here, we will look at comparing the native population of sRNAs and how they might have changes in the leaf due to the chimeric system compared to the control.</p>
<p>As a reminder, the data used here will not yield any results as the treatment and control samples contain the exact same population of eggplant sRNAs, the only difference in the treatment samples are the spiked tomato sRNA clusters.</p>
<p>We have already undertaken the statistical analysis in the previous steps, we will be utilising this dataset going forward.</p>
<div class="section level4">
<h4 id="srna-abundance">sRNA abundance<a class="anchor" aria-label="anchor" href="#srna-abundance"></a>
</h4>
<p>When comparing treatment to control conditions, it might be the case that the same sRNA clusters are found within both yet there could be difference in the total abundance of the shared clusters. For instance, for a given sRNA cluster the samples in the treatment condition might have a greater abundance than the samples in the control condition. These difference could attribute to observed differences.</p>
<p>The statistical analysis calculated the log2FC values for each sRNA cluster by comparing the normalised counts between treatment and control. Here, a positive log2FC indicates an increased abundance of transcripts for a given sRNA cluster in the treatment compared to the control, while negative log2FC indicates decreased abundance of transcripts for a given sRNA cluster. The statistical significance of the log2FC is determined by the (adjusted) pvalue also calculated in the previous step.</p>
<p>Here we will filter the data to select sRNA clusters which are statistically significant, and then plot the results as a heatmap to compare the conditions.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select only significant sRNAs</span></span>
<span><span class="va">significant_sRNAs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAsignificant.html">RNAsignificant</a></span><span class="op">(</span><span class="va">sRNA_edgeR</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">#plot</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">significant_sRNAs</span>, cluster <span class="op">=</span> <span class="cn">FALSE</span>, row.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="mobileRNA_files/figure-html/unnamed-chunk-25-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Lets look at the plot:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span><span class="op">$</span><span class="va">plot</span> </span></code></pre></div>
<p><img src="mobileRNA_files/figure-html/unnamed-chunk-26-1.png" width="700" style="display: block; margin: auto;"><br></p>
</div>
</div>
<div class="section level3">
<h3 id="step-6-identify-absolute-srna-population-difference">Step 6: Identify absolute sRNA population difference<a class="anchor" aria-label="anchor" href="#step-6-identify-absolute-srna-population-difference"></a>
</h3>
<p>Finally, the <code><a href="../reference/RNApopulation.html">RNApopulation()</a></code> function can be utilised to identify unique sRNA populations found in the treatment or control condition.</p>
<p>First lets look at the sRNA clusters unique to the treatment condition. In the chimeric heterografts, we expect that the foreign sRNAs will also be selected in this pick-up, therefore, we can use the parameter <code>genome.ID</code> to remove sRNA cluster related to the foreign genome.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select sRNA clusters only found in treatment &amp; not in the control samples</span></span>
<span><span class="va">treatment_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span> , <span class="st">"heterograft_3"</span><span class="op">)</span></span>
<span><span class="va">unique_treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNApopulation.html">RNApopulation</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>, </span>
<span>                                   conditions <span class="op">=</span> <span class="va">treatment_reps</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># look at number of sRNA cluster only found in treatment </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">unique_treatment</span><span class="op">)</span></span></code></pre></div>
<p>Now, the sRNA clusters unique to the control condition:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select sRNA clusters only found in control &amp; not in the treatment samples</span></span>
<span><span class="va">control_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span> , <span class="st">"selfgraft_3"</span><span class="op">)</span></span>
<span><span class="va">unique_control</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNApopulation.html">RNApopulation</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>, </span>
<span>                                   conditions <span class="op">=</span> <span class="va">control_reps</span><span class="op">)</span></span>
<span><span class="co"># look at number of sRNA cluster only found in control  </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">unique_control</span><span class="op">)</span> </span></code></pre></div>
<p><br></p>
<p>The plotting functions within the package can be used to display these results, specifically and . Similarly, the functions to assist with functional analysis can also be utilsied to pull the consensus sRNA sequence for each cluster in the population, or find overlaps with known annotations in the native genome. <br></p>
</div>
</div>
<div class="section level2">
<h2 id="additional-features">Additional features<a class="anchor" aria-label="anchor" href="#additional-features"></a>
</h2>
<p>The <em>mobileRNA</em> package offers several optional extra-helpful functions to aid analysis. This includes a function to calculate the mean RPM and counts across specific samples and also calculate the overlap between genomic features and sRNA clusters.</p>
<p><br></p>
<div class="section level3">
<h3 id="calculate-rpm-and-count-means-for-specific-samples">Calculate RPM and Count means for specific samples<a class="anchor" aria-label="anchor" href="#calculate-rpm-and-count-means-for-specific-samples"></a>
</h3>
<p>This function calculates the RPM mean and Count mean across specific samples.The output of the calculation is added as an additional column to the input dataframe.</p>
<p>This can be manipulated to be used before or after identified mobile RNAs. When calculating means after identifying mobile molecules, it is recommend to base this on the heterograft samples rather than the full dataset (selfgrafts + heterograft replicates) as all potential mobile sRNAs should have zero counts or RPM value in the selfgraft control replicates. By including the full dataset in these calculations, the mean values for the mobile molecules will be skewed.</p>
<p>Therefore, depending on your intended output this function could be approached in multiple ways.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate mean RPM and mean Counts for heterograft samples</span></span>
<span><span class="va">selected_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span>, <span class="st">"heterograft_3"</span><span class="op">)</span></span>
<span><span class="va">means</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RNAmean.html">RNAmean</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">sRNA_data_consensus</span>, conditions <span class="op">=</span> <span class="va">selected_samples</span><span class="op">)</span></span></code></pre></div>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.0.2 (2020-06-22)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS  10.16</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] mobileRNA_0.99.0 BiocStyle_2.16.1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] systemfonts_1.0.4           splines_4.0.2              </span></span>
<span><span class="co">#&gt;   [3] BiocParallel_1.22.0         listenv_0.8.0              </span></span>
<span><span class="co">#&gt;   [5] GenomeInfoDb_1.24.2         ggplot2_3.4.2              </span></span>
<span><span class="co">#&gt;   [7] Repitools_1.34.0            digest_0.6.29              </span></span>
<span><span class="co">#&gt;   [9] foreach_1.5.2               htmltools_0.5.5            </span></span>
<span><span class="co">#&gt;  [11] viridis_0.6.2               fansi_1.0.2                </span></span>
<span><span class="co">#&gt;  [13] magrittr_2.0.3              Rsolnp_1.16                </span></span>
<span><span class="co">#&gt;  [15] memoise_2.0.1               BSgenome_1.56.0            </span></span>
<span><span class="co">#&gt;  [17] cluster_2.1.2               doParallel_1.0.17          </span></span>
<span><span class="co">#&gt;  [19] limma_3.44.3                globals_0.14.0             </span></span>
<span><span class="co">#&gt;  [21] Biostrings_2.56.0           annotate_1.66.0            </span></span>
<span><span class="co">#&gt;  [23] matrixStats_0.61.0          pkgdown_2.0.7              </span></span>
<span><span class="co">#&gt;  [25] beepr_1.3                   colorspace_2.0-3           </span></span>
<span><span class="co">#&gt;  [27] blob_1.2.2                  ggrepel_0.9.1              </span></span>
<span><span class="co">#&gt;  [29] textshaping_0.3.6           xfun_0.37                  </span></span>
<span><span class="co">#&gt;  [31] dplyr_1.1.2                 crayon_1.5.0               </span></span>
<span><span class="co">#&gt;  [33] RCurl_1.98-1.6              jsonlite_1.8.4             </span></span>
<span><span class="co">#&gt;  [35] genefilter_1.70.0           progressr_0.10.0           </span></span>
<span><span class="co">#&gt;  [37] survival_3.3-1              iterators_1.0.14           </span></span>
<span><span class="co">#&gt;  [39] glue_1.6.2                  gtable_0.3.0               </span></span>
<span><span class="co">#&gt;  [41] zlibbioc_1.34.0             XVector_0.28.0             </span></span>
<span><span class="co">#&gt;  [43] DelayedArray_0.14.1         future.apply_1.8.1         </span></span>
<span><span class="co">#&gt;  [45] BiocGenerics_0.34.0         scales_1.2.1               </span></span>
<span><span class="co">#&gt;  [47] pheatmap_1.0.12             vsn_3.56.0                 </span></span>
<span><span class="co">#&gt;  [49] DBI_1.1.2                   edgeR_3.30.3               </span></span>
<span><span class="co">#&gt;  [51] Rcpp_1.0.8.3                viridisLite_0.4.0          </span></span>
<span><span class="co">#&gt;  [53] xtable_1.8-4                bit_4.0.4                  </span></span>
<span><span class="co">#&gt;  [55] preprocessCore_1.50.0       stats4_4.0.2               </span></span>
<span><span class="co">#&gt;  [57] truncnorm_1.0-8             gplots_3.1.1               </span></span>
<span><span class="co">#&gt;  [59] RColorBrewer_1.1-2          farver_2.1.0               </span></span>
<span><span class="co">#&gt;  [61] pkgconfig_2.0.3             XML_3.99-0.9               </span></span>
<span><span class="co">#&gt;  [63] sass_0.4.0                  locfit_1.5-9.4             </span></span>
<span><span class="co">#&gt;  [65] utf8_1.2.2                  DNAcopy_1.62.0             </span></span>
<span><span class="co">#&gt;  [67] tidyselect_1.2.0            labeling_0.4.2             </span></span>
<span><span class="co">#&gt;  [69] rlang_1.1.0                 AnnotationDbi_1.50.3       </span></span>
<span><span class="co">#&gt;  [71] munsell_0.5.0               tools_4.0.2                </span></span>
<span><span class="co">#&gt;  [73] cachem_1.0.6                cli_3.6.0                  </span></span>
<span><span class="co">#&gt;  [75] generics_0.1.2              RSQLite_2.2.10             </span></span>
<span><span class="co">#&gt;  [77] audio_0.1-10                RPushbullet_0.3.4          </span></span>
<span><span class="co">#&gt;  [79] evaluate_0.15               stringr_1.5.0              </span></span>
<span><span class="co">#&gt;  [81] fastmap_1.1.0               yaml_2.3.5                 </span></span>
<span><span class="co">#&gt;  [83] ragg_1.2.2                  knitr_1.42                 </span></span>
<span><span class="co">#&gt;  [85] bit64_4.0.5                 fs_1.5.2                   </span></span>
<span><span class="co">#&gt;  [87] caTools_1.18.2              purrr_1.0.1                </span></span>
<span><span class="co">#&gt;  [89] pbapply_1.7-0               future_1.24.0              </span></span>
<span><span class="co">#&gt;  [91] compiler_4.0.2              rstudioapi_0.14            </span></span>
<span><span class="co">#&gt;  [93] curl_4.3.2                  png_0.1-7                  </span></span>
<span><span class="co">#&gt;  [95] affyio_1.58.0               tibble_3.2.1               </span></span>
<span><span class="co">#&gt;  [97] geneplotter_1.66.0          bslib_0.3.1                </span></span>
<span><span class="co">#&gt;  [99] stringi_1.7.6               SimDesign_2.10.1           </span></span>
<span><span class="co">#&gt; [101] highr_0.9                   desc_1.4.1                 </span></span>
<span><span class="co">#&gt; [103] gsmoothr_0.1.7              lattice_0.20-45            </span></span>
<span><span class="co">#&gt; [105] Matrix_1.4-0                vctrs_0.6.2                </span></span>
<span><span class="co">#&gt; [107] pillar_1.9.0                lifecycle_1.0.3            </span></span>
<span><span class="co">#&gt; [109] BiocManager_1.30.16         jquerylib_0.1.4            </span></span>
<span><span class="co">#&gt; [111] data.table_1.14.2           bitops_1.0-7               </span></span>
<span><span class="co">#&gt; [113] rtracklayer_1.48.0          GenomicRanges_1.40.0       </span></span>
<span><span class="co">#&gt; [115] R6_2.5.1                    affy_1.66.0                </span></span>
<span><span class="co">#&gt; [117] bookdown_0.32               KernSmooth_2.23-20         </span></span>
<span><span class="co">#&gt; [119] gridExtra_2.3               IRanges_2.22.2             </span></span>
<span><span class="co">#&gt; [121] parallelly_1.30.0           sessioninfo_1.2.2          </span></span>
<span><span class="co">#&gt; [123] codetools_0.2-18            Ringo_1.52.0               </span></span>
<span><span class="co">#&gt; [125] MASS_7.3-55                 gtools_3.9.2               </span></span>
<span><span class="co">#&gt; [127] SummarizedExperiment_1.18.2 DESeq2_1.28.1              </span></span>
<span><span class="co">#&gt; [129] rprojroot_2.0.2             withr_2.5.0                </span></span>
<span><span class="co">#&gt; [131] GenomicAlignments_1.24.0    Rsamtools_2.4.0            </span></span>
<span><span class="co">#&gt; [133] S4Vectors_0.26.1            GenomeInfoDbData_1.2.3     </span></span>
<span><span class="co">#&gt; [135] parallel_4.0.2              grid_4.0.2                 </span></span>
<span><span class="co">#&gt; [137] tidyr_1.3.0                 rmarkdown_2.20             </span></span>
<span><span class="co">#&gt; [139] Biobase_2.48.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
