<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Load and organise either sRNAseq or mRNAseq results into a
single dataframe containing all experimental replicates specified where rows
represent either a sRNA locus or gene, respectively."><title>Import and organise sRNAseq or mRNA data sets — RNAimport • mobileRNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Import and organise sRNAseq or mRNA data sets — RNAimport"><meta property="og:description" content="Load and organise either sRNAseq or mRNAseq results into a
single dataframe containing all experimental replicates specified where rows
represent either a sRNA locus or gene, respectively."><meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/mobileRNA.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/mobileRNA_standard.html">Analyse sRNA populations using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_chimeric.html">Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_mRNA.html">Explore canditate mobile mRNAs &amp; population-scale mRNA changes in chimeric systems using mobileRNA</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Import and organise sRNAseq or mRNA data sets</h1>
      <small class="dont-index">Source: <a href="https://github.com/KJeynesCupper/mobileRNA.git/blob/HEAD/R/RNAimport.R" class="external-link"><code>R/RNAimport.R</code></a></small>
      <div class="d-none name"><code>RNAimport.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Load and organise either sRNAseq or mRNAseq results into a
single dataframe containing all experimental replicates specified where rows
represent either a sRNA locus or gene, respectively.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RNAimport</span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sRNA"</span>, <span class="st">"mRNA"</span><span class="op">)</span>,</span>
<span>  <span class="va">directory</span>,</span>
<span>  <span class="va">samples</span>,</span>
<span>  annotation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tidy <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>string; define type of Next-Generation Sequencing dataset
originates from, either "sRNA" or "mRNA" are the only valid inputs.</p></dd>


<dt>directory</dt>
<dd><p>Path to directory containing of sample folders. NOTE:
Following the suggested pre-processing steps, these can be found in second
alignment folders.</p></dd>


<dt>samples</dt>
<dd><p>Vector of characters naming the sample names correlating
to outputted folders located in the <code>`directory`</code> argument path.</p></dd>


<dt>annotation</dt>
<dd><p>Path to GFF genome annotation file. Required only for
mRNAseq data.</p></dd>


<dt>tidy</dt>
<dd><p>Logical; removes genes from analysis where there are zero counts
across all replicates.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><strong>For sRNAseq:</strong></p>


<p>A dataframe where rows represent sRNA clusters and columns represent
replicate information. Replicate information includes Dicercall, Counts, RPM
and MajorRNA sequence. Each replicate information is distinguishable as
the replicate name is ajoined as a prefix to each column name.</p>


<p><strong>For mRNAseq:</strong></p>


<p>A dataframe where rows represent gene names and columns representing
replicate information. Replicate information includes Counts and FPKM.
Additional columns include SampleCounts (the number of replicates which
contains counts), chr (chromosome), start and end coordinates are
also included in the output. Each replicate information is distinguishable as
the replicate name is adjoined as a prefix to each column name.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><strong>For sRNAseq:</strong>
Supply the directory to the folder storing the various sample replicate
folders produced by ShortStack analysis. Within each sample replicate folder,
there is "Result.txt". This file contains many different columns, but
for the analysis the columns of interest are  "Locus", "DicerCall", "Reads",
and "RPM" which will all be imported and loads into the new data frame.
Note that "Reads" will be converted to "Counts" as output of the function.
Locus contains the genomic locations of clusters,
Dicercall contains the most likely size of the cluster, Reads contains the
number of reads mapped to the cluster (recommended to use unique mapping,
hence these will be the number of uniquely mapped reads, ie, not including
multimapped reads). Lastly, RPM contains the Reads per Million score.</p>
<p><code>"RNAimport"</code> takes the results from  pre-processed sRNAseq data which
has been mapped and undertaken cluster analysis through ShortStack, organises
into a single data frame. Each row represent an sRNA dicer-derived cluster
within the analysis and columns 1 - 5 supply information on the cluster
including the locus, the separated coordinates
(chromosome, start, end) and cluster name.</p>
<p>Further columns represent data imported for each samples including DicerCall,
Counts and RPM. The DicerCall represents the size of most
abundant small RNA size based on the parameter used in ShortStack.
The Count column represents the number of aligned sRNA-seq reads that overlap
the locus. The RPM represents the reads per million.
For each replicate included in the analysis, these columns are labeled with
the type and then then name of the sample, for example, for
a sample called "Sample1", the information from this sample will be stored in
columns DicerCall_Sample1, Count_Sample1 and RPM_Sample1.</p>
<p><strong>For mRNAseq:</strong></p>
<p>Supply the directory to the folder storing the various sample replicate
files outputted by the alignment program. Each file contains many the gene
name (column 1) and the raw count value (column 2). The function identifies
all genes across the analysis, and organises the count values for each sample
replicate into a column. For example, this will look like
"Counts_heterograft_1" for the sample replicate with the file name
"heterograft_1". Further columns include the gene locus information (chr,
start, end, width) are add to the start of the dataframe, and the FPKM
for each sample replicate is calculated. The FPKM columns are organised and
titled in the same format at the "Counts_" columns.</p>
<p>When working with mRNAseq data, a GFF genome annotation file is required.
When working with a chimeric biological system and a merged annotation file,
ensure the genomes remain distinguishable. The function
<code><a href="RNAmergeAnnotations.html">RNAmergeAnnotations()</a></code> can help with this.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># import sRNAseq data</span></span></span>
<span class="r-in"><span><span class="va">df_sRNA</span> <span class="op">&lt;-</span> <span class="fu">RNAimport</span><span class="op">(</span>input <span class="op">=</span> <span class="st">"sRNA"</span>,</span></span>
<span class="r-in"><span>                     directory <span class="op">=</span> <span class="st">"./analysis/sRNA_mapping_results/"</span>,</span></span>
<span class="r-in"><span>                     samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span>,</span></span>
<span class="r-in"><span>                     <span class="st">"heterograft_3"</span>,<span class="st">"selfgraft_1"</span> , <span class="st">"selfgraft_2"</span> ,</span></span>
<span class="r-in"><span>                     <span class="st">"selfgraft_3"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># import mRNAseq data</span></span></span>
<span class="r-in"><span><span class="va">df_mRNA</span> <span class="op">&lt;-</span> <span class="fu">RNAimport</span><span class="op">(</span>input <span class="op">=</span> <span class="st">"mRNA"</span>,</span></span>
<span class="r-in"><span>                     directory <span class="op">=</span> <span class="st">"./analysis/mRNA_counts/"</span>,</span></span>
<span class="r-in"><span>                     samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span>,</span></span>
<span class="r-in"><span>                     <span class="st">"heterograft_3"</span>,<span class="st">"selfgraft_1"</span> , <span class="st">"selfgraft_2"</span> ,</span></span>
<span class="r-in"><span>                     <span class="st">"selfgraft_3"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span><span class="co"># The output of this function can be explore in the data object sRNA_data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sRNA_data"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sRNA_data</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                   Locus chr     start       end   Cluster</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 1:100000956-100001076   1 100000956 100001076 cluster_1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 1:100001486-100001554   1 100001486 100001554 cluster_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 1:100001646-100001731   1 100001646 100001731 cluster_3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4   1:10001034-10001073   1  10001034  10001073 cluster_4</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 1:100011795-100011930   1 100011795 100011930 cluster_5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6   1:10002000-10002065   1  10002000  10002065 cluster_6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   DicerCall_heterograft_1 Count_heterograft_1 RPM_heterograft_1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                      24                  19             1.908</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                      24                   3             0.301</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                       N                   3             0.301</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                      24                   1             0.100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                      24                  11             1.105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                      24                   9             0.904</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>      MajorRNA_heterograft_1 DicerCall_heterograft_2 Count_heterograft_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  AGUGACGGUCUCAUGUAACAUAGA                      24                  14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  AAAGAGUAGACAUAUAUUAAAUGG                      20                   5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 CUGUCUCAUGUGACACUCUUUCACU                      23                   6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  AAAUUGUCAUUUAGUCAGAGGAAU                       N                   0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  AUAGACAAAUCGCACUCGCUCUCG                      24                   2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  UUAGAUUUCUGGAGUAACAACUUU                       N                   2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   RPM_heterograft_2   MajorRNA_heterograft_2 DicerCall_heterograft_3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1             1.803 AGUGACGGUCUCAUGUAACAUAGA                      24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2             0.644     ACAUAUUUUUUUGGACGAAC                      24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3             0.773  UCUGUCUCAUGUGACACUCUUUC                      24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4             0.000                        N                      24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5             0.258 AAGUGUCUACCCAUUGUGAUUCGC                      24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6             0.258      UUAUUGACAAACUAUAACU                      22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Count_heterograft_3 RPM_heterograft_3   MajorRNA_heterograft_3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                  30             2.953 AGUGACGGUCUCAUGUAACAUAGA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                   1             0.098 UAUGUCUACUCUUUAUUACUCUCC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                  11             1.083 AUCUUGUCAAGUGAAAGAGUGUCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                  11             1.083 AUUCCUCUGACUAAAUGACAAUUU</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                   5             0.492 AAUCACAAUGGGUAGACACUUAGA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                   3             0.295 UUAGAUUUCUGGAGUAACAACUUU</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   DicerCall_selfgraft_1 Count_selfgraft_1 RPM_selfgraft_1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                    24                19           1.909</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                    24                 3           0.301</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                     N                 3           0.301</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                    24                 1           0.100</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                    24                11           1.105</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                    24                 9           0.904</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        MajorRNA_selfgraft_1 DicerCall_selfgraft_2 Count_selfgraft_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1  AGUGACGGUCUCAUGUAACAUAGA                    24                14</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2  AAAGAGUAGACAUAUAUUAAAUGG                    20                 5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 CUGUCUCAUGUGACACUCUUUCACU                    23                 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4  AAAUUGUCAUUUAGUCAGAGGAAU                     N                 0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5  AAUGGGUAGACACUUAGAUACAUG                    24                 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6  UUAGAUUUCUGGAGUAACAACUUU                     N                 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   RPM_selfgraft_2     MajorRNA_selfgraft_2 DicerCall_selfgraft_3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1           1.804 AGUGACGGUCUCAUGUAACAUAGA                    24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           0.644     ACAUAUUUUUUUGGACGAAC                    24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           0.773  UCUGUCUCAUGUGACACUCUUUC                    24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           0.000                        N                    24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5           0.258 AUAGACAAAUCGCACUCGCUCUCG                    24</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6           0.258  AGAUUUCUGGAGUAACAACUUUG                    22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Count_selfgraft_3 RPM_selfgraft_3     MajorRNA_selfgraft_3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                30           2.954 AGUGACGGUCUCAUGUAACAUAGA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                 1           0.098 UAUGUCUACUCUUUAUUACUCUCC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3                11           1.083 AUCUUGUCAAGUGAAAGAGUGUCA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                11           1.083 AUUCCUCUGACUAAAUGACAAUUU</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                 5           0.492 AAUCACAAUGGGUAGACACUUAGA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6                 3           0.295   UUGACAAACUAUAACUUAAAUU</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

