<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Load and organise either sRNAseq or mRNAseq results into a
single dataframe containing all experimental replicates specified where rows
represent either a sRNA locus or gene, respectively."><title>Import and organise sRNAseq or mRNA data sets — RNAimport • mobileRNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Import and organise sRNAseq or mRNA data sets — RNAimport"><meta property="og:description" content="Load and organise either sRNAseq or mRNAseq results into a
single dataframe containing all experimental replicates specified where rows
represent either a sRNA locus or gene, respectively."><meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/mobileRNA.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Import and organise sRNAseq or mRNA data sets</h1>
      <small class="dont-index">Source: <a href="https://github.com/KJeynesCupper/mobileRNA.git/blob/HEAD/R/RNAimport.R" class="external-link"><code>R/RNAimport.R</code></a></small>
      <div class="d-none name"><code>RNAimport.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Load and organise either sRNAseq or mRNAseq results into a
single dataframe containing all experimental replicates specified where rows
represent either a sRNA locus or gene, respectively.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RNAimport</span><span class="op">(</span></span>
<span>  input <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sRNA"</span>, <span class="st">"mRNA"</span><span class="op">)</span>,</span>
<span>  <span class="va">directory</span>,</span>
<span>  <span class="va">samples</span>,</span>
<span>  report <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tidy <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>input</dt>
<dd><p>string; define type of Next-Generation Sequencing dataset
originates from, either "sRNA" or "mRNA" are the only valid inputs.</p></dd>


<dt>directory</dt>
<dd><p>Path to directory containing of sample folders. NOTE:
Following the suggested pre-processing steps, these can be found in second
alignment folders.</p></dd>


<dt>samples</dt>
<dd><p>Vector of characters naming the sample names correlating
to outputted folders located in the <code>`directory`</code> argument path.</p></dd>


<dt>report</dt>
<dd><p>Logical; prompts progress.</p></dd>


<dt>tidy</dt>
<dd><p>Logical; removes genes from analysis where there are zero counts
across all replicates.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A dataframe where rows represent sRNA clusters and columns represent
replicate information. Replicate information includes Dicercall, Counts, RPM
and MajorRNA sequence. Each replicate information is distinguishable as
the replicate name is ajoined as a prefix to each column name.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Supply the directory to the folder storing the various sample replicate
folders produced by ShortStack analysis. Within each sample replicate folder,
there is "Result.txt". This file contains many different columns, but
for the analysis the columns of interest are  "Locus", "DicerCall", "Reads",
and "RPM" which will all be imported and loads into the new data frame.
Note that "Reads" will be converted to "Counts" as output of the function.
Locus contains the genomic locations of clusters,
Dicercall contains the most likely size of the cluster, Reads contains the
number of reads mapped to the cluster (recommended to use unique mapping,
hence these will be the number of uniquely mapped reads, ie, not including
multimapped reads). Lastly, RPM contains the Reads per Million score.</p>
<p><code>"RNAimport"</code> takes the results from  pre-processed sRNAseq data which
has been mapped and undertaken cluster analysis through ShortStack, organises
into a single data frame. Each row represent an sRNA dicer-derived cluster
within the analysis and columns 1 - 5 supply information on the cluster
including the locus, the separated coordinates
(chromosome, start, end) and cluster name.</p>
<p>Further columns represent data imported for each samples including DicerCall,
Counts and RPM. The DicerCall represents the size of most
abundant small RNA size based on the parameter used in ShortStack.
The Count column represents the number of aligned sRNA-seq reads that overlap
the locus. The RPM represents the reads per million.
For each replicate included in the analysis, these columns are labeled with
the type and then then name of the sample, for example, for
a sample called "Sample1", the information from this sample will be stored in
columns DicerCall_Sample1, Count_Sample1 and RPM_Sample1.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># import sRNAseq data</span></span></span>
<span class="r-in"><span><span class="va">df_sRNA</span> <span class="op">&lt;-</span> <span class="fu">RNAimport</span><span class="op">(</span>input <span class="op">=</span> <span class="st">"sRNA"</span>,</span></span>
<span class="r-in"><span>                     directory <span class="op">=</span> <span class="st">"./analysis/sRNA_mapping_results/"</span>,</span></span>
<span class="r-in"><span>                     samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span>,</span></span>
<span class="r-in"><span>                     <span class="st">"heterograft_3"</span>,<span class="st">"selfgraft_1"</span> , <span class="st">"selfgraft_2"</span> ,</span></span>
<span class="r-in"><span>                     <span class="st">"selfgraft_3"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># import mRNAseq data</span></span></span>
<span class="r-in"><span><span class="va">df_mRNA</span> <span class="op">&lt;-</span> <span class="fu">RNAimport</span><span class="op">(</span>input <span class="op">=</span> <span class="st">"mRNA"</span>,</span></span>
<span class="r-in"><span>                     directory <span class="op">=</span> <span class="st">"./analysis/mRNA_counts/"</span>,</span></span>
<span class="r-in"><span>                     samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"heterograft_1"</span>, <span class="st">"heterograft_2"</span>,</span></span>
<span class="r-in"><span>                     <span class="st">"heterograft_3"</span>,<span class="st">"selfgraft_1"</span> , <span class="st">"selfgraft_2"</span> ,</span></span>
<span class="r-in"><span>                     <span class="st">"selfgraft_3"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span><span class="co"># The output of this function can be explore in the data object sRNA_data</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sRNA_data"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in check_for_XQuartz():</span> X11 library is missing: install XQuartz from xquartz.macosforge.org</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

