<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="A function to identify potential mobile sRNA or mRNA moleculea
detected in a tissue which originated from a different genotype, usually
related to a chimeric system."><title>Identify potential mobile sRNA &amp; mRNA molecules — RNAmobile • mobileRNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Identify potential mobile sRNA &amp; mRNA molecules — RNAmobile"><meta property="og:description" content="A function to identify potential mobile sRNA or mRNA moleculea
detected in a tissue which originated from a different genotype, usually
related to a chimeric system."><meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/mobileRNA.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Tutorials</h6>
    <a class="dropdown-item" href="../articles/mobileRNA_standard.html">Analyse sRNA populations using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_chimeric.html">Explore canditate mobile sRNAs &amp; population-scale sRNA changes in chimeric systems using mobileRNA</a>
    <a class="dropdown-item" href="../articles/mobileRNA_mRNA.html">Explore canditate mobile mRNAs &amp; population-scale mRNA changes in chimeric systems using mobileRNA</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Identify potential mobile sRNA &amp; mRNA molecules</h1>
      <small class="dont-index">Source: <a href="https://github.com/KJeynesCupper/mobileRNA.git/blob/HEAD/R/RNAmobile.R" class="external-link"><code>R/RNAmobile.R</code></a></small>
      <div class="d-none name"><code>RNAmobile.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A function to identify potential mobile sRNA or mRNA moleculea
detected in a tissue which originated from a different genotype, usually
related to a chimeric system.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RNAmobile</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">controls</span>,</span>
<span>  <span class="va">genome.ID</span>,</span>
<span>  task <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  statistical <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  padj <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  input <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  p.value <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>data</dt>
<dd><p>Numeric data frame</p></dd>


<dt>controls</dt>
<dd><p>Character vector; containing names of control samples.</p></dd>


<dt>genome.ID</dt>
<dd><p>a character string related to the chromosomes in a particular
genome. A distinguishing feature of the genome of interest or non-interest in
the chromosome name (<code>chr</code> column).</p></dd>


<dt>task</dt>
<dd><p>an option to keep or remove the chromosomes containing the
identifying string. To keep the chromosomes with the ID, set task=keep.
To remove, set <code>task="remove"</code>. As default, task is set to <code>keep</code>.</p></dd>


<dt>statistical</dt>
<dd><p>If TRUE, will undertake statistical filtering based on the
a p-value or adjusted p-value threshold stated by <code>padj</code> &amp; <code>p.value</code>.Default
set at FALSE. Requires presence of columns containing statistical data.
In order to filter by the adjusted p-value, a column named <code>padjusted</code> must
be present. Similarly, to filter by the p-value, a column named <code>pvalue</code> must
be present. See <code><a href="RNAanalysis.html">RNAanalysis()</a></code> to calculate statistical values.</p></dd>


<dt>padj</dt>
<dd><p>A user defined numeric value to represent the adjusted p-value
threshold to define statistic significance. Defaults set at 0.05.Only mobile
molecules with adjusted p-values equal or lower than specified are returned.</p></dd>


<dt>threshold</dt>
<dd><p>numeric; set a threshold level. For sRNAseq, this represents
filtering by the minimum number of replicates that defined the
dicer-consensus which is stored in the <code>DicerCounts</code> column. For mRNAseq,
this represents the minimum number of replicates which contained reads at the
gene locus, this information is storeds in the <code>SampleCounts</code> column.</p></dd>


<dt>input</dt>
<dd><p>character; must be either "sRNA" or "mRNA" to represent the type
of data, required when setting theshold.</p></dd>


<dt>p.value</dt>
<dd><p>A user defined numeric value to represent the p-value
threshold to define statistic significance. There is no default value, set
this instead of using an adjusted p-value to filter molecules. Only mobile
molecules with p-values equal or lower than specified are returned.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data-frame containing candidate mobile sRNAs, which could be
further filtered based on statistical significance and the ability to
by-pass the threshold which determines the number of replicates that
defined the dicer-consensus.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><strong>For sRNAseq:</strong>
The function identifies candidate mobile sRNAs, by selecting sRNA clusters
mapped to the foreign genome. It does so by either keeping or removing sRNA
mapped to a given genome. To do so, it requires a common pre-fix across
chromosomes of the same genome. See <code><a href="RNAmergeGenomes.html">RNAmergeGenomes()</a></code> for more
information.</p>
<p>A greater confidence in the mobile sRNA candidates can be achieved by setting
a threshold that considers the number of replicates which contributed to
defining the consensus dicer-call (ie. consensus sRNA classification). This
parameter filters based on the <code>DicerCounts</code> column introduced by the
<code><a href="RNAdicercall.html">RNAdicercall()</a></code> function.</p>
<p><strong>For mRNAseq:</strong>
The function identifies candidate mobile mRNAs, by selecting mRNA molecules
mapped to the foreign genome. It does so by either keeping or removing mRNA
mapped to a given genome. To do so, it requires a common pre-fix across
chromosomes of the same genome. See <code><a href="RNAmergeGenomes.html">RNAmergeGenomes()</a></code> and
<code><a href="RNAmergeAnnotations.html">RNAmergeAnnotations()</a></code> for more information and assistance.</p>
<p>A greater confidence in the mobile mRNA candidates can be achieved by setting
a threshold that considers the number of replicates which contained reads
for the mRNA molecule. This parameter filters based on the <code>SampleCounts</code>
column introduced by the <code><a href="RNAimport.html">RNAimport()</a></code> function.</p>
<p><strong>Statistical Analysis</strong>
The function also allows for statistical analysis based on the results
collect from differential analysis of the total dataset using the function
<code><a href="RNAanalysis.html">RNAanalysis()</a></code>. When <code>statistical=TRUE</code>, the feature is enabled
and selected mobile molecules that meet a specific p-value or
adjusted p-values.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"sRNA_data_consensus"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># vector of control names</span></span></span>
<span class="r-in"><span><span class="va">controls</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"selfgraft_1"</span>, <span class="st">"selfgraft_2"</span> , <span class="st">"selfgraft_3"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Locate potentially mobile sRNA clusters associated to tomato, no</span></span></span>
<span class="r-in"><span><span class="co"># statistical analysis</span></span></span>
<span class="r-in"><span><span class="va">mobile_df1</span> <span class="op">&lt;-</span> <span class="fu">RNAmobile</span><span class="op">(</span>data <span class="op">=</span>  <span class="va">sRNA_data_consensus</span>,</span></span>
<span class="r-in"><span>                    controls <span class="op">=</span> <span class="va">controls</span>,</span></span>
<span class="r-in"><span>                    genome.ID <span class="op">=</span> <span class="st">"SL40"</span>,</span></span>
<span class="r-in"><span>                    task <span class="op">=</span> <span class="st">"keep"</span>,</span></span>
<span class="r-in"><span>                    statistical <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

