<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Merges two genomes annotation files (GFF format) into one
single GFF file saved to desired directory. Adds prefix to each chromosome
name in the genome annotations to ensure each assembly is distinguishable
within the merged file, and saved the individually altered references to
local directory."><title>Merge two GFF genome annotation files — RNAmergeAnnotations • mobileRNA</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Merge two GFF genome annotation files — RNAmergeAnnotations"><meta property="og:description" content="Merges two genomes annotation files (GFF format) into one
single GFF file saved to desired directory. Adds prefix to each chromosome
name in the genome annotations to ensure each assembly is distinguishable
within the merged file, and saved the individually altered references to
local directory."><meta property="og:image" content="https://kjeynescupper.github.io/mobileRNA/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">mobileRNA</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/mobileRNA.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/mobileRNA_local.html">Analyse local sRNA populations using mobileRNA</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/KJeynesCupper/mobileRNA.git/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Merge two GFF genome annotation files</h1>
      <small class="dont-index">Source: <a href="https://github.com/KJeynesCupper/mobileRNA.git/blob/HEAD/R/RNAmergeAnnotations.R" class="external-link"><code>R/RNAmergeAnnotations.R</code></a></small>
      <div class="d-none name"><code>RNAmergeAnnotations.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Merges two genomes annotation files (<code>GFF format</code>) into one
single GFF file saved to desired directory. Adds prefix to each chromosome
name in the genome annotations to ensure each assembly is distinguishable
within the merged file, and saved the individually altered references to
local directory.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RNAmergeAnnotations</span><span class="op">(</span></span>
<span>  <span class="va">annotationA</span>,</span>
<span>  <span class="va">annotationB</span>,</span>
<span>  <span class="va">out_dir</span>,</span>
<span>  abbreviationAnnoA <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>  abbreviationAnnoB <span class="op">=</span> <span class="st">"B"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>annotationA</dt>
<dd><p>GRanges object; a genome annotation assembly file in
GFF format.</p></dd>


<dt>annotationB</dt>
<dd><p>GRanges object; a genome annotation assembly file in
GFF format.</p></dd>


<dt>out_dir</dt>
<dd><p>either a character string or a <code><a href="https://rdrr.io/r/base/connections.html" class="external-link">base::connections()</a></code> open
for writing. Place path to output directory in "", including file output name
with extension of either <code>.gff</code> or <code>.gff3</code>.</p></dd>


<dt>abbreviationAnnoA</dt>
<dd><p>a string placed in "", to replace chromosome names
within <code>annotationA</code>.Default set as "A".</p></dd>


<dt>abbreviationAnnoB</dt>
<dd><p>a string placed in "", to replace chromosome names
within <code>annotationB</code>.Default set as "B".</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>The function output the individually altered annotation files, plus, a merged
annotation file to the global environment and the given directory.
The altered annotation files contain either default or custom alterations to
the chromosome names and are written to the same directory as the merged file,
with the given names addition of <code>annotationA_altered.gff</code> and
<code>annotationB_altered.gff</code>, respectively.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The functions primary goal is to merge two GFF files, however, when
merging genomic files it is critical that the two genomes are distinguishable
by the chromosome names. As a default setting, the function extracts the
chromosome names for a given GFF file and adds an ID before each chromosome
name and removes any periods.</p>
<p>As default, the function will rename the chromosome names in <code>annotationA</code>
to "A". For example, A0, A1, A2 etc. To set a custom chromosome name for
<code>annotationA</code> alter the argument <code>abbreviationAnnoA</code>. While, for
<code>annotationB</code> as default the chromosome names will be named "B", for example,
B0, B1, B2 etc. To set a custom chromosome name for <code>annotationB</code> alter the
argument <code>abbreviationAnnoB</code>.</p>
<p>The merged genome is saved to the specified output directory, and requires
the user to set the name with a <code>gff</code> extension. The user must load the
annotations into the R Global Environment, hence, the alterations made by
this function will alter the variable. The altered annotations will also
be saved to the same directory as supplied for the output file, with a fixed
output name of <code>annotationA_altered.gff</code> and <code>annotationB_altered.gff</code>,
respectively.</p>
<p>IMPORTANT:  The genome reference and annotation of a species
must have chromosomes with matching names. It is critical that if you used
the <code><a href="RNAmergeGenomes.html">RNAmergeGenomes()</a></code> function to to create a merged reference
genome,that you treat the input annotations in the same way.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># import GFF files into R</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Read reference genomes</span></span></span>
<span class="r-in"><span><span class="va">url_remote</span> <span class="op">&lt;-</span> <span class="st">"https://github.com/KJeynesCupper/assemblies/raw/main/"</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">anno1_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">url_remote</span>,<span class="st">"chr12_Eggplant_V4.1_function_IPR_final.gff"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">anno2_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">url_remote</span>, <span class="st">"chr2_ITAG4.0_gene_models.gff"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="co"># Read in GFF files</span></span></span>
<span class="r-in"><span><span class="va">anno1</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html" class="external-link">import</a></span><span class="op">(</span><span class="va">anno1_url</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">anno2</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html" class="external-link">import</a></span><span class="op">(</span><span class="va">anno2_url</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># edit and merge</span></span></span>
<span class="r-in"><span><span class="va">merged_anno</span> <span class="op">&lt;-</span> <span class="fu">RNAmergeAnnotations</span><span class="op">(</span>annotationA <span class="op">=</span> <span class="va">anno1</span>,</span></span>
<span class="r-in"><span>                                 annotationB <span class="op">=</span> <span class="va">anno2</span>,</span></span>
<span class="r-in"><span>             out_dir <span class="op">=</span> <span class="st">"./merged_annotation.gff3"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding abbreviations to chomosome names ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> New annotation file created: ./annotationA_altered.gff</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> New annotation file created: ./annotationB_altered.gff</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Merging altered annotation files ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Set specific pre-fixes:</span></span></span>
<span class="r-in"><span><span class="co">###  annotationA represents Solanum melongena, so abbreviated to `SM`</span></span></span>
<span class="r-in"><span><span class="co">###  annotationB represents Solanum lycopersicum, so abbreviated to `SL`</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">merged_anno_2</span> <span class="op">&lt;-</span> <span class="fu">RNAmergeAnnotations</span><span class="op">(</span>annotationA <span class="op">=</span> <span class="va">anno1</span>,annotationB <span class="op">=</span> <span class="va">anno2</span>,</span></span>
<span class="r-in"><span>              out_dir <span class="op">=</span> <span class="st">"./merged_annotation.gff3"</span>,</span></span>
<span class="r-in"><span>              abbreviationAnnoA <span class="op">=</span> <span class="st">"SM"</span>,abbreviationAnnoB <span class="op">=</span> <span class="st">"SL"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Adding abbreviations to chomosome names ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> New annotation file created: ./annotationA_altered.gff</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> New annotation file created: ./annotationB_altered.gff</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Merging altered annotation files ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Katie Jeynes-Cupper, Marco Catoni.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

